<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en-US"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Ancestral character reconstruction. — ancestral.pml • phangorn</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Ancestral character reconstruction. — ancestral.pml"><meta name="description" content="Marginal reconstruction of the ancestral character states."><meta property="og:description" content="Marginal reconstruction of the ancestral character states."><meta property="og:image" content="https://klausvigo.github.io/phangorn/logo.png"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">phangorn</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.12.1.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/Trees.html">Estimating phylogenetic trees with phangorn</a></li>
    <li><a class="dropdown-item" href="../articles/Morphological.html">Phylogenetic trees from morphological data</a></li>
    <li><a class="dropdown-item" href="../articles/AdvancedFeatures.html">Markov models and transition rate matrices</a></li>
    <li><a class="dropdown-item" href="../articles/Ancestral.html">Ancestral Sequence Reconstruction</a></li>
    <li><a class="dropdown-item" href="../articles/Networx.html">Splits and Networx</a></li>
    <li><a class="dropdown-item" href="../articles/MLbyHand.html">Maximum likelihood by hand</a></li>
    <li><a class="dropdown-item" href="../articles/IntertwiningTreesAndNetworks.html">Intertwining phylogenetic trees and networks: R Example Script</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/KlausVigo/phangorn/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Ancestral character reconstruction.</h1>
      <small class="dont-index">Source: <a href="https://github.com/KlausVigo/phangorn/blob/master/R/ancestral.R" class="external-link"><code>R/ancestral.R</code></a></small>
      <div class="d-none name"><code>ancestral.pml.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Marginal reconstruction of the ancestral character states.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">ancestral.pml</span><span class="op">(</span><span class="va">object</span>, type <span class="op">=</span> <span class="st">"marginal"</span>, return <span class="op">=</span> <span class="st">"prob"</span>, <span class="va">...</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">anc_pml</span><span class="op">(</span><span class="va">object</span>, type <span class="op">=</span> <span class="st">"marginal"</span>, <span class="va">...</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">ancestral.pars</span><span class="op">(</span><span class="va">tree</span>, <span class="va">data</span>, type <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"MPR"</span>, <span class="st">"ACCTRAN"</span>, <span class="st">"POSTORDER"</span><span class="op">)</span>,</span>
<span>  cost <span class="op">=</span> <span class="cn">NULL</span>, return <span class="op">=</span> <span class="st">"prob"</span>, <span class="va">...</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">anc_pars</span><span class="op">(</span><span class="va">tree</span>, <span class="va">data</span>, type <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"MPR"</span>, <span class="st">"ACCTRAN"</span>, <span class="st">"POSTORDER"</span><span class="op">)</span>, cost <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  <span class="va">...</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">pace</span><span class="op">(</span><span class="va">tree</span>, <span class="va">data</span>, type <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"MPR"</span>, <span class="st">"ACCTRAN"</span>, <span class="st">"POSTORDER"</span><span class="op">)</span>, cost <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  return <span class="op">=</span> <span class="st">"prob"</span>, <span class="va">...</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-object">object<a class="anchor" aria-label="anchor" href="#arg-object"></a></dt>
<dd><p>an object of class pml</p></dd>


<dt id="arg-type">type<a class="anchor" aria-label="anchor" href="#arg-type"></a></dt>
<dd><p>method used to assign characters to internal nodes, see details.</p></dd>


<dt id="arg-return">return<a class="anchor" aria-label="anchor" href="#arg-return"></a></dt>
<dd><p>return a <code>phyDat</code> object or matrix of probabilities.</p></dd>


<dt id="arg--">...<a class="anchor" aria-label="anchor" href="#arg--"></a></dt>
<dd><p>Further arguments passed to or from other methods.</p></dd>


<dt id="arg-tree">tree<a class="anchor" aria-label="anchor" href="#arg-tree"></a></dt>
<dd><p>a tree, i.e. an object of class pml</p></dd>


<dt id="arg-data">data<a class="anchor" aria-label="anchor" href="#arg-data"></a></dt>
<dd><p>an object of class phyDat</p></dd>


<dt id="arg-cost">cost<a class="anchor" aria-label="anchor" href="#arg-cost"></a></dt>
<dd><p>A cost matrix for the transitions between two states.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>An object of class ancestral. This is a list containing the tree with
node labels, the original alignment as an <code>phyDat</code> object, a
<code>data.frame</code> containing the probabilities belonging to a state for all
(internal nodes) and the most likely state.</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>The argument "type" defines the criterion to assign the internal nodes. For
<code>ancestral.pml</code> so far "ml and marginal (empirical) "bayes" and for
<code>ancestral.pars</code> "MPR" and "ACCTRAN" are possible.</p>
<p>The function return a list containing the tree with node labels, the original
alignment as an <code>phyDat</code> object, a data.frame containing the
probabilities belonging to a state for all (internal nodes) and the most
likely state. For parsimony and nucleotide data the most likely state might
be ambiguous. For ML this is very unlikely to be the case.</p>
<p>If the input tree does not contain unique node labels the function
<code>ape::MakeNodeLabel</code> is used to create them.</p>
<p>With parsimony reconstruction one has to keep in mind that there will be
often no unique solution.</p>
<p>The functions use the node labels of the provided tree (also if part of the
<code>pml</code> object) if these are unique. Otherwise the function
<code>ape::MakeNodeLabel</code> is used to create them.</p>
<p>For further details see vignette("Ancestral").</p>
    </div>
    <div class="section level2">
    <h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a></h2>
    <p>Felsenstein, J. (2004). <em>Inferring Phylogenies</em>. Sinauer
Associates, Sunderland.</p>
<p>Swofford, D.L., Maddison, W.P. (1987) Reconstructing ancestral character
states under Wagner parsimony. <em>Math. Biosci.</em> <b>87</b>: 199–229</p>
<p>Yang, Z. (2006). <em>Computational Molecular evolution</em>. Oxford University
Press, Oxford.</p>
    </div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p><code><a href="pml.html">pml</a></code>, <code><a href="parsimony.html">parsimony</a></code>, <code><a href="https://rdrr.io/pkg/ape/man/ace.html" class="external-link">ace</a></code>,
<code><a href="plot.ancestral.html">plotAnc</a></code>, <code><a href="latag2n.phyDat.html">latag2n.phyDat</a></code>,
<code><a href="https://rdrr.io/pkg/ape/man/latag2n.html" class="external-link">latag2n</a></code>, <code><a href="gap_as_state.html">gap_as_state</a></code>,
<code><a href="https://rdrr.io/pkg/ape/man/root.html" class="external-link">root</a></code>, <code><a href="https://rdrr.io/pkg/ape/man/makeNodeLabel.html" class="external-link">makeNodeLabel</a></code></p></div>
    </div>
    <div class="section level2">
    <h2 id="author">Author<a class="anchor" aria-label="anchor" href="#author"></a></h2>
    <p>Klaus Schliep <a href="mailto:klaus.schliep@gmail.com">klaus.schliep@gmail.com</a></p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/example.html" class="external-link">example</a></span><span class="op">(</span><span class="va">NJ</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> NJ&gt; data(Laurasiatherian)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> NJ&gt; dm &lt;- dist.ml(Laurasiatherian)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> NJ&gt; tree &lt;- NJ(dm)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> NJ&gt; plot(tree)</span>
<span class="r-plt img"><img src="ancestral.pml-1.png" alt="" width="700" height="433"></span>
<span class="r-in"><span><span class="co"># generate node labels to ensure plotting will work</span></span></span>
<span class="r-in"><span><span class="va">tree</span> <span class="op">&lt;-</span> <span class="fu">makeNodeLabel</span><span class="op">(</span><span class="va">tree</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="pml.html">pml</a></span><span class="op">(</span><span class="va">tree</span>, <span class="va">Laurasiatherian</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">anc.ml</span> <span class="op">&lt;-</span> <span class="fu">anc_pml</span><span class="op">(</span><span class="va">fit</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">anc.p</span> <span class="op">&lt;-</span> <span class="fu">anc_pars</span><span class="op">(</span><span class="va">tree</span>, <span class="va">Laurasiatherian</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co"># plot ancestral sequences at the root</span></span></span>
<span class="r-in"><span><span class="fu"><a href="plot.ancestral.html">plotSeqLogo</a></span><span class="op">(</span> <span class="va">anc.ml</span>, <span class="fl">48</span>, <span class="fl">1</span>, <span class="fl">20</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="ancestral.pml-2.png" alt="" width="700" height="433"></span>
<span class="r-in"><span><span class="fu"><a href="plot.ancestral.html">plotSeqLogo</a></span><span class="op">(</span> <span class="va">anc.p</span>, <span class="fl">48</span>, <span class="fl">1</span>, <span class="fl">20</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="ancestral.pml-3.png" alt="" width="700" height="433"></span>
<span class="r-in"><span><span class="co"># plot the first character</span></span></span>
<span class="r-in"><span><span class="fu"><a href="plot.ancestral.html">plotAnc</a></span><span class="op">(</span><span class="va">anc.ml</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="ancestral.pml-4.png" alt="" width="700" height="433"></span>
<span class="r-in"><span><span class="co"># plot the third character</span></span></span>
<span class="r-in"><span><span class="fu"><a href="plot.ancestral.html">plotAnc</a></span><span class="op">(</span><span class="va">anc.ml</span>, <span class="fl">3</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="ancestral.pml-5.png" alt="" width="700" height="433"></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Klaus Schliep, Emmanuel Paradis, Leonardo de Oliveira Martins, Alastair Potts, Iris Bardel-Kahr.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.2.</p>
</div>

    </footer></div>





  </body></html>

