<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Conversion among Sequence Formats — phyDat • phangorn</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Conversion among Sequence Formats — phyDat"><meta name="description" content="These functions transform several DNA formats into the phyDat format.
allSitePattern generates an alignment of all possible site patterns."><meta property="og:description" content="These functions transform several DNA formats into the phyDat format.
allSitePattern generates an alignment of all possible site patterns."><meta property="og:image" content="https://klausvigo.github.io/phangorn/logo.png"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">phangorn</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">3.0.0.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/Trees.html">Estimating phylogenetic trees with phangorn</a></li>
    <li><a class="dropdown-item" href="../articles/Morphological.html">Phylogenetic trees from morphological data</a></li>
    <li><a class="dropdown-item" href="../articles/AdvancedFeatures.html">Markov models and transition rate matrices</a></li>
    <li><a class="dropdown-item" href="../articles/Ancestral.html">Ancestral Sequence Reconstruction</a></li>
    <li><a class="dropdown-item" href="../articles/Networx.html">Splits and Networx</a></li>
    <li><a class="dropdown-item" href="../articles/MLbyHand.html">Maximum likelihood by hand</a></li>
    <li><a class="dropdown-item" href="../articles/IntertwiningTreesAndNetworks.html">Intertwining phylogenetic trees and networks: R Example Script</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/KlausVigo/phangorn/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Conversion among Sequence Formats</h1>
      <small class="dont-index">Source: <a href="https://github.com/KlausVigo/phangorn/blob/v2.12.0.0-alpha/R/phyDat_conversion.R" class="external-link"><code>R/phyDat_conversion.R</code></a></small>
      <div class="d-none name"><code>as.phyDat.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>These functions transform several DNA formats into the <code>phyDat</code> format.
<code>allSitePattern</code> generates an alignment of all possible site patterns.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">phyDat</span><span class="op">(</span><span class="va">data</span>, type <span class="op">=</span> <span class="st">"DNA"</span>, levels <span class="op">=</span> <span class="cn">NULL</span>, return.index <span class="op">=</span> <span class="cn">TRUE</span>, <span class="va">...</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">as.phyDat</span><span class="op">(</span><span class="va">x</span>, <span class="va">...</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S3 method for class 'factor'</span></span>
<span><span class="fu">as.phyDat</span><span class="op">(</span><span class="va">x</span>, <span class="va">...</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S3 method for class 'DNAbin'</span></span>
<span><span class="fu">as.phyDat</span><span class="op">(</span><span class="va">x</span>, <span class="va">...</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S3 method for class 'AAbin'</span></span>
<span><span class="fu">as.phyDat</span><span class="op">(</span><span class="va">x</span>, <span class="va">...</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S3 method for class 'alignment'</span></span>
<span><span class="fu">as.phyDat</span><span class="op">(</span><span class="va">x</span>, type <span class="op">=</span> <span class="st">"DNA"</span>, <span class="va">...</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">phyDat2alignment</span><span class="op">(</span><span class="va">x</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S3 method for class 'MultipleAlignment'</span></span>
<span><span class="fu">as.phyDat</span><span class="op">(</span><span class="va">x</span>, <span class="va">...</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S3 method for class 'AAStringSet'</span></span>
<span><span class="fu">as.phyDat</span><span class="op">(</span><span class="va">x</span>, <span class="va">...</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S3 method for class 'DNAStringSet'</span></span>
<span><span class="fu">as.phyDat</span><span class="op">(</span><span class="va">x</span>, <span class="va">...</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S3 method for class 'phyDat'</span></span>
<span><span class="fu">as.StringSet</span><span class="op">(</span><span class="va">x</span>, <span class="va">...</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S3 method for class 'phyDat'</span></span>
<span><span class="fu">as.MultipleAlignment</span><span class="op">(</span><span class="va">x</span>, <span class="va">...</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S3 method for class 'phyDat'</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="va">x</span>, allLevels <span class="op">=</span> <span class="cn">TRUE</span>, <span class="va">...</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S3 method for class 'phyDat'</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="va">x</span>, <span class="va">...</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S3 method for class 'phyDat'</span></span>
<span><span class="fu">as.DNAbin</span><span class="op">(</span><span class="va">x</span>, <span class="va">...</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S3 method for class 'phyDat'</span></span>
<span><span class="fu">as.AAbin</span><span class="op">(</span><span class="va">x</span>, <span class="va">...</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">genlight2phyDat</span><span class="op">(</span><span class="va">x</span>, ambiguity <span class="op">=</span> <span class="cn">NA</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">acgt2ry</span><span class="op">(</span><span class="va">obj</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-data">data<a class="anchor" aria-label="anchor" href="#arg-data"></a></dt>
<dd><p>An object containing sequences.</p></dd>


<dt id="arg-type">type<a class="anchor" aria-label="anchor" href="#arg-type"></a></dt>
<dd><p>Type of sequences ("DNA", "AA", "CODON" or "USER").</p></dd>


<dt id="arg-levels">levels<a class="anchor" aria-label="anchor" href="#arg-levels"></a></dt>
<dd><p>Level attributes.</p></dd>


<dt id="arg-return-index">return.index<a class="anchor" aria-label="anchor" href="#arg-return-index"></a></dt>
<dd><p>If TRUE returns a index of the site patterns.</p></dd>


<dt id="arg--">...<a class="anchor" aria-label="anchor" href="#arg--"></a></dt>
<dd><p>further arguments passed to or from other methods.</p></dd>


<dt id="arg-x">x<a class="anchor" aria-label="anchor" href="#arg-x"></a></dt>
<dd><p>An object containing sequences.</p></dd>


<dt id="arg-alllevels">allLevels<a class="anchor" aria-label="anchor" href="#arg-alllevels"></a></dt>
<dd><p>return original data.</p></dd>


<dt id="arg-ambiguity">ambiguity<a class="anchor" aria-label="anchor" href="#arg-ambiguity"></a></dt>
<dd><p>character for ambiguous character and no contrast is
provided.</p></dd>


<dt id="arg-obj">obj<a class="anchor" aria-label="anchor" href="#arg-obj"></a></dt>
<dd><p>as object of class phyDat</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>The functions return an object of class <code>phyDat</code>.</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>If <code>type</code> "USER" a vector has to be give to <code>levels</code>. For example
c("a", "c", "g", "t", "-") would create a data object that can be used in
phylogenetic analysis with gaps as fifth state.  There is a more detailed
example for specifying "USER" defined data formats in the vignette
"phangorn-specials".</p>
<p><code>acgt2ry</code> converts a <code>phyDat</code> object of nucleotides into an binary
ry-coded dataset.</p>
    </div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p><code><a href="https://rdrr.io/pkg/ape/man/DNAbin.html" class="external-link">DNAbin</a></code>, <code><a href="https://rdrr.io/pkg/ape/man/as.alignment.html" class="external-link">as.DNAbin</a></code>,
<code><a href="baseFreq.html">baseFreq</a></code>, <code><a href="baseFreq.html">glance.phyDat</a></code>,
<code><a href="https://rdrr.io/pkg/ape/man/read.dna.html" class="external-link">read.dna</a></code>, <code><a href="https://rdrr.io/pkg/ape/man/read.nexus.data.html" class="external-link">read.nexus.data</a></code>
and the chapter 1 in the <code>vignette("phangorn-specials",
package="phangorn")</code> and the example of <code><a href="pmlMix.html">pmlMix</a></code> for the use of
<code>allSitePattern</code></p></div>
    </div>
    <div class="section level2">
    <h2 id="author">Author<a class="anchor" aria-label="anchor" href="#author"></a></h2>
    <p>Klaus Schliep <a href="mailto:klaus.schliep@gmail.com">klaus.schliep@gmail.com</a></p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">Laurasiatherian</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="va">Laurasiatherian</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "phyDat"</span>
<span class="r-in"><span><span class="va">Laurasiatherian</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 47 sequences with 3179 character and 1605 different site patterns.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> The states are a c g t </span>
<span class="r-in"><span><span class="co"># transform as characters</span></span></span>
<span class="r-in"><span><span class="va">LauraChar</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="va">Laurasiatherian</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co"># and back</span></span></span>
<span class="r-in"><span><span class="va">Laura</span> <span class="op">&lt;-</span> <span class="fu">phyDat</span><span class="op">(</span><span class="va">LauraChar</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/all.equal.html" class="external-link">all.equal</a></span><span class="op">(</span><span class="va">Laurasiatherian</span>, <span class="va">Laura</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] TRUE</span>
<span class="r-in"><span><span class="va">LauraDNAbin</span> <span class="op">&lt;-</span> <span class="fu">as.DNAbin</span><span class="op">(</span><span class="va">Laurasiatherian</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/all.equal.html" class="external-link">all.equal</a></span><span class="op">(</span><span class="va">Laurasiatherian</span>, <span class="fu">as.phyDat</span><span class="op">(</span><span class="va">LauraDNAbin</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] TRUE</span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Klaus Schliep, Emmanuel Paradis, Leonardo de Oliveira Martins, Alastair Potts, Iris Bardel-Kahr.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

    </footer></div>





  </body></html>

