<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en-US"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Pairwise Distances from Sequences — dist.hamming • phangorn</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Pairwise Distances from Sequences — dist.hamming"><meta name="description" content="dist.hamming, dist.ml and dist.logDet compute pairwise
distances for an object of class phyDat. dist.ml uses DNA /
AA sequences to compute distances under different substitution models."><meta property="og:description" content="dist.hamming, dist.ml and dist.logDet compute pairwise
distances for an object of class phyDat. dist.ml uses DNA /
AA sequences to compute distances under different substitution models."><meta property="og:image" content="https://klausvigo.github.io/phangorn/logo.png"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">phangorn</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.12.1.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/Trees.html">Estimating phylogenetic trees with phangorn</a></li>
    <li><a class="dropdown-item" href="../articles/Morphological.html">Phylogenetic trees from morphological data</a></li>
    <li><a class="dropdown-item" href="../articles/AdvancedFeatures.html">Markov models and transition rate matrices</a></li>
    <li><a class="dropdown-item" href="../articles/Ancestral.html">Ancestral Sequence Reconstruction</a></li>
    <li><a class="dropdown-item" href="../articles/Networx.html">Splits and Networx</a></li>
    <li><a class="dropdown-item" href="../articles/MLbyHand.html">Maximum likelihood by hand</a></li>
    <li><a class="dropdown-item" href="../articles/IntertwiningTreesAndNetworks.html">Intertwining phylogenetic trees and networks: R Example Script</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/KlausVigo/phangorn/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Pairwise Distances from Sequences</h1>
      <small class="dont-index">Source: <a href="https://github.com/KlausVigo/phangorn/blob/master/R/distSeq.R" class="external-link"><code>R/distSeq.R</code></a></small>
      <div class="d-none name"><code>dist.hamming.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p><code>dist.hamming</code>, <code>dist.ml</code> and <code>dist.logDet</code> compute pairwise
distances for an object of class <code>phyDat</code>. <code>dist.ml</code> uses DNA /
AA sequences to compute distances under different substitution models.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">dist.hamming</span><span class="op">(</span><span class="va">x</span>, ratio <span class="op">=</span> <span class="cn">TRUE</span>, exclude <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">dist.ml</span><span class="op">(</span><span class="va">x</span>, model <span class="op">=</span> <span class="st">"JC69"</span>, exclude <span class="op">=</span> <span class="st">"none"</span>, bf <span class="op">=</span> <span class="cn">NULL</span>, Q <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  k <span class="op">=</span> <span class="fl">1L</span>, shape <span class="op">=</span> <span class="fl">1</span>, <span class="va">...</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">dist.logDet</span><span class="op">(</span><span class="va">x</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-x">x<a class="anchor" aria-label="anchor" href="#arg-x"></a></dt>
<dd><p>An object of class <code>phyDat</code></p></dd>


<dt id="arg-ratio">ratio<a class="anchor" aria-label="anchor" href="#arg-ratio"></a></dt>
<dd><p>Compute uncorrected ('p') distance or character difference.</p></dd>


<dt id="arg-exclude">exclude<a class="anchor" aria-label="anchor" href="#arg-exclude"></a></dt>
<dd><p>One of "none", "all", "pairwise" indicating whether to delete
the sites with gaps, missing data (or ambiguous states). See details below.</p></dd>


<dt id="arg-model">model<a class="anchor" aria-label="anchor" href="#arg-model"></a></dt>
<dd><p>One of "JC69", "F81" or one of 17 amino acid models see
details.</p></dd>


<dt id="arg-bf">bf<a class="anchor" aria-label="anchor" href="#arg-bf"></a></dt>
<dd><p>A vector of base frequencies.</p></dd>


<dt id="arg-q">Q<a class="anchor" aria-label="anchor" href="#arg-q"></a></dt>
<dd><p>A vector containing the lower triangular part of the rate matrix.</p></dd>


<dt id="arg-k">k<a class="anchor" aria-label="anchor" href="#arg-k"></a></dt>
<dd><p>Number of intervals of the discrete gamma distribution.</p></dd>


<dt id="arg-shape">shape<a class="anchor" aria-label="anchor" href="#arg-shape"></a></dt>
<dd><p>Shape parameter of the gamma distribution.</p></dd>


<dt id="arg--">...<a class="anchor" aria-label="anchor" href="#arg--"></a></dt>
<dd><p>Further arguments passed to or from other methods.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>an object of class <code>dist</code></p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>So far 17 amino acid models are supported ("WAG", "JTT", "LG", "Dayhoff",
"cpREV", "mtmam", "mtArt", "MtZoa", "mtREV24", "VT","RtREV", "HIVw", "HIVb",
"FLU", "Blosum62", "Dayhoff_DCMut" and "JTT_DCMut") and additional rate
matrices and frequencies can be supplied.</p>
<p>The "F81" model uses empirical base frequencies, the "JC69" equal base
frequencies. This is even the case if the data are not nucleotides.</p>
<p>The argument <code>exclude</code> decides how gaps / ambiguous data / missing data
are treated. Usually gaps are treated as ambiguous states, but you can give
gaps its on state <code><a href="gap_as_state.html">gap_as_state</a></code>. <code>exclude="none"</code> keeps all
ambiguous data. The behavior of <code>dist.ml</code> is in this case these same
you would achieve using <code>optim.pml</code> to compute pairwise distances, it
might be a bit odd. <code>exclude="all"</code> removes all sites with ambiguous
states and all gaps if these are coded as ambiguous states. This can lead to
the situation that there only few sites if any fo the alignment left.
Safer is therefore to use <code>exclude="pairwise"</code> which only removes sites
which are ambiguous for each pair of sequences.</p>
    </div>
    <div class="section level2">
    <h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a></h2>
    <p>Lockhart, P. J., Steel, M. A., Hendy, M. D. and Penny, D. (1994)
Recovering evolutionary trees under a more realistic model of sequence
evolution. <em>Molecular Biology and Evolution</em>, <b>11</b>, 605–602.</p>
<p>Jukes TH and Cantor CR (1969). <em>Evolution of Protein Molecules</em>. New
York: Academic Press. 21–132.</p>
<p>McGuire, G., Prentice, M. J. and Wright, F. (1999). Improved error bounds for
genetic distances from DNA sequences. <em>Biometrics</em>, <b>55</b>,
1064–1070.</p>
    </div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p>For more distance methods for nucleotide data see
<code><a href="https://rdrr.io/pkg/ape/man/dist.dna.html" class="external-link">dist.dna</a></code> and <code><a href="dist.p.html">dist.p</a></code> for pairwise
polymorphism p-distances. <code><a href="writeDist.html">writeDist</a></code> for export and import
distances.</p></div>
    </div>
    <div class="section level2">
    <h2 id="author">Author<a class="anchor" aria-label="anchor" href="#author"></a></h2>
    <p>Klaus Schliep <a href="mailto:klaus.schliep@gmail.com">klaus.schliep@gmail.com</a></p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">Laurasiatherian</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">dm1</span> <span class="op">&lt;-</span> <span class="fu">dist.hamming</span><span class="op">(</span><span class="va">Laurasiatherian</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">tree1</span> <span class="op">&lt;-</span> <span class="fu"><a href="NJ.html">NJ</a></span><span class="op">(</span><span class="va">dm1</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">dm2</span> <span class="op">&lt;-</span> <span class="fu">dist.logDet</span><span class="op">(</span><span class="va">Laurasiatherian</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">tree2</span> <span class="op">&lt;-</span> <span class="fu"><a href="NJ.html">NJ</a></span><span class="op">(</span><span class="va">dm2</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="treedist.html">treedist</a></span><span class="op">(</span><span class="va">tree1</span>,<span class="va">tree2</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      symmetric.difference   branch.score.difference           path.difference </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                4.00000000                0.05705091               30.95157508 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> quadratic.path.difference </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                0.80097967 </span>
<span class="r-in"><span><span class="co"># JC model</span></span></span>
<span class="r-in"><span><span class="va">dm3</span> <span class="op">&lt;-</span> <span class="fu">dist.ml</span><span class="op">(</span><span class="va">Laurasiatherian</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">tree3</span> <span class="op">&lt;-</span> <span class="fu"><a href="NJ.html">NJ</a></span><span class="op">(</span><span class="va">dm3</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="treedist.html">treedist</a></span><span class="op">(</span><span class="va">tree1</span>,<span class="va">tree3</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      symmetric.difference   branch.score.difference           path.difference </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                 6.0000000                 0.0412520                30.3644529 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> quadratic.path.difference </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                 0.6106899 </span>
<span class="r-in"><span><span class="co"># F81 + Gamma</span></span></span>
<span class="r-in"><span><span class="va">dm4</span> <span class="op">&lt;-</span> <span class="fu">dist.ml</span><span class="op">(</span><span class="va">Laurasiatherian</span>, model<span class="op">=</span><span class="st">"F81"</span>, k<span class="op">=</span><span class="fl">4</span>, shape<span class="op">=</span><span class="fl">.4</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">tree4</span> <span class="op">&lt;-</span> <span class="fu"><a href="NJ.html">NJ</a></span><span class="op">(</span><span class="va">dm4</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="treedist.html">treedist</a></span><span class="op">(</span><span class="va">tree1</span>,<span class="va">tree4</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      symmetric.difference   branch.score.difference           path.difference </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                12.0000000                 0.1356107                40.7676342 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> quadratic.path.difference </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                 2.0709714 </span>
<span class="r-in"><span><span class="fu"><a href="treedist.html">treedist</a></span><span class="op">(</span><span class="va">tree3</span>,<span class="va">tree4</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      symmetric.difference   branch.score.difference           path.difference </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                8.00000000                0.09494752               39.52214569 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> quadratic.path.difference </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                1.46345381 </span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Klaus Schliep, Emmanuel Paradis, Leonardo de Oliveira Martins, Alastair Potts, Iris Bardel-Kahr.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.2.</p>
</div>

    </footer></div>





  </body></html>

