<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en-US"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Pairwise Polymorphism P-Distances from DNA Sequences — dist.p • phangorn</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Pairwise Polymorphism P-Distances from DNA Sequences — dist.p"><meta name="description" content='This function computes a matrix of pairwise uncorrected polymorphism
p-distances. Polymorphism p-distances include intra-individual site
polymorphisms (2ISPs; e.g. "R") when calculating genetic distances.'><meta property="og:description" content='This function computes a matrix of pairwise uncorrected polymorphism
p-distances. Polymorphism p-distances include intra-individual site
polymorphisms (2ISPs; e.g. "R") when calculating genetic distances.'><meta property="og:image" content="https://klausvigo.github.io/phangorn/logo.png"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">phangorn</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.12.1.3</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/Trees.html">Estimating phylogenetic trees with phangorn</a></li>
    <li><a class="dropdown-item" href="../articles/Morphological.html">Phylogenetic trees from morphological data</a></li>
    <li><a class="dropdown-item" href="../articles/AdvancedFeatures.html">Markov models and transition rate matrices</a></li>
    <li><a class="dropdown-item" href="../articles/Ancestral.html">Ancestral Sequence Reconstruction</a></li>
    <li><a class="dropdown-item" href="../articles/Networx.html">Splits and Networx</a></li>
    <li><a class="dropdown-item" href="../articles/MLbyHand.html">Maximum likelihood by hand</a></li>
    <li><a class="dropdown-item" href="../articles/IntertwiningTreesAndNetworks.html">Intertwining phylogenetic trees and networks: R Example Script</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/KlausVigo/phangorn/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Pairwise Polymorphism P-Distances from DNA Sequences</h1>
      <small class="dont-index">Source: <a href="https://github.com/KlausVigo/phangorn/blob/master/R/dist.p.R" class="external-link"><code>R/dist.p.R</code></a></small>
      <div class="d-none name"><code>dist.p.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This function computes a matrix of pairwise uncorrected polymorphism
p-distances. Polymorphism p-distances include intra-individual site
polymorphisms (2ISPs; e.g. "R") when calculating genetic distances.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">dist.p</span><span class="op">(</span><span class="va">x</span>, cost <span class="op">=</span> <span class="st">"polymorphism"</span>, ignore.indels <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-x">x<a class="anchor" aria-label="anchor" href="#arg-x"></a></dt>
<dd><p>a matrix containing DNA sequences; this must be of class "phyDat"
(use as.phyDat to convert from DNAbin objects).</p></dd>


<dt id="arg-cost">cost<a class="anchor" aria-label="anchor" href="#arg-cost"></a></dt>
<dd><p>A cost matrix or "polymorphism" for a predefined one.</p></dd>


<dt id="arg-ignore-indels">ignore.indels<a class="anchor" aria-label="anchor" href="#arg-ignore-indels"></a></dt>
<dd><p>a logical indicating whether gaps are treated as fifth
state or not. Warning, each gap site is treated as a characters, so an an
indel that spans a number of base positions would be treated as multiple
character states.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>an object of class <code>dist</code>.</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>The polymorphism p-distances (Potts et al. 2014) have been developed to
analyse intra-individual variant polymorphism. For example, the widely used
ribosomal internal transcribed spacer (ITS) region (e.g. Alvarez and Wendel,
2003) consists of 100's to 1000's of units within array across potentially
multiple nucleolus organizing regions (Bailey et al., 2003; Goeker and
Grimm, 2008). This can give rise to intra-individual site polymorphisms
(2ISPs) that can be detected from direct-PCR sequencing or cloning . Clone
consensus sequences (see Goeker and Grimm, 2008) can be analysed with this
function.</p>
    </div>
    <div class="section level2">
    <h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a></h2>
    <p>Alvarez, I., and J. F. Wendel. (2003) Ribosomal ITS sequences
and plant phylogenetic inference. <em> Molecular Phylogenetics and
Evolution</em>, <b>29</b>, 417–434.</p>
<p>Bailey, C. D., T. G. Carr, S. A. Harris, and C. E. Hughes. (2003)
Characterization of angiosperm nrDNA polymorphism, paralogy, and
pseudogenes. <em>Molecular Phylogenetics and Evolution</em> <b>29</b>,
435–455.</p>
<p>Goeker, M., and G. Grimm. (2008) General functions to transform associate
data to host data, and their use in phylogenetic inference from sequences
with intra-individual variability. <em>BMC Evolutionary Biology</em>,
<b>8</b>:86.</p>
<p>Potts, A.J., T.A. Hedderson, and G.W. Grimm. (2014) Constructing phylogenies
in the presence of intra-individual site polymorphisms (2ISPs) with a focus
on the nuclear ribosomal cistron. <em>Systematic Biology</em>, <b>63</b>,
1–16</p>
    </div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p><code><a href="https://rdrr.io/pkg/ape/man/dist.dna.html" class="external-link">dist.dna</a></code>, <code><a href="dist.hamming.html">dist.hamming</a></code></p></div>
    </div>
    <div class="section level2">
    <h2 id="author">Author<a class="anchor" aria-label="anchor" href="#author"></a></h2>
    <p>Klaus Schliep and Alastair Potts</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">Laurasiatherian</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">laura</span> <span class="op">&lt;-</span> <span class="fu">as.DNAbin</span><span class="op">(</span><span class="va">Laurasiatherian</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">dm</span> <span class="op">&lt;-</span> <span class="fu">dist.p</span><span class="op">(</span><span class="va">Laurasiatherian</span>, <span class="st">"polymorphism"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">########################################################</span></span></span>
<span class="r-in"><span><span class="co"># Dealing with indel 2ISPs</span></span></span>
<span class="r-in"><span><span class="co"># These can be coded using an "x" in the alignment. Note</span></span></span>
<span class="r-in"><span><span class="co"># that as.character usage in the read.dna() function.</span></span></span>
<span class="r-in"><span><span class="co">#########################################################</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/cat.html" class="external-link">cat</a></span><span class="op">(</span><span class="st">"3 5"</span>,</span></span>
<span class="r-in"><span>    <span class="st">"No305     ATRA-"</span>,</span></span>
<span class="r-in"><span>    <span class="st">"No304     ATAYX"</span>,</span></span>
<span class="r-in"><span>    <span class="st">"No306     ATAGA"</span>,</span></span>
<span class="r-in"><span>    file <span class="op">=</span> <span class="st">"exdna.txt"</span>, sep <span class="op">=</span> <span class="st">"\n"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">(</span><span class="va">ex.dna</span> <span class="op">&lt;-</span> <span class="fu">read.dna</span><span class="op">(</span><span class="st">"exdna.txt"</span>, format <span class="op">=</span> <span class="st">"sequential"</span>, as.character<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       [,1] [,2] [,3] [,4] [,5]</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> No305 "a"  "t"  "r"  "a"  "-" </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> No304 "a"  "t"  "a"  "y"  "x" </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> No306 "a"  "t"  "a"  "g"  "a" </span>
<span class="r-in"><span><span class="va">dat</span> <span class="op">&lt;-</span> <span class="fu"><a href="as.phyDat.html">phyDat</a></span><span class="op">(</span><span class="va">ex.dna</span>, <span class="st">"USER"</span>, levels<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/vector.html" class="external-link">as.vector</a></span><span class="op">(</span><span class="va">ex.dna</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu">dist.p</span><span class="op">(</span><span class="va">dat</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       No305 No304</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> No304     4      </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> No306     3     4</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/unlink.html" class="external-link">unlink</a></span><span class="op">(</span><span class="st">"exdna.txt"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Klaus Schliep, Emmanuel Paradis, Leonardo de Oliveira Martins, Alastair Potts, Iris Bardel-Kahr.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer></div>





  </body></html>

