<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Translate nucleic acid sequences into codons — dna2codon • phangorn</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Translate nucleic acid sequences into codons — dna2codon"><meta name="description" content="The function transforms dna2codon DNA sequences to codon sequences,
codon2dna transform the other way. dna2codon translates
nucleotide to amino acids using the function trans."><meta property="og:description" content="The function transforms dna2codon DNA sequences to codon sequences,
codon2dna transform the other way. dna2codon translates
nucleotide to amino acids using the function trans."><meta property="og:image" content="https://klausvigo.github.io/phangorn/logo.png"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">phangorn</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">3.0.0.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/Trees.html">Estimating phylogenetic trees with phangorn</a></li>
    <li><a class="dropdown-item" href="../articles/Morphological.html">Phylogenetic trees from morphological data</a></li>
    <li><a class="dropdown-item" href="../articles/AdvancedFeatures.html">Markov models and transition rate matrices</a></li>
    <li><a class="dropdown-item" href="../articles/Ancestral.html">Ancestral Sequence Reconstruction</a></li>
    <li><a class="dropdown-item" href="../articles/Networx.html">Splits and Networx</a></li>
    <li><a class="dropdown-item" href="../articles/MLbyHand.html">Maximum likelihood by hand</a></li>
    <li><a class="dropdown-item" href="../articles/IntertwiningTreesAndNetworks.html">Intertwining phylogenetic trees and networks: R Example Script</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/KlausVigo/phangorn/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Translate nucleic acid sequences into codons</h1>
      <small class="dont-index">Source: <a href="https://github.com/KlausVigo/phangorn/blob/v2.12/R/codon.R" class="external-link"><code>R/codon.R</code></a></small>
      <div class="d-none name"><code>dna2codon.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>The function transforms <code>dna2codon</code> DNA sequences to codon sequences,
<code>codon2dna</code> transform the other way. <code>dna2codon</code> translates
nucleotide to amino acids using the function <code>trans</code>.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">dna2codon</span><span class="op">(</span><span class="va">x</span>, codonstart <span class="op">=</span> <span class="fl">1</span>, code <span class="op">=</span> <span class="fl">1</span>, ambiguity <span class="op">=</span> <span class="st">"---"</span>, <span class="va">...</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">codon2dna</span><span class="op">(</span><span class="va">x</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">dna2aa</span><span class="op">(</span><span class="va">x</span>, codonstart <span class="op">=</span> <span class="fl">1</span>, code <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-x">x<a class="anchor" aria-label="anchor" href="#arg-x"></a></dt>
<dd><p>An object containing sequences.</p></dd>


<dt id="arg-codonstart">codonstart<a class="anchor" aria-label="anchor" href="#arg-codonstart"></a></dt>
<dd><p>an integer giving where to start the translation. This
should be 1, 2, or 3, but larger values are accepted and have for effect to
start the translation further within the sequence.</p></dd>


<dt id="arg-code">code<a class="anchor" aria-label="anchor" href="#arg-code"></a></dt>
<dd><p>The ncbi genetic code number for translation (see details).
By default the standard genetic code is used.</p></dd>


<dt id="arg-ambiguity">ambiguity<a class="anchor" aria-label="anchor" href="#arg-ambiguity"></a></dt>
<dd><p>character for ambiguous character and no contrast is
provided.</p></dd>


<dt id="arg--">...<a class="anchor" aria-label="anchor" href="#arg--"></a></dt>
<dd><p>further arguments passed to or from other methods.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>The functions return an object of class <code>phyDat</code>.</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>The following genetic codes are described here. The number preceding each
corresponds to the code argument.</p><table class="table table"><tr><td>1</td><td>standard</td></tr><tr><td>2</td><td>vertebrate.mitochondrial</td></tr><tr><td>3</td><td>yeast.mitochondrial</td></tr><tr><td>4</td><td>protozoan.mitochondrial+mycoplasma</td></tr><tr><td>5</td><td>invertebrate.mitochondrial</td></tr><tr><td>6</td><td>ciliate+dasycladaceal</td></tr><tr><td>9</td><td>echinoderm+flatworm.mitochondrial</td></tr><tr><td>10</td><td>euplotid</td></tr><tr><td>11</td><td>bacterial+plantplastid</td></tr><tr><td>12</td><td>alternativeyeast</td></tr><tr><td>13</td><td>ascidian.mitochondrial</td></tr><tr><td>14</td><td>alternativeflatworm.mitochondrial</td></tr><tr><td>15</td><td>blepharism</td></tr><tr><td>16</td><td>chlorophycean.mitochondrial</td></tr><tr><td>21</td><td>trematode.mitochondrial</td></tr><tr><td>22</td><td>scenedesmus.mitochondrial</td></tr><tr><td>23</td><td>thraustochytrium.mitochondria</td></tr><tr><td>24</td><td>Pterobranchia.mitochondrial</td></tr><tr><td>25</td><td>CandidateDivision.SR1+Gracilibacteria</td></tr><tr><td>26</td><td>Pachysolen.tannophilus</td></tr></table><p>Alignment gaps and ambiguities are currently ignored and sites containing
these are deleted.</p>
    </div>
    <div class="section level2">
    <h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a></h2>
    <p><a href="https://www.ncbi.nlm.nih.gov/Taxonomy/taxonomyhome.html/index.cgi?chapter=cgencodes" class="external-link">https://www.ncbi.nlm.nih.gov/Taxonomy/taxonomyhome.html/index.cgi?chapter=cgencodes</a></p>
    </div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p><code><a href="https://rdrr.io/pkg/ape/man/trans.html" class="external-link">trans</a></code>, <code><a href="as.phyDat.html">phyDat</a></code> and the chapter 4 in the
<code>vignette("phangorn-specials", package="phangorn")</code></p></div>
    </div>
    <div class="section level2">
    <h2 id="author">Author<a class="anchor" aria-label="anchor" href="#author"></a></h2>
    <p>Klaus Schliep <a href="mailto:klaus.schliep@gmail.com">klaus.schliep@gmail.com</a></p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">Laurasiatherian</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="va">Laurasiatherian</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "phyDat"</span>
<span class="r-in"><span><span class="va">Laurasiatherian</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 47 sequences with 3179 character and 1605 different site patterns.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> The states are a c g t </span>
<span class="r-in"><span><span class="fu">dna2codon</span><span class="op">(</span><span class="va">Laurasiatherian</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 47 sequences with 1059 character and 914 different site patterns.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> The states are aaa aac aag aat aca acc acg act aga agc agg agt ata atc atg att caa cac cag cat cca ccc ccg cct cga cgc cgg cgt cta ctc ctg ctt gaa gac gag gat gca gcc gcg gct gga ggc ggg ggt gta gtc gtg gtt tac tat tca tcc tcg tct tgc tgg tgt tta ttc ttg ttt </span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Klaus Schliep, Emmanuel Paradis, Leonardo de Oliveira Martins, Alastair Potts, Iris Bardel-Kahr.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

    </footer></div>





  </body></html>

