<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en-US"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Package index â€¢ phangorn</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Package index"><meta property="og:image" content="https://klausvigo.github.io/phangorn/logo.png"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">phangorn</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.12.1.3</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/Trees.html">Estimating phylogenetic trees with phangorn</a></li>
    <li><a class="dropdown-item" href="../articles/Morphological.html">Phylogenetic trees from morphological data</a></li>
    <li><a class="dropdown-item" href="../articles/AdvancedFeatures.html">Markov models and transition rate matrices</a></li>
    <li><a class="dropdown-item" href="../articles/Ancestral.html">Ancestral Sequence Reconstruction</a></li>
    <li><a class="dropdown-item" href="../articles/Networx.html">Splits and Networx</a></li>
    <li><a class="dropdown-item" href="../articles/MLbyHand.html">Maximum likelihood by hand</a></li>
    <li><a class="dropdown-item" href="../articles/IntertwiningTreesAndNetworks.html">Intertwining phylogenetic trees and networks: R Example Script</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/KlausVigo/phangorn/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-index">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Package index</h1>
    </div>

    <div class="section level2">
      <h2 id="all-functions">All functions<a class="anchor" aria-label="anchor" href="#all-functions"></a></h2>




    </div><div class="section level2">




      <dl><dt>

          <code><a href="CI.html">CI()</a></code> <code><a href="CI.html">RI()</a></code>

        </dt>
        <dd>Consistency Index and Retention Index</dd>
      </dl><dl><dt>

          <code><a href="Laurasiatherian.html">Laurasiatherian</a></code>

        </dt>
        <dd>Laurasiatherian data (AWCMEE)</dd>
      </dl><dl><dt>

          <code><a href="NJ.html">NJ()</a></code> <code><a href="NJ.html">UNJ()</a></code>

        </dt>
        <dd>Neighbor-Joining</dd>
      </dl><dl><dt>

          <code><a href="SH.test.html">SH.test()</a></code>

        </dt>
        <dd>Shimodaira-Hasegawa Test</dd>
      </dl><dl><dt>

          <code><a href="SOWH.test.html">SOWH.test()</a></code>

        </dt>
        <dd>Swofford-Olsen-Waddell-Hillis Test</dd>
      </dl><dl><dt>

          <code><a href="acctran.html">acctran()</a></code> <code><a href="acctran.html">parsimony_edgelength()</a></code>

        </dt>
        <dd>Assign edge length to tree</dd>
      </dl><dl><dt>

          <code><a href="add.tips.html">add.tips()</a></code>

        </dt>
        <dd>Add tips to a tree</dd>
      </dl><dl><dt>

          <code><a href="addConfidences.html">addConfidences()</a></code> <code><a href="addConfidences.html">presenceAbsence()</a></code> <code><a href="addConfidences.html">createLabel()</a></code>

        </dt>
        <dd>Compare splits and add support values to an object</dd>
      </dl><dl><dt>

          <code><a href="add_ci.html">add_ci()</a></code> <code><a href="add_ci.html">add_boxplot()</a></code>

        </dt>
        <dd>Draw Confidences Intervals on Phylogenies</dd>
      </dl><dl><dt>

          <code><a href="add_edge_length.html">add_edge_length()</a></code>

        </dt>
        <dd>Assign and compute edge lengths from a sample of trees</dd>
      </dl><dl><dt>

          <code><a href="allTrees.html">allTrees()</a></code>

        </dt>
        <dd>Compute all trees topologies.</dd>
      </dl><dl><dt>

          <code><a href="ancestral.pml.html">ancestral.pml()</a></code> <code><a href="ancestral.pml.html">anc_pml()</a></code> <code><a href="ancestral.pml.html">ancestral.pars()</a></code> <code><a href="ancestral.pml.html">anc_pars()</a></code> <code><a href="ancestral.pml.html">pace()</a></code>

        </dt>
        <dd>Ancestral character reconstruction.</dd>
      </dl><dl><dt>

          <code><a href="as.networx.html">as.networx()</a></code>

        </dt>
        <dd>Conversion among phylogenetic network objects</dd>
      </dl><dl><dt>

          <code><a href="as.phyDat.html">phyDat()</a></code> <code><a href="as.phyDat.html">as.phyDat()</a></code> <code><a href="as.phyDat.html">phyDat2alignment()</a></code> <code><a href="as.phyDat.html">as.StringSet()</a></code> <code><a href="as.phyDat.html">as.MultipleAlignment(<i>&lt;phyDat&gt;</i>)</a></code> <code><a href="as.phyDat.html">as.character(<i>&lt;phyDat&gt;</i>)</a></code> <code><a href="as.phyDat.html">as.data.frame(<i>&lt;phyDat&gt;</i>)</a></code> <code><a href="as.phyDat.html">as.DNAbin(<i>&lt;phyDat&gt;</i>)</a></code> <code><a href="as.phyDat.html">as.AAbin(<i>&lt;phyDat&gt;</i>)</a></code> <code><a href="as.phyDat.html">genlight2phyDat()</a></code> <code><a href="as.phyDat.html">acgt2ry()</a></code> <code><a href="as.phyDat.html">unalign()</a></code>

        </dt>
        <dd>Conversion among Sequence Formats</dd>
      </dl><dl><dt>

          <code><a href="as.splits.html">allSplits()</a></code> <code><a href="as.splits.html">allCircularSplits()</a></code> <code><a href="as.splits.html">as.splits()</a></code> <code><a href="as.splits.html">as.matrix(<i>&lt;splits&gt;</i>)</a></code> <code><a href="as.splits.html">as.Matrix(<i>&lt;splits&gt;</i>)</a></code> <code><a href="as.splits.html">print(<i>&lt;splits&gt;</i>)</a></code> <code><a href="as.splits.html">c(<i>&lt;splits&gt;</i>)</a></code> <code><a href="as.splits.html">unique(<i>&lt;splits&gt;</i>)</a></code> <code><a href="as.splits.html">as.prop.part(<i>&lt;splits&gt;</i>)</a></code> <code><a href="as.splits.html">as.bitsplits(<i>&lt;splits&gt;</i>)</a></code> <code><a href="as.splits.html">compatible()</a></code>

        </dt>
        <dd>Splits representation of graphs and trees.</dd>
      </dl><dl><dt>

          <code><a href="bab.html">bab()</a></code>

        </dt>
        <dd>Branch and bound for finding all most parsimonious trees</dd>
      </dl><dl><dt>

          <code><a href="baseFreq.html">baseFreq()</a></code> <code><a href="baseFreq.html">glance(<i>&lt;phyDat&gt;</i>)</a></code> <code><a href="baseFreq.html">composition_test()</a></code> <code><a href="baseFreq.html">summary(<i>&lt;phyDat&gt;</i>)</a></code> <code><a href="baseFreq.html">print(<i>&lt;summary.phyDat&gt;</i>)</a></code>

        </dt>
        <dd>Summaries of alignments</dd>
      </dl><dl><dt>

          <code><a href="bootstrap.pml.html">bootstrap.pml()</a></code> <code><a href="bootstrap.pml.html">bootstrap.phyDat()</a></code>

        </dt>
        <dd>Bootstrap</dd>
      </dl><dl><dt>

          <code><a href="chloroplast.html">chloroplast</a></code>

        </dt>
        <dd>Chloroplast alignment</dd>
      </dl><dl><dt>

          <code><a href="cladePar.html">cladePar()</a></code>

        </dt>
        <dd>Utility function to plot.phylo</dd>
      </dl><dl><dt>

          <code><a href="coalSpeciesTree.html">coalSpeciesTree()</a></code>

        </dt>
        <dd>Species Tree</dd>
      </dl><dl><dt>

          <code><a href="codonTest.html">codonTest()</a></code>

        </dt>
        <dd>codonTest</dd>
      </dl><dl><dt>

          <code><a href="consensusNet.html">consensusNet()</a></code>

        </dt>
        <dd>Computes a consensusNetwork from a list of trees Computes a <code>networx</code> object from a collection of splits.</dd>
      </dl><dl><dt>

          <code><a href="cophenetic.networx.html">cophenetic(<i>&lt;networx&gt;</i>)</a></code>

        </dt>
        <dd>Pairwise Distances from a Phylogenetic Network</dd>
      </dl><dl><dt>

          <code><a href="delta.score.html">delta.score()</a></code>

        </dt>
        <dd>Computes the \(\delta\) score</dd>
      </dl><dl><dt>

          <code><a href="densiTree.html">densiTree()</a></code>

        </dt>
        <dd>Plots a densiTree.</dd>
      </dl><dl><dt>

          <code><a href="designTree.html">designTree()</a></code> <code><a href="designTree.html">nnls.tree()</a></code> <code><a href="designTree.html">nnls.phylo()</a></code> <code><a href="designTree.html">nnls.splits()</a></code> <code><a href="designTree.html">nnls.networx()</a></code> <code><a href="designTree.html">designSplits()</a></code>

        </dt>
        <dd>Compute a design matrix or non-negative LS</dd>
      </dl><dl><dt>

          <code><a href="dfactorial.html">ldfactorial()</a></code> <code><a href="dfactorial.html">dfactorial()</a></code>

        </dt>
        <dd>Arithmetic Operators</dd>
      </dl><dl><dt>

          <code><a href="discrete.gamma.html">discrete.gamma()</a></code> <code><a href="discrete.gamma.html">discrete.beta()</a></code> <code><a href="discrete.gamma.html">plot_gamma_plus_inv()</a></code> <code><a href="discrete.gamma.html">plotRates()</a></code>

        </dt>
        <dd>Discrete Gamma and Beta distribution</dd>
      </dl><dl><dt>

          <code><a href="dist.hamming.html">dist.hamming()</a></code> <code><a href="dist.hamming.html">dist.ml()</a></code> <code><a href="dist.hamming.html">dist.logDet()</a></code>

        </dt>
        <dd>Pairwise Distances from Sequences</dd>
      </dl><dl><dt>

          <code><a href="dist.p.html">dist.p()</a></code>

        </dt>
        <dd>Pairwise Polymorphism P-Distances from DNA Sequences</dd>
      </dl><dl><dt>

          <code><a href="distanceHadamard.html">distanceHadamard()</a></code>

        </dt>
        <dd>Distance Hadamard</dd>
      </dl><dl><dt>

          <code><a href="dna2codon.html">dna2codon()</a></code> <code><a href="dna2codon.html">codon2dna()</a></code> <code><a href="dna2codon.html">dna2aa()</a></code>

        </dt>
        <dd>Translate nucleic acid sequences into codons</dd>
      </dl><dl><dt>

          <code><a href="gap_as_state.html">gap_as_state()</a></code> <code><a href="gap_as_state.html">gap_as_ambiguous()</a></code> <code><a href="gap_as_state.html">has_gap_state()</a></code>

        </dt>
        <dd>Treat gaps as a state</dd>
      </dl><dl><dt>

          <code><a href="getClans.html">getClans()</a></code> <code><a href="getClans.html">getSlices()</a></code> <code><a href="getClans.html">getClips()</a></code> <code><a href="getClans.html">getDiversity()</a></code> <code><a href="getClans.html">summary(<i>&lt;clanistics&gt;</i>)</a></code> <code><a href="getClans.html">diversity()</a></code>

        </dt>
        <dd>Clans, slices and clips</dd>
      </dl><dl><dt>

          <code><a href="hadamard.html">hadamard()</a></code> <code><a href="hadamard.html">fhm()</a></code> <code><a href="hadamard.html">h4st()</a></code> <code><a href="hadamard.html">h2st()</a></code>

        </dt>
        <dd>Hadamard Matrices and Fast Hadamard Multiplication</dd>
      </dl><dl><dt>

          <code><a href="identify.networx.html">identify(<i>&lt;networx&gt;</i>)</a></code>

        </dt>
        <dd>Identify splits in a network</dd>
      </dl><dl><dt>

          <code><a href="image.phyDat.html">image(<i>&lt;phyDat&gt;</i>)</a></code> <code><a href="image.phyDat.html">image(<i>&lt;ancestral&gt;</i>)</a></code>

        </dt>
        <dd>Plot of a Sequence Alignment</dd>
      </dl><dl><dt>

          <code><a href="latag2n.phyDat.html">latag2n.phyDat()</a></code>

        </dt>
        <dd>Replace leading and trailing alignment gaps with an ambiguous state</dd>
      </dl><dl><dt>

          <code><a href="lento.html">lento()</a></code>

        </dt>
        <dd>Lento plot</dd>
      </dl><dl><dt>

          <code><a href="mast.html">mast()</a></code>

        </dt>
        <dd>Maximum agreement subtree</dd>
      </dl><dl><dt>

          <code><a href="maxCladeCred.html">maxCladeCred()</a></code> <code><a href="maxCladeCred.html">mcc()</a></code> <code><a href="maxCladeCred.html">allCompat()</a></code>

        </dt>
        <dd>Maximum clade credibility tree</dd>
      </dl><dl><dt>

          <code><a href="midpoint.html">getRoot()</a></code> <code><a href="midpoint.html">midpoint()</a></code> <code><a href="midpoint.html">pruneTree()</a></code> <code><a href="midpoint.html">keep_as_tip()</a></code>

        </dt>
        <dd>Tree manipulation</dd>
      </dl><dl><dt>

          <code><a href="mites.html">mites</a></code>

        </dt>
        <dd>Morphological characters of Mites (SchÃ¤ffer et al. 2010)</dd>
      </dl><dl><dt>

          <code><a href="modelTest.html">modelTest()</a></code>

        </dt>
        <dd>ModelTest</dd>
      </dl><dl><dt>

          <code><a href="neighborNet.html">neighborNet()</a></code>

        </dt>
        <dd>Computes a neighborNet from a distance matrix</dd>
      </dl><dl><dt>

          <code><a href="nni.html">nni()</a></code> <code><a href="nni.html">rNNI()</a></code> <code><a href="nni.html">rSPR()</a></code>

        </dt>
        <dd>Tree rearrangements.</dd>
      </dl><dl><dt>

          <code><a href="parsimony.html">fitch()</a></code> <code><a href="parsimony.html">random.addition()</a></code> <code><a href="parsimony.html">parsimony()</a></code> <code><a href="parsimony.html">optim.parsimony()</a></code> <code><a href="parsimony.html">pratchet()</a></code> <code><a href="parsimony.html">sankoff()</a></code>

        </dt>
        <dd>Parsimony tree.</dd>
      </dl><dl><dt>

          <code><a href="phyDat.html">print(<i>&lt;phyDat&gt;</i>)</a></code> <code><a href="phyDat.html">cbind(<i>&lt;phyDat&gt;</i>)</a></code> <code><a href="phyDat.html">rbind(<i>&lt;phyDat&gt;</i>)</a></code> <code><a href="phyDat.html">c(<i>&lt;phyDat&gt;</i>)</a></code> <code><a href="phyDat.html">subset(<i>&lt;phyDat&gt;</i>)</a></code> <code><a href="phyDat.html">`[`(<i>&lt;phyDat&gt;</i>)</a></code> <code><a href="phyDat.html">unique(<i>&lt;phyDat&gt;</i>)</a></code> <code><a href="phyDat.html">removeUndeterminedSites()</a></code> <code><a href="phyDat.html">removeAmbiguousSites()</a></code> <code><a href="phyDat.html">allSitePattern()</a></code>

        </dt>
        <dd>Generic functions for class phyDat</dd>
      </dl><dl><dt>

          <code><a href="plot.ancestral.html">plotAnc()</a></code> <code><a href="plot.ancestral.html">plotSeqLogo()</a></code> <code><a href="plot.ancestral.html">add_mutations()</a></code>

        </dt>
        <dd>Plot ancestral character on a tree</dd>
      </dl><dl><dt>

          <code><a href="plot.networx.html">plot(<i>&lt;networx&gt;</i>)</a></code>

        </dt>
        <dd>plot phylogenetic networks</dd>
      </dl><dl><dt>

          <code><a href="plot.pml.html">plot(<i>&lt;pml&gt;</i>)</a></code>

        </dt>
        <dd>Plot phylogeny of a pml object</dd>
      </dl><dl><dt>

          <code><a href="plotBS.html">plotBS()</a></code> <code><a href="plotBS.html">add_support()</a></code>

        </dt>
        <dd>Plotting trees with bootstrap values</dd>
      </dl><dl><dt>

          <code><a href="pml.html">as.pml()</a></code> <code><a href="pml.html">update(<i>&lt;pml&gt;</i>)</a></code> <code><a href="pml.html">pml()</a></code> <code><a href="pml.html">optim.pml()</a></code> <code><a href="pml.html">logLik(<i>&lt;pml&gt;</i>)</a></code> <code><a href="pml.html">anova(<i>&lt;pml&gt;</i>)</a></code> <code><a href="pml.html">vcov(<i>&lt;pml&gt;</i>)</a></code> <code><a href="pml.html">print(<i>&lt;pml&gt;</i>)</a></code>

        </dt>
        <dd>Likelihood of a tree.</dd>
      </dl><dl><dt>

          <code><a href="pml.control.html">pml.control()</a></code> <code><a href="pml.control.html">ratchet.control()</a></code>

        </dt>
        <dd>Auxiliary for Controlling Fitting</dd>
      </dl><dl><dt>

          <code><a href="pml.fit.html">lli()</a></code> <code><a href="pml.fit.html">pml.fit()</a></code>

        </dt>
        <dd>Internal maximum likelihood functions.</dd>
      </dl><dl><dt>

          <code><a href="pmlCluster.html">pmlCluster()</a></code>

        </dt>
        <dd>Stochastic Partitioning</dd>
      </dl><dl><dt>

          <code><a href="pmlMix.html">pmlMix()</a></code>

        </dt>
        <dd>Phylogenetic mixture model</dd>
      </dl><dl><dt>

          <code><a href="pmlPart.html">multiphyDat2pmlPart()</a></code> <code><a href="pmlPart.html">pmlPart2multiPhylo()</a></code> <code><a href="pmlPart.html">pmlPart()</a></code>

        </dt>
        <dd>Partition model.</dd>
      </dl><dl><dt>

          <code><a href="pml_bb.html">pml_bb()</a></code>

        </dt>
        <dd>Likelihood of a tree.</dd>
      </dl><dl><dt>

          <code><a href="read.nexus.partitions.html">read.nexus.partitions()</a></code>

        </dt>
        <dd>Function to import partitioned data from nexus files</dd>
      </dl><dl><dt>

          <code><a href="read.nexus.splits.html">read.nexus.splits()</a></code> <code><a href="read.nexus.splits.html">write.nexus.splits()</a></code> <code><a href="read.nexus.splits.html">write.nexus.networx()</a></code> <code><a href="read.nexus.splits.html">read.nexus.networx()</a></code> <code><a href="read.nexus.splits.html">write.splits()</a></code>

        </dt>
        <dd>Function to import and export splits and networks</dd>
      </dl><dl><dt>

          <code><a href="read.phyDat.html">read.phyDat()</a></code> <code><a href="read.phyDat.html">write.phyDat()</a></code>

        </dt>
        <dd>Import and export sequence alignments</dd>
      </dl><dl><dt>

          <code><a href="simSeq.html">simSeq()</a></code>

        </dt>
        <dd>Simulate sequences.</dd>
      </dl><dl><dt>

          <code><a href="splitsNetwork.html">splitsNetwork()</a></code>

        </dt>
        <dd>Phylogenetic Network</dd>
      </dl><dl><dt>

          <code><a href="superTree.html">superTree()</a></code>

        </dt>
        <dd>Super Tree methods</dd>
      </dl><dl><dt>

          <code><a href="terraces.html">terraces()</a></code>

        </dt>
        <dd>Discrete Gamma and Beta distribution</dd>
      </dl><dl><dt>

          <code><a href="transferBootstrap.html">transferBootstrap()</a></code>

        </dt>
        <dd>Transfer Bootstrap</dd>
      </dl><dl><dt>

          <code><a href="treedist.html">treedist()</a></code> <code><a href="treedist.html">sprdist()</a></code> <code><a href="treedist.html">SPR.dist()</a></code> <code><a href="treedist.html">RF.dist()</a></code> <code><a href="treedist.html">wRF.dist()</a></code> <code><a href="treedist.html">KF.dist()</a></code> <code><a href="treedist.html">path.dist()</a></code>

        </dt>
        <dd>Distances between trees</dd>
      </dl><dl><dt>

          <code><a href="upgma.html">upgma()</a></code> <code><a href="upgma.html">wpgma()</a></code> <code><a href="upgma.html">supgma()</a></code>

        </dt>
        <dd>UPGMA, WPGMA and sUPGMA</dd>
      </dl><dl><dt>

          <code><a href="write.ancestral.html">write.ancestral()</a></code> <code><a href="write.ancestral.html">as.ancestral()</a></code> <code><a href="write.ancestral.html">print(<i>&lt;ancestral&gt;</i>)</a></code> <code><a href="write.ancestral.html">as.phyDat(<i>&lt;ancestral&gt;</i>)</a></code> <code><a href="write.ancestral.html">as.data.frame(<i>&lt;ancestral&gt;</i>)</a></code>

        </dt>
        <dd>Export and convenience functions for ancestral reconstructions</dd>
      </dl><dl><dt>

          <code><a href="write.pml.html">write.pml()</a></code>

        </dt>
        <dd>Export pml objects</dd>
      </dl><dl><dt>

          <code><a href="writeDist.html">writeDist()</a></code> <code><a href="writeDist.html">write.nexus.dist()</a></code> <code><a href="writeDist.html">readDist()</a></code> <code><a href="writeDist.html">read.nexus.dist()</a></code> <code><a href="writeDist.html">unique(<i>&lt;dist&gt;</i>)</a></code>

        </dt>
        <dd>Writing and reading distances in phylip and nexus format</dd>
      </dl><dl><dt>

          <code><a href="yeast.html">yeast</a></code>

        </dt>
        <dd>Yeast alignment (Rokas et al.)</dd>
      </dl></div>
  </main></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Klaus Schliep, Emmanuel Paradis, Leonardo de Oliveira Martins, Alastair Potts, Iris Bardel-Kahr.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer></div>





  </body></html>

