# Package index

## All functions

- [`CI()`](https://klausvigo.github.io/phangorn/reference/CI.md)
  [`RI()`](https://klausvigo.github.io/phangorn/reference/CI.md) :
  Consistency Index and Retention Index

- [`Laurasiatherian`](https://klausvigo.github.io/phangorn/reference/Laurasiatherian.md)
  : Laurasiatherian data (AWCMEE)

- [`NJ()`](https://klausvigo.github.io/phangorn/reference/NJ.md)
  [`UNJ()`](https://klausvigo.github.io/phangorn/reference/NJ.md) :
  Neighbor-Joining

- [`SH.test()`](https://klausvigo.github.io/phangorn/reference/SH.test.md)
  : Shimodaira-Hasegawa Test

- [`SOWH.test()`](https://klausvigo.github.io/phangorn/reference/SOWH.test.md)
  : Swofford-Olsen-Waddell-Hillis Test

- [`acctran()`](https://klausvigo.github.io/phangorn/reference/acctran.md)
  [`parsimony_edgelength()`](https://klausvigo.github.io/phangorn/reference/acctran.md)
  : Assign edge length to tree

- [`add.tips()`](https://klausvigo.github.io/phangorn/reference/add.tips.md)
  : Add tips to a tree

- [`addConfidences()`](https://klausvigo.github.io/phangorn/reference/addConfidences.md)
  [`presenceAbsence()`](https://klausvigo.github.io/phangorn/reference/addConfidences.md)
  [`createLabel()`](https://klausvigo.github.io/phangorn/reference/addConfidences.md)
  : Compare splits and add support values to an object

- [`add_ci()`](https://klausvigo.github.io/phangorn/reference/add_ci.md)
  [`add_boxplot()`](https://klausvigo.github.io/phangorn/reference/add_ci.md)
  : Draw Confidences Intervals on Phylogenies

- [`add_edge_length()`](https://klausvigo.github.io/phangorn/reference/add_edge_length.md)
  : Assign and compute edge lengths from a sample of trees

- [`allTrees()`](https://klausvigo.github.io/phangorn/reference/allTrees.md)
  : Compute all trees topologies.

- [`ancestral.pml()`](https://klausvigo.github.io/phangorn/reference/ancestral.pml.md)
  [`anc_pml()`](https://klausvigo.github.io/phangorn/reference/ancestral.pml.md)
  [`ancestral.pars()`](https://klausvigo.github.io/phangorn/reference/ancestral.pml.md)
  [`anc_pars()`](https://klausvigo.github.io/phangorn/reference/ancestral.pml.md)
  [`pace()`](https://klausvigo.github.io/phangorn/reference/ancestral.pml.md)
  : Ancestral character reconstruction.

- [`as.networx()`](https://klausvigo.github.io/phangorn/reference/as.networx.md)
  : Conversion among phylogenetic network objects

- [`phyDat()`](https://klausvigo.github.io/phangorn/reference/as.phyDat.md)
  [`as.phyDat()`](https://klausvigo.github.io/phangorn/reference/as.phyDat.md)
  [`phyDat2alignment()`](https://klausvigo.github.io/phangorn/reference/as.phyDat.md)
  [`as.StringSet()`](https://klausvigo.github.io/phangorn/reference/as.phyDat.md)
  [`as.MultipleAlignment(`*`<phyDat>`*`)`](https://klausvigo.github.io/phangorn/reference/as.phyDat.md)
  [`as.character(`*`<phyDat>`*`)`](https://klausvigo.github.io/phangorn/reference/as.phyDat.md)
  [`as.data.frame(`*`<phyDat>`*`)`](https://klausvigo.github.io/phangorn/reference/as.phyDat.md)
  [`as.DNAbin(`*`<phyDat>`*`)`](https://klausvigo.github.io/phangorn/reference/as.phyDat.md)
  [`as.AAbin(`*`<phyDat>`*`)`](https://klausvigo.github.io/phangorn/reference/as.phyDat.md)
  [`genlight2phyDat()`](https://klausvigo.github.io/phangorn/reference/as.phyDat.md)
  [`acgt2ry()`](https://klausvigo.github.io/phangorn/reference/as.phyDat.md)
  [`unalign()`](https://klausvigo.github.io/phangorn/reference/as.phyDat.md)
  : Conversion among Sequence Formats

- [`allSplits()`](https://klausvigo.github.io/phangorn/reference/as.splits.md)
  [`allCircularSplits()`](https://klausvigo.github.io/phangorn/reference/as.splits.md)
  [`as.splits()`](https://klausvigo.github.io/phangorn/reference/as.splits.md)
  [`as.matrix(`*`<splits>`*`)`](https://klausvigo.github.io/phangorn/reference/as.splits.md)
  [`as.Matrix(`*`<splits>`*`)`](https://klausvigo.github.io/phangorn/reference/as.splits.md)
  [`print(`*`<splits>`*`)`](https://klausvigo.github.io/phangorn/reference/as.splits.md)
  [`c(`*`<splits>`*`)`](https://klausvigo.github.io/phangorn/reference/as.splits.md)
  [`unique(`*`<splits>`*`)`](https://klausvigo.github.io/phangorn/reference/as.splits.md)
  [`as.prop.part(`*`<splits>`*`)`](https://klausvigo.github.io/phangorn/reference/as.splits.md)
  [`as.bitsplits(`*`<splits>`*`)`](https://klausvigo.github.io/phangorn/reference/as.splits.md)
  [`compatible()`](https://klausvigo.github.io/phangorn/reference/as.splits.md)
  : Splits representation of graphs and trees.

- [`bab()`](https://klausvigo.github.io/phangorn/reference/bab.md) :
  Branch and bound for finding all most parsimonious trees

- [`baseFreq()`](https://klausvigo.github.io/phangorn/reference/baseFreq.md)
  [`glance(`*`<phyDat>`*`)`](https://klausvigo.github.io/phangorn/reference/baseFreq.md)
  [`composition_test()`](https://klausvigo.github.io/phangorn/reference/baseFreq.md)
  [`summary(`*`<phyDat>`*`)`](https://klausvigo.github.io/phangorn/reference/baseFreq.md)
  [`print(`*`<summary.phyDat>`*`)`](https://klausvigo.github.io/phangorn/reference/baseFreq.md)
  : Summaries of alignments

- [`bootstrap.pml()`](https://klausvigo.github.io/phangorn/reference/bootstrap.pml.md)
  [`bootstrap.phyDat()`](https://klausvigo.github.io/phangorn/reference/bootstrap.pml.md)
  : Bootstrap

- [`chloroplast`](https://klausvigo.github.io/phangorn/reference/chloroplast.md)
  : Chloroplast alignment

- [`cladePar()`](https://klausvigo.github.io/phangorn/reference/cladePar.md)
  : Utility function to plot.phylo

- [`coalSpeciesTree()`](https://klausvigo.github.io/phangorn/reference/coalSpeciesTree.md)
  : Species Tree

- [`codonTest()`](https://klausvigo.github.io/phangorn/reference/codonTest.md)
  : codonTest

- [`consensusNet()`](https://klausvigo.github.io/phangorn/reference/consensusNet.md)
  :

  Computes a consensusNetwork from a list of trees Computes a `networx`
  object from a collection of splits.

- [`cophenetic(`*`<networx>`*`)`](https://klausvigo.github.io/phangorn/reference/cophenetic.networx.md)
  : Pairwise Distances from a Phylogenetic Network

- [`delta.score()`](https://klausvigo.github.io/phangorn/reference/delta.score.md)
  : Computes the \\\delta\\ score

- [`densiTree()`](https://klausvigo.github.io/phangorn/reference/densiTree.md)
  : Plots a densiTree.

- [`designTree()`](https://klausvigo.github.io/phangorn/reference/designTree.md)
  [`nnls.tree()`](https://klausvigo.github.io/phangorn/reference/designTree.md)
  [`nnls.phylo()`](https://klausvigo.github.io/phangorn/reference/designTree.md)
  [`nnls.splits()`](https://klausvigo.github.io/phangorn/reference/designTree.md)
  [`nnls.networx()`](https://klausvigo.github.io/phangorn/reference/designTree.md)
  [`designSplits()`](https://klausvigo.github.io/phangorn/reference/designTree.md)
  : Compute a design matrix or non-negative LS

- [`ldfactorial()`](https://klausvigo.github.io/phangorn/reference/dfactorial.md)
  [`dfactorial()`](https://klausvigo.github.io/phangorn/reference/dfactorial.md)
  : Arithmetic Operators

- [`discrete.gamma()`](https://klausvigo.github.io/phangorn/reference/discrete.gamma.md)
  [`discrete.beta()`](https://klausvigo.github.io/phangorn/reference/discrete.gamma.md)
  [`plot_gamma_plus_inv()`](https://klausvigo.github.io/phangorn/reference/discrete.gamma.md)
  [`plotRates()`](https://klausvigo.github.io/phangorn/reference/discrete.gamma.md)
  : Discrete Gamma and Beta distribution

- [`dist.hamming()`](https://klausvigo.github.io/phangorn/reference/dist.hamming.md)
  [`dist.ml()`](https://klausvigo.github.io/phangorn/reference/dist.hamming.md)
  [`dist.logDet()`](https://klausvigo.github.io/phangorn/reference/dist.hamming.md)
  : Pairwise Distances from Sequences

- [`dist.p()`](https://klausvigo.github.io/phangorn/reference/dist.p.md)
  : Pairwise Polymorphism P-Distances from DNA Sequences

- [`distanceHadamard()`](https://klausvigo.github.io/phangorn/reference/distanceHadamard.md)
  : Distance Hadamard

- [`dna2codon()`](https://klausvigo.github.io/phangorn/reference/dna2codon.md)
  [`codon2dna()`](https://klausvigo.github.io/phangorn/reference/dna2codon.md)
  [`dna2aa()`](https://klausvigo.github.io/phangorn/reference/dna2codon.md)
  : Translate nucleic acid sequences into codons

- [`gap_as_state()`](https://klausvigo.github.io/phangorn/reference/gap_as_state.md)
  [`gap_as_ambiguous()`](https://klausvigo.github.io/phangorn/reference/gap_as_state.md)
  [`has_gap_state()`](https://klausvigo.github.io/phangorn/reference/gap_as_state.md)
  : Treat gaps as a state

- [`getClans()`](https://klausvigo.github.io/phangorn/reference/getClans.md)
  [`getSlices()`](https://klausvigo.github.io/phangorn/reference/getClans.md)
  [`getClips()`](https://klausvigo.github.io/phangorn/reference/getClans.md)
  [`getDiversity()`](https://klausvigo.github.io/phangorn/reference/getClans.md)
  [`summary(`*`<clanistics>`*`)`](https://klausvigo.github.io/phangorn/reference/getClans.md)
  [`diversity()`](https://klausvigo.github.io/phangorn/reference/getClans.md)
  : Clans, slices and clips

- [`hadamard()`](https://klausvigo.github.io/phangorn/reference/hadamard.md)
  [`fhm()`](https://klausvigo.github.io/phangorn/reference/hadamard.md)
  [`h4st()`](https://klausvigo.github.io/phangorn/reference/hadamard.md)
  [`h2st()`](https://klausvigo.github.io/phangorn/reference/hadamard.md)
  : Hadamard Matrices and Fast Hadamard Multiplication

- [`identify(`*`<networx>`*`)`](https://klausvigo.github.io/phangorn/reference/identify.networx.md)
  : Identify splits in a network

- [`image(`*`<phyDat>`*`)`](https://klausvigo.github.io/phangorn/reference/image.phyDat.md)
  [`image(`*`<ancestral>`*`)`](https://klausvigo.github.io/phangorn/reference/image.phyDat.md)
  : Plot of a Sequence Alignment

- [`latag2n.phyDat()`](https://klausvigo.github.io/phangorn/reference/latag2n.phyDat.md)
  : Replace leading and trailing alignment gaps with an ambiguous state

- [`lento()`](https://klausvigo.github.io/phangorn/reference/lento.md) :
  Lento plot

- [`mast()`](https://klausvigo.github.io/phangorn/reference/mast.md) :
  Maximum agreement subtree

- [`maxCladeCred()`](https://klausvigo.github.io/phangorn/reference/maxCladeCred.md)
  [`mcc()`](https://klausvigo.github.io/phangorn/reference/maxCladeCred.md)
  [`allCompat()`](https://klausvigo.github.io/phangorn/reference/maxCladeCred.md)
  : Maximum clade credibility tree

- [`getRoot()`](https://klausvigo.github.io/phangorn/reference/midpoint.md)
  [`midpoint()`](https://klausvigo.github.io/phangorn/reference/midpoint.md)
  [`pruneTree()`](https://klausvigo.github.io/phangorn/reference/midpoint.md)
  [`keep_as_tip()`](https://klausvigo.github.io/phangorn/reference/midpoint.md)
  : Tree manipulation

- [`mites`](https://klausvigo.github.io/phangorn/reference/mites.md) :
  Morphological characters of Mites (Sch√§ffer et al. 2010)

- [`modelTest()`](https://klausvigo.github.io/phangorn/reference/modelTest.md)
  : ModelTest

- [`neighborNet()`](https://klausvigo.github.io/phangorn/reference/neighborNet.md)
  : Computes a neighborNet from a distance matrix

- [`nni()`](https://klausvigo.github.io/phangorn/reference/nni.md)
  [`rNNI()`](https://klausvigo.github.io/phangorn/reference/nni.md)
  [`rSPR()`](https://klausvigo.github.io/phangorn/reference/nni.md) :
  Tree rearrangements.

- [`fitch()`](https://klausvigo.github.io/phangorn/reference/parsimony.md)
  [`random.addition()`](https://klausvigo.github.io/phangorn/reference/parsimony.md)
  [`parsimony()`](https://klausvigo.github.io/phangorn/reference/parsimony.md)
  [`optim.parsimony()`](https://klausvigo.github.io/phangorn/reference/parsimony.md)
  [`pratchet()`](https://klausvigo.github.io/phangorn/reference/parsimony.md)
  [`sankoff()`](https://klausvigo.github.io/phangorn/reference/parsimony.md)
  : Parsimony tree.

- [`print(`*`<phyDat>`*`)`](https://klausvigo.github.io/phangorn/reference/phyDat.md)
  [`cbind(`*`<phyDat>`*`)`](https://klausvigo.github.io/phangorn/reference/phyDat.md)
  [`rbind(`*`<phyDat>`*`)`](https://klausvigo.github.io/phangorn/reference/phyDat.md)
  [`c(`*`<phyDat>`*`)`](https://klausvigo.github.io/phangorn/reference/phyDat.md)
  [`subset(`*`<phyDat>`*`)`](https://klausvigo.github.io/phangorn/reference/phyDat.md)
  [`` `[`( ``*`<phyDat>`*`)`](https://klausvigo.github.io/phangorn/reference/phyDat.md)
  [`unique(`*`<phyDat>`*`)`](https://klausvigo.github.io/phangorn/reference/phyDat.md)
  [`removeUndeterminedSites()`](https://klausvigo.github.io/phangorn/reference/phyDat.md)
  [`removeAmbiguousSites()`](https://klausvigo.github.io/phangorn/reference/phyDat.md)
  [`allSitePattern()`](https://klausvigo.github.io/phangorn/reference/phyDat.md)
  : Generic functions for class phyDat

- [`plotAnc()`](https://klausvigo.github.io/phangorn/reference/plot.ancestral.md)
  [`plotSeqLogo()`](https://klausvigo.github.io/phangorn/reference/plot.ancestral.md)
  [`add_mutations()`](https://klausvigo.github.io/phangorn/reference/plot.ancestral.md)
  : Plot ancestral character on a tree

- [`plot(`*`<networx>`*`)`](https://klausvigo.github.io/phangorn/reference/plot.networx.md)
  : plot phylogenetic networks

- [`plot(`*`<pml>`*`)`](https://klausvigo.github.io/phangorn/reference/plot.pml.md)
  : Plot phylogeny of a pml object

- [`plotBS()`](https://klausvigo.github.io/phangorn/reference/plotBS.md)
  [`add_support()`](https://klausvigo.github.io/phangorn/reference/plotBS.md)
  : Plotting trees with bootstrap values

- [`as.pml()`](https://klausvigo.github.io/phangorn/reference/pml.md)
  [`update(`*`<pml>`*`)`](https://klausvigo.github.io/phangorn/reference/pml.md)
  [`pml()`](https://klausvigo.github.io/phangorn/reference/pml.md)
  [`optim.pml()`](https://klausvigo.github.io/phangorn/reference/pml.md)
  [`logLik(`*`<pml>`*`)`](https://klausvigo.github.io/phangorn/reference/pml.md)
  [`anova(`*`<pml>`*`)`](https://klausvigo.github.io/phangorn/reference/pml.md)
  [`vcov(`*`<pml>`*`)`](https://klausvigo.github.io/phangorn/reference/pml.md)
  [`print(`*`<pml>`*`)`](https://klausvigo.github.io/phangorn/reference/pml.md)
  : Likelihood of a tree.

- [`pml.control()`](https://klausvigo.github.io/phangorn/reference/pml.control.md)
  [`ratchet.control()`](https://klausvigo.github.io/phangorn/reference/pml.control.md)
  : Auxiliary for Controlling Fitting

- [`lli()`](https://klausvigo.github.io/phangorn/reference/pml.fit.md)
  [`pml.fit()`](https://klausvigo.github.io/phangorn/reference/pml.fit.md)
  : Internal maximum likelihood functions.

- [`pmlCluster()`](https://klausvigo.github.io/phangorn/reference/pmlCluster.md)
  : Stochastic Partitioning

- [`pmlMix()`](https://klausvigo.github.io/phangorn/reference/pmlMix.md)
  : Phylogenetic mixture model

- [`multiphyDat2pmlPart()`](https://klausvigo.github.io/phangorn/reference/pmlPart.md)
  [`pmlPart2multiPhylo()`](https://klausvigo.github.io/phangorn/reference/pmlPart.md)
  [`pmlPart()`](https://klausvigo.github.io/phangorn/reference/pmlPart.md)
  : Partition model.

- [`pml_bb()`](https://klausvigo.github.io/phangorn/reference/pml_bb.md)
  : Likelihood of a tree.

- [`read.nexus.partitions()`](https://klausvigo.github.io/phangorn/reference/read.nexus.partitions.md)
  : Function to import partitioned data from nexus files

- [`read.nexus.splits()`](https://klausvigo.github.io/phangorn/reference/read.nexus.splits.md)
  [`write.nexus.splits()`](https://klausvigo.github.io/phangorn/reference/read.nexus.splits.md)
  [`write.nexus.networx()`](https://klausvigo.github.io/phangorn/reference/read.nexus.splits.md)
  [`read.nexus.networx()`](https://klausvigo.github.io/phangorn/reference/read.nexus.splits.md)
  [`write.splits()`](https://klausvigo.github.io/phangorn/reference/read.nexus.splits.md)
  : Function to import and export splits and networks

- [`read.phyDat()`](https://klausvigo.github.io/phangorn/reference/read.phyDat.md)
  [`write.phyDat()`](https://klausvigo.github.io/phangorn/reference/read.phyDat.md)
  : Import and export sequence alignments

- [`simSeq()`](https://klausvigo.github.io/phangorn/reference/simSeq.md)
  : Simulate sequences.

- [`splitsNetwork()`](https://klausvigo.github.io/phangorn/reference/splitsNetwork.md)
  : Phylogenetic Network

- [`superTree()`](https://klausvigo.github.io/phangorn/reference/superTree.md)
  : Super Tree methods

- [`terraces()`](https://klausvigo.github.io/phangorn/reference/terraces.md)
  : Explore likelihood parsimony surface

- [`transferBootstrap()`](https://klausvigo.github.io/phangorn/reference/transferBootstrap.md)
  : Transfer Bootstrap

- [`treedist()`](https://klausvigo.github.io/phangorn/reference/treedist.md)
  [`sprdist()`](https://klausvigo.github.io/phangorn/reference/treedist.md)
  [`SPR.dist()`](https://klausvigo.github.io/phangorn/reference/treedist.md)
  [`RF.dist()`](https://klausvigo.github.io/phangorn/reference/treedist.md)
  [`wRF.dist()`](https://klausvigo.github.io/phangorn/reference/treedist.md)
  [`KF.dist()`](https://klausvigo.github.io/phangorn/reference/treedist.md)
  [`path.dist()`](https://klausvigo.github.io/phangorn/reference/treedist.md)
  : Distances between trees

- [`upgma()`](https://klausvigo.github.io/phangorn/reference/upgma.md)
  [`wpgma()`](https://klausvigo.github.io/phangorn/reference/upgma.md)
  [`supgma()`](https://klausvigo.github.io/phangorn/reference/upgma.md)
  : UPGMA, WPGMA and sUPGMA

- [`write.ancestral()`](https://klausvigo.github.io/phangorn/reference/write.ancestral.md)
  [`as.ancestral()`](https://klausvigo.github.io/phangorn/reference/write.ancestral.md)
  [`print(`*`<ancestral>`*`)`](https://klausvigo.github.io/phangorn/reference/write.ancestral.md)
  [`as.phyDat(`*`<ancestral>`*`)`](https://klausvigo.github.io/phangorn/reference/write.ancestral.md)
  [`as.data.frame(`*`<ancestral>`*`)`](https://klausvigo.github.io/phangorn/reference/write.ancestral.md)
  : Export and convenience functions for ancestral reconstructions

- [`write.pml()`](https://klausvigo.github.io/phangorn/reference/write.pml.md)
  : Export pml objects

- [`writeDist()`](https://klausvigo.github.io/phangorn/reference/writeDist.md)
  [`write.nexus.dist()`](https://klausvigo.github.io/phangorn/reference/writeDist.md)
  [`readDist()`](https://klausvigo.github.io/phangorn/reference/writeDist.md)
  [`read.nexus.dist()`](https://klausvigo.github.io/phangorn/reference/writeDist.md)
  [`unique(`*`<dist>`*`)`](https://klausvigo.github.io/phangorn/reference/writeDist.md)
  : Writing and reading distances in phylip and nexus format

- [`yeast`](https://klausvigo.github.io/phangorn/reference/yeast.md) :
  Yeast alignment (Rokas et al.)
