<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en-US"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Generic functions for class phyDat — print.phyDat • phangorn</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Generic functions for class phyDat — print.phyDat"><meta name="description" content="These functions help to manipulate alignments of class phyDat."><meta property="og:description" content="These functions help to manipulate alignments of class phyDat."><meta property="og:image" content="https://klausvigo.github.io/phangorn/logo.png"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">phangorn</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.12.1.3</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/Trees.html">Estimating phylogenetic trees with phangorn</a></li>
    <li><a class="dropdown-item" href="../articles/Morphological.html">Phylogenetic trees from morphological data</a></li>
    <li><a class="dropdown-item" href="../articles/AdvancedFeatures.html">Markov models and transition rate matrices</a></li>
    <li><a class="dropdown-item" href="../articles/Ancestral.html">Ancestral Sequence Reconstruction</a></li>
    <li><a class="dropdown-item" href="../articles/Networx.html">Splits and Networx</a></li>
    <li><a class="dropdown-item" href="../articles/MLbyHand.html">Maximum likelihood by hand</a></li>
    <li><a class="dropdown-item" href="../articles/IntertwiningTreesAndNetworks.html">Intertwining phylogenetic trees and networks: R Example Script</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/KlausVigo/phangorn/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Generic functions for class phyDat</h1>
      <small class="dont-index">Source: <a href="https://github.com/KlausVigo/phangorn/blob/main/R/phyDat.R" class="external-link"><code>R/phyDat.R</code></a></small>
      <div class="d-none name"><code>phyDat.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>These functions help to manipulate alignments of class phyDat.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="co"># S3 method for class 'phyDat'</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">x</span>, <span class="va">...</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S3 method for class 'phyDat'</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="va">...</span>, gaps <span class="op">=</span> <span class="st">"-"</span>, compress <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S3 method for class 'phyDat'</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">rbind</a></span><span class="op">(</span><span class="va">...</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S3 method for class 'phyDat'</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">...</span>, gaps <span class="op">=</span> <span class="st">"-"</span>, compress <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S3 method for class 'phyDat'</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">x</span>, <span class="va">subset</span>, <span class="va">select</span>, site.pattern <span class="op">=</span> <span class="cn">TRUE</span>, <span class="va">...</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S3 method for class 'phyDat'</span></span>
<span><span class="va">x</span><span class="op">[</span><span class="va">i</span>, <span class="va">j</span>, <span class="va">...</span>, drop <span class="op">=</span> <span class="cn">FALSE</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># S3 method for class 'phyDat'</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">x</span>, incomparables <span class="op">=</span> <span class="cn">FALSE</span>, identical <span class="op">=</span> <span class="cn">TRUE</span>, <span class="va">...</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">removeUndeterminedSites</span><span class="op">(</span><span class="va">x</span>, <span class="va">...</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">removeAmbiguousSites</span><span class="op">(</span><span class="va">x</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">allSitePattern</span><span class="op">(</span><span class="va">n</span>, levels <span class="op">=</span> <span class="cn">NULL</span>, names <span class="op">=</span> <span class="cn">NULL</span>, type <span class="op">=</span> <span class="st">"DNA"</span>, code <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-x">x<a class="anchor" aria-label="anchor" href="#arg-x"></a></dt>
<dd><p>An object containing sequences.</p></dd>


<dt id="arg--">...<a class="anchor" aria-label="anchor" href="#arg--"></a></dt>
<dd><p>further arguments passed to or from other methods.</p></dd>


<dt id="arg-gaps">gaps<a class="anchor" aria-label="anchor" href="#arg-gaps"></a></dt>
<dd><p>character for gaps (default is '-').</p></dd>


<dt id="arg-compress">compress<a class="anchor" aria-label="anchor" href="#arg-compress"></a></dt>
<dd><p>logical, compress data to store only site patterns.</p></dd>


<dt id="arg-subset">subset<a class="anchor" aria-label="anchor" href="#arg-subset"></a></dt>
<dd><p>a subset of taxa.</p></dd>


<dt id="arg-select">select<a class="anchor" aria-label="anchor" href="#arg-select"></a></dt>
<dd><p>a subset of characters.</p></dd>


<dt id="arg-site-pattern">site.pattern<a class="anchor" aria-label="anchor" href="#arg-site-pattern"></a></dt>
<dd><p>select site pattern or sites (see details).</p></dd>


<dt id="arg-i-j">i, j<a class="anchor" aria-label="anchor" href="#arg-i-j"></a></dt>
<dd><p>indices of the rows and/or columns to select or to drop. They
may be numeric, logical, or character (in the same way than for standard R
objects).</p></dd>


<dt id="arg-drop">drop<a class="anchor" aria-label="anchor" href="#arg-drop"></a></dt>
<dd><p>for compatibility with the generic (unused).</p></dd>


<dt id="arg-incomparables">incomparables<a class="anchor" aria-label="anchor" href="#arg-incomparables"></a></dt>
<dd><p>for compatibility with unique.</p></dd>


<dt id="arg-identical">identical<a class="anchor" aria-label="anchor" href="#arg-identical"></a></dt>
<dd><p>if TRUE (default) sequences have to be identical, if FALSE
sequences are considered duplicates if distance between sequences is zero
(happens frequently with ambiguous sites).</p></dd>


<dt id="arg-n">n<a class="anchor" aria-label="anchor" href="#arg-n"></a></dt>
<dd><p>Number of sequences.</p></dd>


<dt id="arg-levels">levels<a class="anchor" aria-label="anchor" href="#arg-levels"></a></dt>
<dd><p>Level attributes.</p></dd>


<dt id="arg-names">names<a class="anchor" aria-label="anchor" href="#arg-names"></a></dt>
<dd><p>Names of sequences.</p></dd>


<dt id="arg-type">type<a class="anchor" aria-label="anchor" href="#arg-type"></a></dt>
<dd><p>Type of sequences ("DNA", "AA" or "USER").</p></dd>


<dt id="arg-code">code<a class="anchor" aria-label="anchor" href="#arg-code"></a></dt>
<dd><p>The ncbi genetic code number for translation.
By default the standard genetic code is used.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>The functions return an object of class <code>phyDat</code>.</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p><code>allSitePattern</code> generates all possible site patterns and can be useful
in simulation studies. For further details see the vignette
AdvancedFeatures.</p>
<p>The generic function <code>c</code> can be used to to combine sequences and
<code>unique</code> to get all unique sequences or unique haplotypes.</p>
<p><code>phyDat</code> stores identical columns of an alignment only once and keeps an
index of the original positions. This saves memory and especially
computations as these are usually need to be done only once for each site
pattern.
In the example below the matrix x in the example has 8 columns, but column 1
and 2 and also 3 and 5 are identical. The <code>phyDat</code> object y has only 6
site pattern. If argument <code>site.pattern=FALSE</code> the indexing behaves like
on the original matrix x. <code>site.pattern=TRUE</code> can be useful inside
functions.</p>
    </div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p><code><a href="https://rdrr.io/pkg/ape/man/DNAbin.html" class="external-link">DNAbin</a></code>, <code><a href="https://rdrr.io/pkg/ape/man/as.alignment.html" class="external-link">as.DNAbin</a></code>,
<code><a href="baseFreq.html">baseFreq</a></code>, <code><a href="baseFreq.html">glance.phyDat</a></code>, <code><a href="dna2codon.html">dna2codon</a></code>,
<code><a href="https://rdrr.io/pkg/ape/man/read.dna.html" class="external-link">read.dna</a></code>, <code><a href="https://rdrr.io/pkg/ape/man/read.nexus.data.html" class="external-link">read.nexus.data</a></code>
and the chapter 1 in the <code><a href="../articles/AdvancedFeatures.html">vignette("AdvancedFeatures",
package="phangorn")</a></code> and the example of <code><a href="pmlMix.html">pmlMix</a></code> for the use of
<code>allSitePattern</code>.</p></div>
    </div>
    <div class="section level2">
    <h2 id="author">Author<a class="anchor" aria-label="anchor" href="#author"></a></h2>
    <p>Klaus Schliep <a href="mailto:klaus.schliep@gmail.com">klaus.schliep@gmail.com</a></p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">Laurasiatherian</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="va">Laurasiatherian</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "phyDat"</span>
<span class="r-in"><span><span class="va">Laurasiatherian</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 47 sequences with 3179 character and 1605 different site patterns.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> The states are a c g t </span>
<span class="r-in"><span><span class="co"># base frequencies</span></span></span>
<span class="r-in"><span><span class="fu"><a href="baseFreq.html">baseFreq</a></span><span class="op">(</span><span class="va">Laurasiatherian</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         a         c         g         t </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 0.3321866 0.1990791 0.2040652 0.2646691 </span>
<span class="r-in"><span><span class="co"># summary statistics</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">Laurasiatherian</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Alignment statistics </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Type:  DNA </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Sequences:  47 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Columns:  3179 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Site pattern:  1605 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Parsimony informative sites:  1400 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Constant sites:  1354 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Number of gaps:  0 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Number of ambiguous states:  0 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Duplicated sequences:  0 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> State frequencies (empirical):</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> p(a) = 0.3321866 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> p(c) = 0.1990791 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> p(g) = 0.2040652 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> p(t) = 0.2646691 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Composition Test (Chisq) </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>           statistic df  p-value</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Opposum      12.518  3 0.005805</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Baboon        8.888  3 0.030823</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Human         8.319  3 0.039866</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Bandicoot     7.909  3 0.047937</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># subsetting phyDat objects</span></span></span>
<span class="r-in"><span><span class="co"># the first 5 sequences</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">Laurasiatherian</span>, subset<span class="op">=</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5 sequences with 3179 character and 1605 different site patterns.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> The states are a c g t </span>
<span class="r-in"><span><span class="co"># the first 5 characters</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">Laurasiatherian</span>, select<span class="op">=</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, site.pattern <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 47 sequences with 5 character and 5 different site patterns.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> The states are a c g t </span>
<span class="r-in"><span><span class="co"># subsetting with []</span></span></span>
<span class="r-in"><span><span class="va">Laurasiatherian</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">20</span><span class="op">]</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5 sequences with 20 character and 17 different site patterns.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> The states are a c g t </span>
<span class="r-in"><span><span class="co"># short for</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">Laurasiatherian</span>, subset<span class="op">=</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, select<span class="op">=</span><span class="fl">1</span><span class="op">:</span><span class="fl">20</span>, site.pattern <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5 sequences with 20 character and 17 different site patterns.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> The states are a c g t </span>
<span class="r-in"><span><span class="co"># the first 5 site patterns (often more than 5 characters)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">Laurasiatherian</span>, select<span class="op">=</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, site.pattern <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 47 sequences with 454 character and 5 different site patterns.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> The states are a c g t </span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">x</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"a"</span>, <span class="st">"a"</span>, <span class="st">"c"</span>, <span class="st">"g"</span>, <span class="st">"c"</span>, <span class="st">"t"</span>, <span class="st">"a"</span>, <span class="st">"g"</span>,</span></span>
<span class="r-in"><span>              <span class="st">"a"</span>, <span class="st">"a"</span>, <span class="st">"c"</span>, <span class="st">"g"</span>, <span class="st">"c"</span>, <span class="st">"t"</span>, <span class="st">"a"</span>, <span class="st">"g"</span>,</span></span>
<span class="r-in"><span>              <span class="st">"a"</span>, <span class="st">"a"</span>, <span class="st">"c"</span>, <span class="st">"c"</span>, <span class="st">"c"</span>, <span class="st">"t"</span>, <span class="st">"t"</span>, <span class="st">"g"</span><span class="op">)</span>, nrow<span class="op">=</span><span class="fl">3</span>, byrow <span class="op">=</span> <span class="cn">TRUE</span>,</span></span>
<span class="r-in"><span>            dimnames <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"t1"</span>, <span class="st">"t2"</span>, <span class="st">"t3"</span><span class="op">)</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">8</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">(</span><span class="va">y</span> <span class="op">&lt;-</span> <span class="fu"><a href="as.phyDat.html">phyDat</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3 sequences with 8 character and 6 different site patterns.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> The states are a c g t </span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">y</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2 sequences with 8 character and 6 different site patterns.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> The states are a c g t </span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">y</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">2</span>, compress<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2 sequences with 8 character and 4 different site patterns.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> The states are a c g t </span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">y</span>, select<span class="op">=</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span>, site.pattern <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    [,1] [,2] [,3]</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> t1 "a"  "a"  "c" </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> t2 "a"  "a"  "c" </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> t3 "a"  "a"  "c" </span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">y</span>, select<span class="op">=</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span>, site.pattern <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    [,1] [,2] [,3] [,4] [,5]</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> t1 "a"  "a"  "c"  "c"  "g" </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> t2 "a"  "a"  "c"  "c"  "g" </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> t3 "a"  "a"  "c"  "c"  "c" </span>
<span class="r-in"><span><span class="va">y</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">]</span> <span class="co"># same as subset(y, select=1:3, site.pattern = FALSE)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3 sequences with 3 character and 2 different site patterns.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> The states are a c g t </span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Compute all possible site patterns</span></span></span>
<span class="r-in"><span><span class="co"># for nucleotides there $4 ^ (number of tips)$ patterns</span></span></span>
<span class="r-in"><span><span class="fu">allSitePattern</span><span class="op">(</span><span class="fl">5</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5 sequences with 1024 character and 1024 different site patterns.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> The states are a c g t </span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Klaus Schliep, Emmanuel Paradis, Leonardo de Oliveira Martins, Alastair Potts, Iris Bardel-Kahr.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer></div>





  </body></html>

