<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en-US"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Plot ancestral character on a tree — plotAnc • phangorn</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Plot ancestral character on a tree — plotAnc"><meta name="description" content="plotAnc plots a phylogeny and adds character to the nodes. Either
takes output from  ancestral.pars or ancestral.pml or from an
alignment where there are node labels in the tree match the constructed
sequences in the alignment."><meta property="og:description" content="plotAnc plots a phylogeny and adds character to the nodes. Either
takes output from  ancestral.pars or ancestral.pml or from an
alignment where there are node labels in the tree match the constructed
sequences in the alignment."><meta property="og:image" content="https://klausvigo.github.io/phangorn/logo.png"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">phangorn</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.12.1.3</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/Trees.html">Estimating phylogenetic trees with phangorn</a></li>
    <li><a class="dropdown-item" href="../articles/Morphological.html">Phylogenetic trees from morphological data</a></li>
    <li><a class="dropdown-item" href="../articles/AdvancedFeatures.html">Markov models and transition rate matrices</a></li>
    <li><a class="dropdown-item" href="../articles/Ancestral.html">Ancestral Sequence Reconstruction</a></li>
    <li><a class="dropdown-item" href="../articles/Networx.html">Splits and Networx</a></li>
    <li><a class="dropdown-item" href="../articles/MLbyHand.html">Maximum likelihood by hand</a></li>
    <li><a class="dropdown-item" href="../articles/IntertwiningTreesAndNetworks.html">Intertwining phylogenetic trees and networks: R Example Script</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/KlausVigo/phangorn/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Plot ancestral character on a tree</h1>
      <small class="dont-index">Source: <a href="https://github.com/KlausVigo/phangorn/blob/main/R/plotAnc.R" class="external-link"><code>R/plotAnc.R</code></a></small>
      <div class="d-none name"><code>plot.ancestral.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p><code>plotAnc</code> plots a phylogeny and adds character to the nodes. Either
takes output from  <code>ancestral.pars</code> or <code>ancestral.pml</code> or from an
alignment where there are node labels in the tree match the constructed
sequences in the alignment.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">plotAnc</span><span class="op">(</span><span class="va">x</span>, i <span class="op">=</span> <span class="fl">1</span>, type <span class="op">=</span> <span class="st">"phylogram"</span>, <span class="va">...</span>, col <span class="op">=</span> <span class="cn">NULL</span>, cex.pie <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>  pos <span class="op">=</span> <span class="st">"bottomright"</span>, scheme <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">plotSeqLogo</span><span class="op">(</span><span class="va">x</span>, node <span class="op">=</span> <span class="fu"><a href="midpoint.html">getRoot</a></span><span class="op">(</span><span class="va">x</span><span class="op">$</span><span class="va">tree</span><span class="op">)</span>, start <span class="op">=</span> <span class="fl">1</span>, end <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  scheme <span class="op">=</span> <span class="st">"Ape_NT"</span>, <span class="va">...</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">add_mutations</span><span class="op">(</span><span class="va">x</span>, pos <span class="op">=</span> <span class="cn">NULL</span>, frame <span class="op">=</span> <span class="st">"none"</span>, <span class="va">...</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-x">x<a class="anchor" aria-label="anchor" href="#arg-x"></a></dt>
<dd><p>an object of class <code>ancestral</code>.</p></dd>


<dt id="arg-i">i<a class="anchor" aria-label="anchor" href="#arg-i"></a></dt>
<dd><p>plots the i-th site.</p></dd>


<dt id="arg-type">type<a class="anchor" aria-label="anchor" href="#arg-type"></a></dt>
<dd><p>a character string specifying the type of phylogeny to be drawn;
it must be one of "phylogram" (the default), "cladogram", "fan", "unrooted",
"radial", "tidy", or any unambiguous abbreviation of these.</p></dd>


<dt id="arg--">...<a class="anchor" aria-label="anchor" href="#arg--"></a></dt>
<dd><p>Further arguments passed to or from other methods.</p></dd>


<dt id="arg-col">col<a class="anchor" aria-label="anchor" href="#arg-col"></a></dt>
<dd><p>a vector containing the colors for all possible states.</p></dd>


<dt id="arg-cex-pie">cex.pie<a class="anchor" aria-label="anchor" href="#arg-cex-pie"></a></dt>
<dd><p>a numeric defining the size of the pie graphs.</p></dd>


<dt id="arg-pos">pos<a class="anchor" aria-label="anchor" href="#arg-pos"></a></dt>
<dd><p>position or range in the alignment to add mutations. If NULL all
mutations are written out.</p></dd>


<dt id="arg-scheme">scheme<a class="anchor" aria-label="anchor" href="#arg-scheme"></a></dt>
<dd><p>a predefined color scheme. For amino acid options are "Ape_AA",
"Zappo_AA", "Clustal", "Polarity" and "Transmembrane_tendency", for
nucleotides "Ape_NT" and"RY_NT". Names can be abbreviated.</p></dd>


<dt id="arg-node">node<a class="anchor" aria-label="anchor" href="#arg-node"></a></dt>
<dd><p>to plot for which the probabilities should be plotted.</p></dd>


<dt id="arg-start">start<a class="anchor" aria-label="anchor" href="#arg-start"></a></dt>
<dd><p>start position to plot.</p></dd>


<dt id="arg-end">end<a class="anchor" aria-label="anchor" href="#arg-end"></a></dt>
<dd><p>end position to plot.</p></dd>


<dt id="arg-frame">frame<a class="anchor" aria-label="anchor" href="#arg-frame"></a></dt>
<dd><p>a character string specifying the kind of frame to be printed
around the text. See <code><a href="https://rdrr.io/pkg/ape/man/nodelabels.html" class="external-link">edgelabels</a></code>.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p><code>plotAnc</code> returns silently x.</p>
<p><code>plotSeqLogo</code> returns a ggplot object.</p>
<p><code>add_mutations</code> adds the position and changes of possible
mutations to a phylogeny.</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>For further details see vignette("Ancestral").</p>
    </div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p><code><a href="ancestral.pml.html">anc_pml</a></code>, <code><a href="https://rdrr.io/pkg/ape/man/plot.phylo.html" class="external-link">plot.phylo</a></code>,
<code><a href="https://rdrr.io/pkg/ape/man/image.DNAbin.html" class="external-link">image.DNAbin</a></code>, <code><a href="https://rdrr.io/pkg/ape/man/AAbin.html" class="external-link">image.AAbin</a></code>
<code><a href="https://rdrr.io/pkg/ggseqlogo/man/ggseqlogo.html" class="external-link">ggseqlogo</a></code>, <code><a href="https://rdrr.io/pkg/ape/man/nodelabels.html" class="external-link">edgelabels</a></code></p></div>
    </div>
    <div class="section level2">
    <h2 id="author">Author<a class="anchor" aria-label="anchor" href="#author"></a></h2>
    <p>Klaus Schliep <a href="mailto:klaus.schliep@gmail.com">klaus.schliep@gmail.com</a></p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/example.html" class="external-link">example</a></span><span class="op">(</span><span class="va">NJ</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> NJ&gt; data(Laurasiatherian)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> NJ&gt; dm &lt;- dist.ml(Laurasiatherian)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> NJ&gt; tree &lt;- NJ(dm)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> NJ&gt; plot(tree)</span>
<span class="r-plt img"><img src="plot.ancestral-1.png" alt="" width="700" height="433"></span>
<span class="r-in"><span><span class="co"># generate node labels to ensure plotting will work</span></span></span>
<span class="r-in"><span><span class="va">tree</span> <span class="op">&lt;-</span> <span class="fu">makeNodeLabel</span><span class="op">(</span><span class="va">tree</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">anc.p</span> <span class="op">&lt;-</span> <span class="fu"><a href="ancestral.pml.html">anc_pars</a></span><span class="op">(</span><span class="va">tree</span>, <span class="va">Laurasiatherian</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co"># plot the third character</span></span></span>
<span class="r-in"><span><span class="fu">plotAnc</span><span class="op">(</span><span class="va">anc.p</span>, <span class="fl">3</span>, pos<span class="op">=</span><span class="st">"topright"</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="plot.ancestral-2.png" alt="" width="700" height="433"></span>
<span class="r-in"><span><span class="fu">plotSeqLogo</span><span class="op">(</span><span class="va">anc.p</span>, node<span class="op">=</span><span class="st">"Node10"</span>, <span class="fl">1</span>, <span class="fl">25</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="plot.ancestral-3.png" alt="" width="700" height="433"></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">chloroplast</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">tree</span> <span class="op">&lt;-</span> <span class="fu"><a href="parsimony.html">pratchet</a></span><span class="op">(</span><span class="va">chloroplast</span>,  maxit<span class="op">=</span><span class="fl">10</span>, trace<span class="op">=</span><span class="fl">0</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">tree</span> <span class="op">&lt;-</span> <span class="fu">makeNodeLabel</span><span class="op">(</span><span class="va">tree</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">anc.ch</span> <span class="op">&lt;-</span> <span class="fu"><a href="ancestral.pml.html">anc_pars</a></span><span class="op">(</span><span class="va">tree</span>, <span class="va">chloroplast</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/graphics/image.html" class="external-link">image</a></span><span class="op">(</span><span class="fu"><a href="as.phyDat.html">as.phyDat</a></span><span class="op">(</span><span class="va">anc.ch</span><span class="op">)</span><span class="op">[</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">25</span><span class="op">]</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="plot.ancestral-4.png" alt="" width="700" height="433"></span>
<span class="r-in"><span><span class="fu">plotAnc</span><span class="op">(</span><span class="va">anc.ch</span>, <span class="fl">21</span>, scheme<span class="op">=</span><span class="st">"Ape_AA"</span>, pos<span class="op">=</span><span class="st">"topleft"</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="plot.ancestral-5.png" alt="" width="700" height="433"></span>
<span class="r-in"><span><span class="fu">plotAnc</span><span class="op">(</span><span class="va">anc.ch</span>, <span class="fl">21</span>, scheme<span class="op">=</span><span class="st">"Clustal"</span>, pos<span class="op">=</span><span class="st">"topleft"</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="plot.ancestral-6.png" alt="" width="700" height="433"></span>
<span class="r-in"><span><span class="fu">plotSeqLogo</span><span class="op">(</span><span class="va">anc.ch</span>, node<span class="op">=</span><span class="st">"Node1"</span>, <span class="fl">1</span>, <span class="fl">25</span>, scheme<span class="op">=</span><span class="st">"Clustal"</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="plot.ancestral-7.png" alt="" width="700" height="433"></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">woodmouse</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">tree</span> <span class="op">&lt;-</span> <span class="fu"><a href="pml_bb.html">pml_bb</a></span><span class="op">(</span><span class="va">woodmouse</span>, <span class="st">"JC"</span>, rearrangement<span class="op">=</span><span class="st">"NNI"</span><span class="op">)</span><span class="op">$</span><span class="va">tree</span> <span class="op">|&gt;</span> <span class="fu"><a href="midpoint.html">midpoint</a></span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> optimize edge weights:  -1861.356 --&gt; -1857.165 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> optimize edge weights:  -1857.165 --&gt; -1857.165 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> candidates: 6 swap: 1 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> candidates: 4 swap: 1 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> candidates: 0 swap: 0 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> optimize topology:  -1857.165 --&gt; -1856.056  NNI moves:  2 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> candidates: 0 swap: 0 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> optimize topology:  -1856.056 --&gt; -1856.056  NNI moves:  0 </span>
<span class="r-in"><span><span class="va">woodmouse_aa</span> <span class="op">&lt;-</span> <span class="fu">trans</span><span class="op">(</span><span class="va">woodmouse</span>, <span class="fl">2</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="as.phyDat.html">as.phyDat</a></span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>sequence length not a multiple of 3: 2 nucleotides dropped</span>
<span class="r-in"><span><span class="va">anc_aa</span> <span class="op">&lt;-</span> <span class="fu"><a href="ancestral.pml.html">anc_pars</a></span><span class="op">(</span><span class="va">tree</span>, <span class="va">woodmouse_aa</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">tree</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu">add_mutations</span><span class="op">(</span><span class="va">anc_aa</span>, adj <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">.5</span>, <span class="op">-</span><span class="fl">0.3</span><span class="op">)</span>, col<span class="op">=</span><span class="fl">2</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="plot.ancestral-8.png" alt="" width="700" height="433"></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Klaus Schliep, Emmanuel Paradis, Leonardo de Oliveira Martins, Alastair Potts, Iris Bardel-Kahr.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer></div>





  </body></html>

