<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en-US"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Auxiliary for Controlling Fitting — pml.control • phangorn</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Auxiliary for Controlling Fitting — pml.control"><meta name="description" content="Auxiliary functions for providing optim.pml, pml_bb
fitting. Use it to construct a control or ratchet.par argument."><meta property="og:description" content="Auxiliary functions for providing optim.pml, pml_bb
fitting. Use it to construct a control or ratchet.par argument."><meta property="og:image" content="https://klausvigo.github.io/phangorn/logo.png"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">phangorn</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.12.1.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/Trees.html">Estimating phylogenetic trees with phangorn</a></li>
    <li><a class="dropdown-item" href="../articles/Morphological.html">Phylogenetic trees from morphological data</a></li>
    <li><a class="dropdown-item" href="../articles/AdvancedFeatures.html">Markov models and transition rate matrices</a></li>
    <li><a class="dropdown-item" href="../articles/Ancestral.html">Ancestral Sequence Reconstruction</a></li>
    <li><a class="dropdown-item" href="../articles/Networx.html">Splits and Networx</a></li>
    <li><a class="dropdown-item" href="../articles/MLbyHand.html">Maximum likelihood by hand</a></li>
    <li><a class="dropdown-item" href="../articles/IntertwiningTreesAndNetworks.html">Intertwining phylogenetic trees and networks: R Example Script</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/KlausVigo/phangorn/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Auxiliary for Controlling Fitting</h1>
      <small class="dont-index">Source: <a href="https://github.com/KlausVigo/phangorn/blob/master/R/pml_control.R" class="external-link"><code>R/pml_control.R</code></a></small>
      <div class="d-none name"><code>pml.control.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Auxiliary functions for providing <code><a href="pml.html">optim.pml</a>, <a href="pml_bb.html">pml_bb</a></code>
fitting. Use it to construct a <code>control</code> or <code>ratchet.par</code> argument.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">pml.control</span><span class="op">(</span>epsilon <span class="op">=</span> <span class="fl">1e-08</span>, maxit <span class="op">=</span> <span class="fl">10</span>, trace <span class="op">=</span> <span class="fl">1</span>, tau <span class="op">=</span> <span class="fl">1e-08</span>,</span>
<span>  statefreq <span class="op">=</span> <span class="st">"empirical"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">ratchet.control</span><span class="op">(</span>iter <span class="op">=</span> <span class="fl">20L</span>, maxit <span class="op">=</span> <span class="fl">200L</span>, minit <span class="op">=</span> <span class="fl">100L</span>, prop <span class="op">=</span> <span class="fl">1</span><span class="op">/</span><span class="fl">2</span>,</span>
<span>  rell <span class="op">=</span> <span class="cn">TRUE</span>, bs <span class="op">=</span> <span class="fl">1000L</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-epsilon">epsilon<a class="anchor" aria-label="anchor" href="#arg-epsilon"></a></dt>
<dd><p>Stop criterion for optimization (see details).</p></dd>


<dt id="arg-maxit">maxit<a class="anchor" aria-label="anchor" href="#arg-maxit"></a></dt>
<dd><p>Maximum number of iterations (see details).</p></dd>


<dt id="arg-trace">trace<a class="anchor" aria-label="anchor" href="#arg-trace"></a></dt>
<dd><p>Show output during optimization (see details).</p></dd>


<dt id="arg-tau">tau<a class="anchor" aria-label="anchor" href="#arg-tau"></a></dt>
<dd><p>minimal edge length.</p></dd>


<dt id="arg-statefreq">statefreq<a class="anchor" aria-label="anchor" href="#arg-statefreq"></a></dt>
<dd><p>take "empirical" or "estimate" state frequencies.</p></dd>


<dt id="arg-iter">iter<a class="anchor" aria-label="anchor" href="#arg-iter"></a></dt>
<dd><p>Number of iterations to stop if there is no change.</p></dd>


<dt id="arg-minit">minit<a class="anchor" aria-label="anchor" href="#arg-minit"></a></dt>
<dd><p>Minimum number of iterations.</p></dd>


<dt id="arg-prop">prop<a class="anchor" aria-label="anchor" href="#arg-prop"></a></dt>
<dd><p>Only used if <code>rearrangement=stochastic</code>. How many NNI moves
should be added to the tree in proportion of the number of taxa.´</p></dd>


<dt id="arg-rell">rell<a class="anchor" aria-label="anchor" href="#arg-rell"></a></dt>
<dd><p>logical, if TRUE approximate bootstraping similar Minh et al.
(2013) is performed.</p></dd>


<dt id="arg-bs">bs<a class="anchor" aria-label="anchor" href="#arg-bs"></a></dt>
<dd><p>number of approximate bootstrap samples.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A list with components named as the arguments for controlling the
fitting process.</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p><code>pml.control</code> controls the fitting process. <code>epsilon</code> and
<code>maxit</code> are only defined for the most outer loop, this affects
<code>pmlCluster</code>, <code>pmlPart</code> and <code>pmlMix</code>.</p>
<p><code>epsilon</code> is not an absolute difference between, but instead is
defined as (logLik(k)-logLik(k+1))/logLik(k+1). This seems to be a good
compromise and to work reasonably well for small and large trees or
alignments.</p>
<p>If <code>trace</code> is set to zero than no out put is shown, if functions are
called internally than the trace is decreased by one, so a higher of trace
produces more feedback. It can be useful to figure out how long an run will
take and for debugging.</p>
<p><code>statefreq</code> controls if base/state frequencies are optimized or
empirical estimates are taken, when this applies. For some nucleotide models
(e.g. JC, SYM) equal base frequencies and for amino acid models precomputed
state frequencies are used, if not '+F' is specified.</p>
<p><code>tau</code> might be exactly zero if duplicated sequences in the alignment are
observed. In this case the analysis is performed only on unique sequences and
duplicated taxa are added to the tree with zero edge length. This may lead to
multifurcations if there are three or more identical sequences. After
optimization it is good practice to prune away edges of length <code>tau</code>
using <code>di2multi</code>. See also Janzen et al. (2021).</p>
    </div>
    <div class="section level2">
    <h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a></h2>
    <p>Minh, B. Q., Nguyen, M. A. T., &amp; von Haeseler, A. (2013).
Ultrafast approximation for phylogenetic bootstrap. <em>Molecular biology
and evolution</em>, <b>30(5)</b>, 1188-1195.</p>
<p>Janzen, T., Bokma, F.,Etienne,  R. S. (2021) Nucleotide Substitutions during
Speciation may Explain Substitution Rate Variation,
<em>Systematic Biology</em>, <b>71(5)</b>, 1244–1254.</p>
    </div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p><code><a href="pml_bb.html">pml_bb</a></code>, <code><a href="pml.html">optim.pml</a></code></p></div>
    </div>
    <div class="section level2">
    <h2 id="author">Author<a class="anchor" aria-label="anchor" href="#author"></a></h2>
    <p>Klaus Schliep <a href="mailto:klaus.schliep@gmail.com">klaus.schliep@gmail.com</a></p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="fu">pml.control</span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $epsilon</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 1e-08</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $maxit</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 10</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $trace</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $tau</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 1e-08</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $statefreq</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "empirical"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-in"><span><span class="fu">pml.control</span><span class="op">(</span>maxit<span class="op">=</span><span class="fl">25</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $epsilon</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 1e-08</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $maxit</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 25</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $trace</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $tau</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 1e-08</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $statefreq</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "empirical"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Klaus Schliep, Emmanuel Paradis, Leonardo de Oliveira Martins, Alastair Potts, Iris Bardel-Kahr.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.2.</p>
</div>

    </footer></div>





  </body></html>

