<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en-US"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Likelihood of a tree. — pml_bb • phangorn</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Likelihood of a tree. — pml_bb"><meta name="description" content="pml_bb for pml black box infers a phylogenetic tree infers a tree
using maximum likelihood (ML)."><meta property="og:description" content="pml_bb for pml black box infers a phylogenetic tree infers a tree
using maximum likelihood (ML)."><meta property="og:image" content="https://klausvigo.github.io/phangorn/logo.png"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">phangorn</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.12.1.3</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/Trees.html">Estimating phylogenetic trees with phangorn</a></li>
    <li><a class="dropdown-item" href="../articles/Morphological.html">Phylogenetic trees from morphological data</a></li>
    <li><a class="dropdown-item" href="../articles/AdvancedFeatures.html">Markov models and transition rate matrices</a></li>
    <li><a class="dropdown-item" href="../articles/Ancestral.html">Ancestral Sequence Reconstruction</a></li>
    <li><a class="dropdown-item" href="../articles/Networx.html">Splits and Networx</a></li>
    <li><a class="dropdown-item" href="../articles/MLbyHand.html">Maximum likelihood by hand</a></li>
    <li><a class="dropdown-item" href="../articles/IntertwiningTreesAndNetworks.html">Intertwining phylogenetic trees and networks: R Example Script</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/KlausVigo/phangorn/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Likelihood of a tree.</h1>
      <small class="dont-index">Source: <a href="https://github.com/KlausVigo/phangorn/blob/main/R/pml_bb.R" class="external-link"><code>R/pml_bb.R</code></a></small>
      <div class="d-none name"><code>pml_bb.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p><code>pml_bb</code> for pml black box infers a phylogenetic tree infers a tree
using maximum likelihood (ML).</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">pml_bb</span><span class="op">(</span><span class="va">x</span>, model <span class="op">=</span> <span class="cn">NULL</span>, rearrangement <span class="op">=</span> <span class="st">"stochastic"</span>,</span>
<span>  method <span class="op">=</span> <span class="st">"unrooted"</span>, start <span class="op">=</span> <span class="cn">NULL</span>, tip.dates <span class="op">=</span> <span class="cn">NULL</span>, <span class="va">...</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-x">x<a class="anchor" aria-label="anchor" href="#arg-x"></a></dt>
<dd><p>An alignment of class (either class <code>phyDat</code>, <code>DNAbin</code> or
<code>AAbin</code>) or an object of class <code>modelTest</code>.</p></dd>


<dt id="arg-model">model<a class="anchor" aria-label="anchor" href="#arg-model"></a></dt>
<dd><p>A string providing model (e.g. "GTR+G(4)+I"). Not necessary if
a modelTest object is supplied.</p></dd>


<dt id="arg-rearrangement">rearrangement<a class="anchor" aria-label="anchor" href="#arg-rearrangement"></a></dt>
<dd><p>Type of tree tree rearrangements to perform, one of
"none", "NNI", "stochastic" or "ratchet"</p></dd>


<dt id="arg-method">method<a class="anchor" aria-label="anchor" href="#arg-method"></a></dt>
<dd><p>One of "unrooted", "ultrametric" or "tiplabeled".</p></dd>


<dt id="arg-start">start<a class="anchor" aria-label="anchor" href="#arg-start"></a></dt>
<dd><p>A starting tree can be supplied.</p></dd>


<dt id="arg-tip-dates">tip.dates<a class="anchor" aria-label="anchor" href="#arg-tip-dates"></a></dt>
<dd><p>A named vector of sampling times associated to the tips /
sequences.</p></dd>


<dt id="arg--">...<a class="anchor" aria-label="anchor" href="#arg--"></a></dt>
<dd><p>Further arguments passed to or from other methods.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p><code>pml_bb</code> returns an object of class pml.</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p><code>pml_bb</code> is a convenience function combining <code>pml</code> and
<code>optim.pml</code>. If no tree is supplied, the function will generate a
starting tree. If a modelTest object is supplied the model will be chosen
according to BIC.</p>
<p><code>tip.dates</code> should be a named vector of sampling times, in any time
unit, with time increasing toward the present. For example, this may be in
units of “days since study start” or “years since 10,000 BCE”, but not
“millions of years ago”.</p>
<p><code>model</code> takes a string and tries to extract the model. When an
<code>modelTest</code> object the best BIC model is chosen by default.
The string should contain a substitution model (e.g. JC, GTR, WAG) and can
additional have  a term "+I" for invariant sites, "+G(4)" for a discrete
gamma model, "+R(4)" for a free rate model. In case of amino acid models
a term "+F" for estimating the amino acid frequencies. Whether nucleotide
frequencies are estimated is defined by <code><a href="pml.control.html">pml.control</a></code>.</p>
<p>Currently very experimental and likely to change.</p>
    </div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p><code><a href="pml.html">optim.pml</a></code>, <code><a href="modelTest.html">modelTest</a></code>,
<code><a href="https://rdrr.io/pkg/ape/man/rtt.html" class="external-link">rtt</a></code>, <code><a href="pml.control.html">pml.control</a></code></p></div>
    </div>
    <div class="section level2">
    <h2 id="author">Author<a class="anchor" aria-label="anchor" href="#author"></a></h2>
    <p>Klaus Schliep <a href="mailto:klaus.schliep@gmail.com">klaus.schliep@gmail.com</a></p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">woodmouse</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">tmp</span> <span class="op">&lt;-</span> <span class="fu">pml_bb</span><span class="op">(</span><span class="va">woodmouse</span>, model<span class="op">=</span><span class="st">"HKY+I"</span>, rearrangement<span class="op">=</span><span class="st">"NNI"</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> optimize edge weights:  -1812.646 --&gt; -1810.473 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> optimize rate matrix:  -1810.473 --&gt; -1758.757 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> optimize invariant sites:  -1758.757 --&gt; -1744.355 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> optimize edge weights:  -1744.355 --&gt; -1744.199 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> candidates: 0 swap: 0 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> optimize topology:  -1744.199 --&gt; -1744.199  NNI moves:  0 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> optimize rate matrix:  -1744.199 --&gt; -1744.186 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> optimize invariant sites:  -1744.186 --&gt; -1744.186 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> optimize edge weights:  -1744.186 --&gt; -1744.186 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> optimize rate matrix:  -1744.186 --&gt; -1744.186 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> optimize invariant sites:  -1744.186 --&gt; -1744.186 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> optimize edge weights:  -1744.186 --&gt; -1744.186 </span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span> <span class="co"># \dontrun{</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">Laurasiatherian</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">mt</span> <span class="op">&lt;-</span> <span class="fu"><a href="modelTest.html">modelTest</a></span><span class="op">(</span><span class="va">Laurasiatherian</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">fit</span> <span class="op">&lt;-</span> <span class="fu">pml_bb</span><span class="op">(</span><span class="va">mt</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># estimate free rate model with 2 rate categories</span></span></span>
<span class="r-in"><span><span class="va">fit_HKY_R2</span> <span class="op">&lt;-</span> <span class="fu">pml_bb</span><span class="op">(</span><span class="va">woodmouse</span>, model<span class="op">=</span><span class="st">"HKY+R(2)"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span> <span class="co"># }</span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Klaus Schliep, Emmanuel Paradis, Leonardo de Oliveira Martins, Alastair Potts, Iris Bardel-Kahr.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer></div>





  </body></html>

