<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Simulate sequences. — simSeq • phangorn</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Simulate sequences. — simSeq"><meta name="description" content="Simulate sequences from a given evolutionary tree."><meta property="og:description" content="Simulate sequences from a given evolutionary tree."><meta property="og:image" content="https://klausvigo.github.io/phangorn/logo.png"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">phangorn</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">3.0.0.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/Trees.html">Estimating phylogenetic trees with phangorn</a></li>
    <li><a class="dropdown-item" href="../articles/Morphological.html">Phylogenetic trees from morphological data</a></li>
    <li><a class="dropdown-item" href="../articles/AdvancedFeatures.html">Markov models and transition rate matrices</a></li>
    <li><a class="dropdown-item" href="../articles/Ancestral.html">Ancestral Sequence Reconstruction</a></li>
    <li><a class="dropdown-item" href="../articles/Networx.html">Splits and Networx</a></li>
    <li><a class="dropdown-item" href="../articles/MLbyHand.html">Maximum likelihood by hand</a></li>
    <li><a class="dropdown-item" href="../articles/IntertwiningTreesAndNetworks.html">Intertwining phylogenetic trees and networks: R Example Script</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/KlausVigo/phangorn/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Simulate sequences.</h1>
      <small class="dont-index">Source: <a href="https://github.com/KlausVigo/phangorn/blob/v2.12/R/simSeq.R" class="external-link"><code>R/simSeq.R</code></a></small>
      <div class="d-none name"><code>simSeq.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Simulate sequences from a given evolutionary tree.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">simSeq</span><span class="op">(</span><span class="va">x</span>, <span class="va">...</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S3 method for class 'phylo'</span></span>
<span><span class="fu">simSeq</span><span class="op">(</span><span class="va">x</span>, l <span class="op">=</span> <span class="fl">1000</span>, Q <span class="op">=</span> <span class="cn">NULL</span>, bf <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  rootseq <span class="op">=</span> <span class="cn">NULL</span>, type <span class="op">=</span> <span class="st">"DNA"</span>, model <span class="op">=</span> <span class="cn">NULL</span>, levels <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  rate <span class="op">=</span> <span class="fl">1</span>, ancestral <span class="op">=</span> <span class="cn">FALSE</span>, code <span class="op">=</span> <span class="fl">1</span>, <span class="va">...</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S3 method for class 'pml'</span></span>
<span><span class="fu">simSeq</span><span class="op">(</span><span class="va">x</span>, ancestral <span class="op">=</span> <span class="cn">FALSE</span>, <span class="va">...</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-x">x<a class="anchor" aria-label="anchor" href="#arg-x"></a></dt>
<dd><p>a phylogenetic tree <code>tree</code>, i.e. an object of class
<code>phylo</code> or and object of class <code>pml</code>.</p></dd>


<dt id="arg--">...<a class="anchor" aria-label="anchor" href="#arg--"></a></dt>
<dd><p>Further arguments passed to or from other methods.</p></dd>


<dt id="arg-l">l<a class="anchor" aria-label="anchor" href="#arg-l"></a></dt>
<dd><p>The length of the sequence to simulate.</p></dd>


<dt id="arg-q">Q<a class="anchor" aria-label="anchor" href="#arg-q"></a></dt>
<dd><p>The rate matrix.</p></dd>


<dt id="arg-bf">bf<a class="anchor" aria-label="anchor" href="#arg-bf"></a></dt>
<dd><p>Base frequencies.</p></dd>


<dt id="arg-rootseq">rootseq<a class="anchor" aria-label="anchor" href="#arg-rootseq"></a></dt>
<dd><p>A vector of length <code>l</code> containing the root sequence.
If not provided, the root sequence is randomly generated.</p></dd>


<dt id="arg-type">type<a class="anchor" aria-label="anchor" href="#arg-type"></a></dt>
<dd><p>Type of sequences ("DNA", "AA", "CODON" or "USER").</p></dd>


<dt id="arg-model">model<a class="anchor" aria-label="anchor" href="#arg-model"></a></dt>
<dd><p>Amino acid model of evolution to employ, for example "WAG",
"JTT", "Dayhoff" or "LG". For a full list of supported models, type
<code>phangorn:::.aamodels</code>. Ignored if type is not equal to "AA".</p></dd>


<dt id="arg-levels">levels<a class="anchor" aria-label="anchor" href="#arg-levels"></a></dt>
<dd><p>A character vector of the different character tokens.
Ignored unless type = "USER".</p></dd>


<dt id="arg-rate">rate<a class="anchor" aria-label="anchor" href="#arg-rate"></a></dt>
<dd><p>A numerical value greater than zero giving the mutation rate
or scaler for edge lengths.</p></dd>


<dt id="arg-ancestral">ancestral<a class="anchor" aria-label="anchor" href="#arg-ancestral"></a></dt>
<dd><p>Logical specifying whether to return ancestral sequences.</p></dd>


<dt id="arg-code">code<a class="anchor" aria-label="anchor" href="#arg-code"></a></dt>
<dd><p>The ncbi genetic code number for translation (see details). By
default the standard genetic code is used.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p><code>simSeq</code> returns an object of class phyDat.</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p><code>simSeq</code> is a generic function to simulate sequence alignments
along a phylogeny. It is quite flexible and can generate DNA, RNA,
amino acids, codon, morphological or binary sequences.
simSeq can take as input a phylogenetic tree of class <code>phylo</code>,
or a <code>pml</code> object; it will return an object of class <code>phyDat</code>.
There is also a more low level
version, which lacks rate variation, but one can combine different
alignments with their own rates (see example). The rate parameter acts like
a scaler for the edge lengths.</p>
<p>For codon models <code>type="CODON"</code>, two additional arguments <code>dnds</code>
for the dN/dS ratio and <code>tstv</code> for the transition transversion ratio
can be supplied.</p>
<p><strong>Defaults:</strong></p>
<p>If <code>x</code> is a tree of class <code>phylo</code>, then sequences will be generated
with the default Jukes-Cantor DNA model (<code>"JC"</code>).</p>
<p>If <code>bf</code> is not specified, then all states will be treated as equally
probable.</p>
<p>If <code>Q</code> is not specified, then a uniform rate matrix will be employed.</p>
    </div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p><code><a href="as.phyDat.html">phyDat</a></code>, <code><a href="pml.html">pml</a></code>, <code><a href="SOWH.test.html">SOWH.test</a></code></p></div>
    </div>
    <div class="section level2">
    <h2 id="author">Author<a class="anchor" aria-label="anchor" href="#author"></a></h2>
    <p>Klaus Schliep <a href="mailto:klaus.schliep@gmail.com">klaus.schliep@gmail.com</a></p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span></span></span>
<span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span> <span class="co"># \dontrun{</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">Laurasiatherian</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">tree</span> <span class="op">&lt;-</span> <span class="fu">nj</span><span class="op">(</span><span class="fu"><a href="dist.hamming.html">dist.ml</a></span><span class="op">(</span><span class="va">Laurasiatherian</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="pml.html">pml</a></span><span class="op">(</span><span class="va">tree</span>, <span class="va">Laurasiatherian</span>, k<span class="op">=</span><span class="fl">4</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="pml.html">optim.pml</a></span><span class="op">(</span><span class="va">fit</span>, optNni<span class="op">=</span><span class="cn">TRUE</span>, model<span class="op">=</span><span class="st">"GTR"</span>, optGamma<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu">simSeq</span><span class="op">(</span><span class="va">fit</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span> <span class="co"># }</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">tree</span> <span class="op">&lt;-</span> <span class="fu">rtree</span><span class="op">(</span><span class="fl">5</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">tree</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu">nodelabels</span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="simSeq-1.png" alt="" width="700" height="433"></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Example for simple DNA alignment</span></span></span>
<span class="r-in"><span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu">simSeq</span><span class="op">(</span><span class="va">tree</span>, l <span class="op">=</span> <span class="fl">10</span>, type<span class="op">=</span><span class="st">"DNA"</span>, bf<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">.1</span>,<span class="fl">.2</span>,<span class="fl">.3</span>,<span class="fl">.4</span><span class="op">)</span>, Q<span class="op">=</span><span class="fl">1</span><span class="op">:</span><span class="fl">6</span>,</span></span>
<span class="r-in"><span>               ancestral<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    [,1] [,2] [,3] [,4] [,5] [,6] [,7] [,8] [,9] [,10]</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> t5 "t"  "t"  "a"  "a"  "a"  "c"  "g"  "g"  "c"  "g"  </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> t4 "t"  "a"  "a"  "a"  "a"  "c"  "t"  "g"  "c"  "g"  </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> t1 "c"  "t"  "a"  "g"  "c"  "a"  "t"  "g"  "g"  "t"  </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> t3 "c"  "c"  "a"  "g"  "g"  "t"  "t"  "t"  "t"  "t"  </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> t2 "c"  "t"  "c"  "t"  "t"  "t"  "t"  "t"  "t"  "g"  </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6  "c"  "t"  "t"  "t"  "t"  "g"  "t"  "t"  "t"  "g"  </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 7  "c"  "t"  "a"  "g"  "t"  "t"  "t"  "t"  "t"  "g"  </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 8  "t"  "a"  "a"  "a"  "a"  "c"  "t"  "g"  "c"  "g"  </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 9  "c"  "c"  "a"  "g"  "g"  "t"  "t"  "t"  "t"  "t"  </span>
<span class="r-in"><span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Example to simulate discrete Gamma rate variation</span></span></span>
<span class="r-in"><span><span class="va">rates</span> <span class="op">&lt;-</span> <span class="fu"><a href="discrete.gamma.html">discrete.gamma</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">4</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">data1</span> <span class="op">&lt;-</span> <span class="fu">simSeq</span><span class="op">(</span><span class="va">tree</span>, l <span class="op">=</span> <span class="fl">100</span>, type<span class="op">=</span><span class="st">"AA"</span>, model<span class="op">=</span><span class="st">"WAG"</span>, rate<span class="op">=</span><span class="va">rates</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">data2</span> <span class="op">&lt;-</span> <span class="fu">simSeq</span><span class="op">(</span><span class="va">tree</span>, l <span class="op">=</span> <span class="fl">100</span>, type<span class="op">=</span><span class="st">"AA"</span>, model<span class="op">=</span><span class="st">"WAG"</span>, rate<span class="op">=</span><span class="va">rates</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">data3</span> <span class="op">&lt;-</span> <span class="fu">simSeq</span><span class="op">(</span><span class="va">tree</span>, l <span class="op">=</span> <span class="fl">100</span>, type<span class="op">=</span><span class="st">"AA"</span>, model<span class="op">=</span><span class="st">"WAG"</span>, rate<span class="op">=</span><span class="va">rates</span><span class="op">[</span><span class="fl">3</span><span class="op">]</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">data4</span> <span class="op">&lt;-</span> <span class="fu">simSeq</span><span class="op">(</span><span class="va">tree</span>, l <span class="op">=</span> <span class="fl">100</span>, type<span class="op">=</span><span class="st">"AA"</span>, model<span class="op">=</span><span class="st">"WAG"</span>, rate<span class="op">=</span><span class="va">rates</span><span class="op">[</span><span class="fl">4</span><span class="op">]</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">data1</span>,<span class="va">data2</span>, <span class="va">data3</span>, <span class="va">data4</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu"><a href="read.phyDat.html">write.phyDat</a></span><span class="op">(</span><span class="va">data</span>, file<span class="op">=</span><span class="st">"temp.dat"</span>, format<span class="op">=</span><span class="st">"sequential"</span>, nbcol <span class="op">=</span> <span class="op">-</span><span class="fl">1</span>,</span></span>
<span class="r-in"><span>  colsep <span class="op">=</span> <span class="st">""</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/unlink.html" class="external-link">unlink</a></span><span class="op">(</span><span class="st">"temp.dat"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Klaus Schliep, Emmanuel Paradis, Leonardo de Oliveira Martins, Alastair Potts, Iris Bardel-Kahr.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

    </footer></div>





  </body></html>

