<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="phangorn">
<title>Markov models and transition rate matrices • phangorn</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Markov models and transition rate matrices">
<meta property="og:description" content="phangorn">
<meta property="og:image" content="https://klausvigo.github.io/phangorn/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">phangorn</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.10.0.900</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/Trees.html">Estimating phylogenetic trees with phangorn</a>
    <a class="dropdown-item" href="../articles/Morphological.html">Phylogenetic trees from morphological data</a>
    <a class="dropdown-item" href="../articles/AdvancedFeatures.html">Markov models and transition rate matrices</a>
    <a class="dropdown-item" href="../articles/Ancestral.html">Ancestral Sequence Reconstruction</a>
    <a class="dropdown-item" href="../articles/Networx.html">Splits and Networx</a>
    <a class="dropdown-item" href="../articles/MLbyHand.html">Maximum likelihood by hand</a>
    <a class="dropdown-item" href="../articles/IntertwiningTreesAndNetworks.html">Intertwining phylogenetic trees and networks: R Example Script</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/KlausVigo/phangorn/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Markov models and transition rate matrices</h1>
                        <h4 data-toc-skip class="author">Klaus Schliep,
Iris Bardel-Kahr</h4>
            <address class="author_afil">
      Graz University of Technology, University of
Graz<br><a class="author_email" href="mailto:#"></a><a href="mailto:klaus.schliep@gmail.com" class="email">klaus.schliep@gmail.com</a>
      </address>
                  
            <h4 data-toc-skip class="date">2023-01-17</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/KlausVigo/phangorn/blob/HEAD/vignettes/AdvancedFeatures.Rmd" class="external-link"><code>vignettes/AdvancedFeatures.Rmd</code></a></small>
      <div class="d-none name"><code>AdvancedFeatures.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>This document illustrates some of the <em>phangorn</em> <span class="citation">(Schliep 2011)</span> specialized features which are
useful but maybe not as well-known or just not (yet) described
elsewhere. This is mainly interesting for someone who wants to explore
different models or set up some simulation studies. We show how to
construct data objects for different character states other than
nucleotides or amino acids or how to set up different models to estimate
transition rate.</p>
<p>The vignettes <em>Estimating phylogenetic trees with phangorn</em>
and <em>Phylogenetic trees from morphological data</em> describe in
detail how to estimate phylogenies from nucleotides, amino acids or
morphological data.</p>
</div>
<div class="section level2">
<h2 id="user-defined-data-formats">User defined data formats<a class="anchor" aria-label="anchor" href="#user-defined-data-formats"></a>
</h2>
<p>To better understand how to define our own data type it is useful to
know a bit more about the internal representation of <code>phyDat</code>
objects. The internal representation of <code>phyDat</code> object is
very similar to <code>factor</code> objects.</p>
<p>As an example we will show here several possibilities to define
nucleotide data with gaps defined as a fifth state. Ignoring gaps or
coding them as ambiguous sites - as it is done in most programs, also in
phangorn as default - may be misleading (see <span class="citation">(Warnow 2012)</span>). When the number of gaps is low
and randomly distributed, coding gaps as separate state may be not
important.</p>
<p>Let’s assume we have given a matrix where each row contains a
character vector of a taxonomic unit:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/KlausVigo/phangorn" class="external-link">phangorn</a></span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Loading required package: ape</span></span></code></pre>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"r"</span>,<span class="st">"a"</span>,<span class="st">"y"</span>,<span class="st">"g"</span>,<span class="st">"g"</span>,<span class="st">"a"</span>,<span class="st">"c"</span>,<span class="st">"-"</span>,<span class="st">"c"</span>,<span class="st">"t"</span>,<span class="st">"c"</span>,<span class="st">"g"</span>,</span>
<span>    <span class="st">"a"</span>,<span class="st">"a"</span>,<span class="st">"t"</span>,<span class="st">"g"</span>,<span class="st">"g"</span>,<span class="st">"a"</span>,<span class="st">"t"</span>,<span class="st">"-"</span>,<span class="st">"c"</span>,<span class="st">"t"</span>,<span class="st">"c"</span>,<span class="st">"a"</span>,</span>
<span>    <span class="st">"a"</span>,<span class="st">"a"</span>,<span class="st">"t"</span>,<span class="st">"-"</span>,<span class="st">"g"</span>,<span class="st">"a"</span>,<span class="st">"c"</span>,<span class="st">"c"</span>,<span class="st">"c"</span>,<span class="st">"t"</span>,<span class="st">"?"</span>,<span class="st">"g"</span><span class="op">)</span>,</span>
<span>    dimnames <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"t1"</span>, <span class="st">"t2"</span>, <span class="st">"t3"</span><span class="op">)</span>,<span class="cn">NULL</span><span class="op">)</span>, nrow<span class="op">=</span><span class="fl">3</span>, byrow<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">data</span></span></code></pre></div>
<pre><code><span><span class="co">##    [,1] [,2] [,3] [,4] [,5] [,6] [,7] [,8] [,9] [,10] [,11] [,12]</span></span>
<span><span class="co">## t1 "r"  "a"  "y"  "g"  "g"  "a"  "c"  "-"  "c"  "t"   "c"   "g"  </span></span>
<span><span class="co">## t2 "a"  "a"  "t"  "g"  "g"  "a"  "t"  "-"  "c"  "t"   "c"   "a"  </span></span>
<span><span class="co">## t3 "a"  "a"  "t"  "-"  "g"  "a"  "c"  "c"  "c"  "t"   "?"   "g"</span></span></code></pre>
<p>Normally we would transform this matrix into a phyDat object and gaps
are handled as ambiguous character (like “?”).</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gapsdata1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/as.phyDat.html">phyDat</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span></span>
<span><span class="va">gapsdata1</span></span></code></pre></div>
<pre><code><span><span class="co">## 3 sequences with 12 character and 11 different site patterns.</span></span>
<span><span class="co">## The states are a c g t</span></span></code></pre>
<p>Now we will define a “USER” defined object and have to supply a
vector of levels of the character states for the new data – in our case
the four nucleotide states and the gap. Additionally we can define
ambiguous states which can be any of the states.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gapsdata2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/as.phyDat.html">phyDat</a></span><span class="op">(</span><span class="va">data</span>, type<span class="op">=</span><span class="st">"USER"</span>, levels<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"a"</span>,<span class="st">"c"</span>,<span class="st">"g"</span>,<span class="st">"t"</span>,<span class="st">"-"</span><span class="op">)</span>,</span>
<span>    ambiguity <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"?"</span>, <span class="st">"n"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Warning in phyDat.default(data, levels = levels, return.index = return.index, :</span></span>
<span><span class="co">## Found unknown characters (not supplied in levels). Deleted sites with unknown</span></span>
<span><span class="co">## states.</span></span></code></pre>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gapsdata2</span></span></code></pre></div>
<pre><code><span><span class="co">## 3 sequences with 10 character and 9 different site patterns.</span></span>
<span><span class="co">## The states are a c g t -</span></span></code></pre>
<p>This is not yet what we wanted, as two sites of our alignment - which
contain the ambiguous characters “r” and “y” - got deleted. To define
ambiguous characters like “r” and “y” explicitly we have to supply a
contrast matrix similar to contrasts for factors.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">contrast</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span>data <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">0</span>,<span class="fl">0</span>,<span class="fl">0</span>,<span class="fl">0</span>,</span>
<span>    <span class="fl">0</span>,<span class="fl">1</span>,<span class="fl">0</span>,<span class="fl">0</span>,<span class="fl">0</span>,</span>
<span>    <span class="fl">0</span>,<span class="fl">0</span>,<span class="fl">1</span>,<span class="fl">0</span>,<span class="fl">0</span>,</span>
<span>    <span class="fl">0</span>,<span class="fl">0</span>,<span class="fl">0</span>,<span class="fl">1</span>,<span class="fl">0</span>,</span>
<span>    <span class="fl">1</span>,<span class="fl">0</span>,<span class="fl">1</span>,<span class="fl">0</span>,<span class="fl">0</span>,</span>
<span>    <span class="fl">0</span>,<span class="fl">1</span>,<span class="fl">0</span>,<span class="fl">1</span>,<span class="fl">0</span>,</span>
<span>    <span class="fl">0</span>,<span class="fl">0</span>,<span class="fl">0</span>,<span class="fl">0</span>,<span class="fl">1</span>,</span>
<span>    <span class="fl">1</span>,<span class="fl">1</span>,<span class="fl">1</span>,<span class="fl">1</span>,<span class="fl">0</span>,</span>
<span>    <span class="fl">1</span>,<span class="fl">1</span>,<span class="fl">1</span>,<span class="fl">1</span>,<span class="fl">1</span><span class="op">)</span>,</span>
<span>    ncol <span class="op">=</span> <span class="fl">5</span>, byrow <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dimnames.html" class="external-link">dimnames</a></span><span class="op">(</span><span class="va">contrast</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"a"</span>,<span class="st">"c"</span>,<span class="st">"g"</span>,<span class="st">"t"</span>,<span class="st">"r"</span>,<span class="st">"y"</span>,<span class="st">"-"</span>,<span class="st">"n"</span>,<span class="st">"?"</span><span class="op">)</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"a"</span>, <span class="st">"c"</span>, <span class="st">"g"</span>, <span class="st">"t"</span>, <span class="st">"-"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">contrast</span></span></code></pre></div>
<pre><code><span><span class="co">##   a c g t -</span></span>
<span><span class="co">## a 1 0 0 0 0</span></span>
<span><span class="co">## c 0 1 0 0 0</span></span>
<span><span class="co">## g 0 0 1 0 0</span></span>
<span><span class="co">## t 0 0 0 1 0</span></span>
<span><span class="co">## r 1 0 1 0 0</span></span>
<span><span class="co">## y 0 1 0 1 0</span></span>
<span><span class="co">## - 0 0 0 0 1</span></span>
<span><span class="co">## n 1 1 1 1 0</span></span>
<span><span class="co">## ? 1 1 1 1 1</span></span></code></pre>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gapsdata3</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/as.phyDat.html">phyDat</a></span><span class="op">(</span><span class="va">data</span>, type<span class="op">=</span><span class="st">"USER"</span>, contrast<span class="op">=</span><span class="va">contrast</span><span class="op">)</span></span>
<span><span class="va">gapsdata3</span></span></code></pre></div>
<pre><code><span><span class="co">## 3 sequences with 12 character and 11 different site patterns.</span></span>
<span><span class="co">## The states are a c g t -</span></span></code></pre>
<p>Here we defined “n” as a state which can be any nucleotide, but not a
gap. “-” and “?” can be any state including a gap.</p>
<p>These data can be used in all functions available in
<em>phangorn</em> to compute distance matrices or perform parsimony and
maximum likelihood analysis.</p>
</div>
<div class="section level2">
<h2 id="markov-models-of-character-evolution">Markov models of character evolution<a class="anchor" aria-label="anchor" href="#markov-models-of-character-evolution"></a>
</h2>
<p>To model nucleotide substitutions across the edges of a tree T we can
assign a transition matrix. In the case of nucleotides, with four
character states, each 4 <span class="math inline">\(\times\)</span> 4
transition matrix has, at most, 12 free parameters.</p>
<p>Time-reversible Markov models are used to describe how characters
change over time, and use fewer parameters. Time-reversible means that
these models need not be directed in time, and the Markov property
states that these models depend only on the current state. These models
are used in analyses of phylogenies using maximum likelihood and MCMC,
computing pairwise distances, as well as in simulating sequence
evolution.</p>
<p>We will now describe the General Time-Reversible (GTR) model <span class="citation">(Tavaré 1986)</span>. The parameters of the GTR model
are the equilibrium frequencies <span class="math inline">\(\pi = (\pi_A
,\pi_C ,\pi_G ,\pi_T)\)</span> and a rate matrix <span class="math inline">\(Q\)</span> which has the form <span class="math display">\[\begin{equation}
Q =
\begin{pmatrix}
\ast  &amp; \alpha\pi_C &amp; \beta\pi_G &amp; \gamma\pi_T \\
\alpha\pi_A &amp; \ast &amp; \delta\pi_G &amp; \epsilon\pi_T \\
\beta\pi_A &amp; \delta\pi_C &amp; \ast &amp; \eta\pi_T \\
\gamma\pi_A &amp;  \epsilon\pi_C &amp; \eta\pi_G &amp; \ast \\
\end{pmatrix}
(1)
\end{equation}\]</span></p>
<p>where we need to assign 6 parameters <span class="math inline">\(\alpha, \dots, \eta\)</span>. The elements on the
diagonal are chosen so that the rows sum to zero. The Jukes-Cantor (JC)
<span class="citation">(Jukes and Cantor 1969)</span> model can be
derived as special case from the GTR model, for equal equilibrium
frequencies <span class="math inline">\(\pi_A = \pi_C = \pi_G = \pi_T =
0.25\)</span> and equal rates set to <span class="math inline">\(\alpha
= \beta = \gamma = \delta = \eta\)</span>. Table 2 lists all built-in
nucleotide models in <em>phangorn</em>. The transition probabilities,
which describe the probabilities of change from character <span class="math inline">\(i\)</span> to <span class="math inline">\(j\)</span> in time <span class="math inline">\(t\)</span>, are given by the corresponding entries
of the matrix exponential <span class="math display">\[
P(t) = (p_{ij}(t)) = e^{Qt}, \qquad \sum_j p_{ij}=1
\]</span> where <span class="math inline">\(P(t)\)</span> is the
transition matrix spanning a period of time <span class="math inline">\(t\)</span>.</p>
</div>
<div class="section level2">
<h2 id="estimation-of-non-standard-transition-rate-matrices">Estimation of non-standard transition rate matrices<a class="anchor" aria-label="anchor" href="#estimation-of-non-standard-transition-rate-matrices"></a>
</h2>
<p>In the section <em>User defined data formats</em> we described how to
set up user defined data formats. Now we describe how to estimate
transition matrices with pml.</p>
<p>Again for nucleotide data the most common models can be called
directly in the <code>optim.pml</code> function (e.g. “JC69”, “HKY”,
“GTR” to name a few). Table 2 lists all the available nucleotide models,
which can estimated directly in <code>optim.pml</code>. For amino acids
several transition matrices are available (“WAG”, “JTT”, “LG”,
“Dayhoff”, “cpREV”, “mtmam”, “mtArt”, “MtZoa”, “mtREV24”, “VT”,“RtREV”,
“HIVw”, “HIVb”, “FLU”, “Blosum62”, “Dayhoff_DCMut” and “JTT-DCMut”) or
can be estimated with <code>optim.pml</code>. For example <span class="citation">Mathews, Clements, and Beilstein (2010)</span> used
this function to estimate a phytochrome amino acid transition
matrix.</p>
<p>We will now show how to estimate a rate matrix with different
transition (<span class="math inline">\(\alpha\)</span>) and
transversion ratio (<span class="math inline">\(\beta\)</span>) and a
fixed rate to the gap state (<span class="math inline">\(\gamma\)</span>) - a kind of Kimura two-parameter
model (K81) for nucleotide data with gaps as fifth state (see table
1).</p>
<table class="table">
<caption>Tab 1. Rate matrix with 3 parameters to optimize.</caption>
<thead><tr class="header">
<th>a</th>
<th>c</th>
<th>g</th>
<th>t</th>
<th>-</th>
<th></th>
</tr></thead>
<tbody>
<tr class="odd">
<td>a</td>
<td></td>
<td></td>
<td></td>
<td></td>
<td></td>
</tr>
<tr class="even">
<td>c</td>
<td><span class="math inline">\(\beta\)</span></td>
<td></td>
<td></td>
<td></td>
<td></td>
</tr>
<tr class="odd">
<td>g</td>
<td><span class="math inline">\(\alpha\)</span></td>
<td><span class="math inline">\(\beta\)</span></td>
<td></td>
<td></td>
<td></td>
</tr>
<tr class="even">
<td>t</td>
<td><span class="math inline">\(\beta\)</span></td>
<td><span class="math inline">\(\alpha\)</span></td>
<td><span class="math inline">\(\beta\)</span></td>
<td></td>
<td></td>
</tr>
<tr class="odd">
<td>-</td>
<td><span class="math inline">\(\gamma\)</span></td>
<td><span class="math inline">\(\gamma\)</span></td>
<td><span class="math inline">\(\gamma\)</span></td>
<td><span class="math inline">\(\gamma\)</span></td>
<td></td>
</tr>
</tbody>
</table>
<p>If we want to fit a non-standard transition rate matrix, we have to
tell <code>optim.pml</code> which transitions in Q get the same rate.
The parameter vector subs accepts a vector of consecutive integers and
at least one element has to be zero (these get the reference rate of 1).
Negative values indicate that there is no direct transition possible and
the rate is set to zero.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://ape-package.ird.fr/" class="external-link">ape</a></span><span class="op">)</span></span>
<span><span class="va">tree</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ape/man/root.html" class="external-link">unroot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/ape/man/rtree.html" class="external-link">rtree</a></span><span class="op">(</span><span class="fl">3</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/pml.html">pml</a></span><span class="op">(</span><span class="va">tree</span>, <span class="va">gapsdata3</span><span class="op">)</span></span>
<span><span class="va">fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/pml.html">optim.pml</a></span><span class="op">(</span><span class="va">fit</span>, optQ<span class="op">=</span><span class="cn">TRUE</span>, subs<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">0</span>,<span class="fl">1</span>,<span class="fl">2</span>,<span class="fl">1</span>,<span class="fl">0</span>,<span class="fl">2</span>,<span class="fl">1</span>,<span class="fl">2</span>,<span class="fl">2</span><span class="op">)</span>,</span>
<span>    control<span class="op">=</span><span class="fu"><a href="../reference/pml.control.html">pml.control</a></span><span class="op">(</span>trace<span class="op">=</span><span class="fl">0</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">fit</span></span></code></pre></div>
<pre><code><span><span class="co">## model: Mk </span></span>
<span><span class="co">## loglikelihood: -33.01 </span></span>
<span><span class="co">## unconstrained loglikelihood: -28.43 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Rate matrix:</span></span>
<span><span class="co">##           a         c         g         t      -</span></span>
<span><span class="co">## a 0.000e+00 2.584e-06 1.000e+00 2.584e-06 0.6912</span></span>
<span><span class="co">## c 2.584e-06 0.000e+00 2.584e-06 1.000e+00 0.6912</span></span>
<span><span class="co">## g 1.000e+00 2.584e-06 0.000e+00 2.584e-06 0.6912</span></span>
<span><span class="co">## t 2.584e-06 1.000e+00 2.584e-06 0.000e+00 0.6912</span></span>
<span><span class="co">## - 6.912e-01 6.912e-01 6.912e-01 6.912e-01 0.0000</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Base frequencies:  </span></span>
<span><span class="co">##   a   c   g   t   - </span></span>
<span><span class="co">## 0.2 0.2 0.2 0.2 0.2</span></span></code></pre>
<p>Here are some conventions how the models are estimated:</p>
<p>If a model is supplied the base frequencies bf and rate matrix Q are
optimized according to the model (nucleotides) or the adequate rate
matrix and frequencies are chosen (for amino acids). If optQ=TRUE and
neither a model nor subs are supplied then a symmetric (optBf=FALSE) or
reversible model (optBf=TRUE, i.e. the GTR for nucleotides) is
estimated. This can be slow if the there are many character states,
e.g. for amino acids. Table 2 shows how parameters are optimized and the
number of parameters to estimate. The elements of the vector subs
correspond to <span class="math inline">\(\alpha, \dots, \eta\)</span>
in equation (1)</p>
<table class="table">
<caption>Tab 2. DNA models available in phangorn.</caption>
<thead><tr class="header">
<th>model</th>
<th>optQ</th>
<th>optBf</th>
<th>subs</th>
<th>df</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>JC</td>
<td>FALSE</td>
<td>FALSE</td>
<td><span class="math inline">\(c(0, 0, 0, 0, 0, 0)\)</span></td>
<td>0</td>
</tr>
<tr class="even">
<td>F81</td>
<td>FALSE</td>
<td>TRUE</td>
<td><span class="math inline">\(c(0, 0, 0, 0, 0, 0)\)</span></td>
<td>3</td>
</tr>
<tr class="odd">
<td>K80</td>
<td>TRUE</td>
<td>FALSE</td>
<td><span class="math inline">\(c(0, 1, 0, 0, 1, 0)\)</span></td>
<td>1</td>
</tr>
<tr class="even">
<td>HKY</td>
<td>TRUE</td>
<td>TRUE</td>
<td><span class="math inline">\(c(0, 1, 0, 0, 1, 0)\)</span></td>
<td>4</td>
</tr>
<tr class="odd">
<td>TrNe</td>
<td>TRUE</td>
<td>FALSE</td>
<td><span class="math inline">\(c(0, 1, 0, 0, 2, 0)\)</span></td>
<td>2</td>
</tr>
<tr class="even">
<td>TrN</td>
<td>TRUE</td>
<td>TRUE</td>
<td><span class="math inline">\(c(0, 1, 0, 0, 2, 0)\)</span></td>
<td>5</td>
</tr>
<tr class="odd">
<td>TPM1</td>
<td>TRUE</td>
<td>FALSE</td>
<td><span class="math inline">\(c(0, 1, 2, 2, 1, 0)\)</span></td>
<td>2</td>
</tr>
<tr class="even">
<td>K81</td>
<td>TRUE</td>
<td>FALSE</td>
<td><span class="math inline">\(c(0, 1, 2, 2, 1, 0)\)</span></td>
<td>2</td>
</tr>
<tr class="odd">
<td>TPM1u</td>
<td>TRUE</td>
<td>TRUE</td>
<td><span class="math inline">\(c(0, 1, 2, 2, 1, 0)\)</span></td>
<td>5</td>
</tr>
<tr class="even">
<td>TPM2</td>
<td>TRUE</td>
<td>FALSE</td>
<td><span class="math inline">\(c(1, 2, 1, 0, 2, 0)\)</span></td>
<td>2</td>
</tr>
<tr class="odd">
<td>TPM2u</td>
<td>TRUE</td>
<td>TRUE</td>
<td><span class="math inline">\(c(1, 2, 1, 0, 2, 0)\)</span></td>
<td>5</td>
</tr>
<tr class="even">
<td>TPM3</td>
<td>TRUE</td>
<td>FALSE</td>
<td><span class="math inline">\(c(1, 2, 0, 1, 2, 0)\)</span></td>
<td>2</td>
</tr>
<tr class="odd">
<td>TPM3u</td>
<td>TRUE</td>
<td>TRUE</td>
<td><span class="math inline">\(c(1, 2, 0, 1, 2, 0)\)</span></td>
<td>5</td>
</tr>
<tr class="even">
<td>TIM1e</td>
<td>TRUE</td>
<td>FALSE</td>
<td><span class="math inline">\(c(0, 1, 2, 2, 3, 0)\)</span></td>
<td>3</td>
</tr>
<tr class="odd">
<td>TIM1</td>
<td>TRUE</td>
<td>TRUE</td>
<td><span class="math inline">\(c(0, 1, 2, 2, 3, 0)\)</span></td>
<td>6</td>
</tr>
<tr class="even">
<td>TIM2e</td>
<td>TRUE</td>
<td>FALSE</td>
<td><span class="math inline">\(c(1, 2, 1, 0, 3, 0)\)</span></td>
<td>3</td>
</tr>
<tr class="odd">
<td>TIM2</td>
<td>TRUE</td>
<td>TRUE</td>
<td><span class="math inline">\(c(1, 2, 1, 0, 3, 0)\)</span></td>
<td>6</td>
</tr>
<tr class="even">
<td>TIM3e</td>
<td>TRUE</td>
<td>FALSE</td>
<td><span class="math inline">\(c(1, 2, 0, 1, 3, 0)\)</span></td>
<td>3</td>
</tr>
<tr class="odd">
<td>TIM3</td>
<td>TRUE</td>
<td>TRUE</td>
<td><span class="math inline">\(c(1, 2, 0, 1, 3, 0)\)</span></td>
<td>6</td>
</tr>
<tr class="even">
<td>TVMe</td>
<td>TRUE</td>
<td>FALSE</td>
<td><span class="math inline">\(c(1, 2, 3, 4, 2, 0)\)</span></td>
<td>4</td>
</tr>
<tr class="odd">
<td>TVM</td>
<td>TRUE</td>
<td>TRUE</td>
<td><span class="math inline">\(c(1, 2, 3, 4, 2, 0)\)</span></td>
<td>7</td>
</tr>
<tr class="even">
<td>SYM</td>
<td>TRUE</td>
<td>FALSE</td>
<td><span class="math inline">\(c(1, 2, 3, 4, 5, 0)\)</span></td>
<td>5</td>
</tr>
<tr class="odd">
<td>GTR</td>
<td>TRUE</td>
<td>TRUE</td>
<td><span class="math inline">\(c(1, 2, 3, 4, 5, 0)\)</span></td>
<td>8</td>
</tr>
</tbody>
</table>
<div class="section level3">
<h3 id="predefined-models-for-user-defined-data">Predefined models for user defined data<a class="anchor" aria-label="anchor" href="#predefined-models-for-user-defined-data"></a>
</h3>
<p>So far there are 4 models which are just a generalization from
nucleotide models allowing different number of states. In many cases
only the equal rates (ER) model will be appropriate.</p>
<table class="table">
<caption>Tab 2: Build in models which are available for USER defined
data.</caption>
<thead><tr class="header">
<th>DNA</th>
<th>USER</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>JC</td>
<td>ER</td>
</tr>
<tr class="even">
<td>F81</td>
<td>FREQ</td>
</tr>
<tr class="odd">
<td>SYM</td>
<td>SYM</td>
</tr>
<tr class="even">
<td>GTR</td>
<td>GTR</td>
</tr>
</tbody>
</table>
<p>There is an additional model ORDERED, which assumes ordered
characters and only allows to switch between neighboring states. Table 3
show the corresponding rate matrix.</p>
<table class="table">
<caption>Tab 3: Rate matrix for “ORDERED” model with 5 states.</caption>
<thead><tr class="header">
<th>a</th>
<th>b</th>
<th>c</th>
<th>d</th>
<th>e</th>
<th></th>
</tr></thead>
<tbody>
<tr class="odd">
<td>a</td>
<td></td>
<td></td>
<td></td>
<td></td>
<td></td>
</tr>
<tr class="even">
<td>b</td>
<td>1</td>
<td></td>
<td></td>
<td></td>
<td></td>
</tr>
<tr class="odd">
<td>c</td>
<td>0</td>
<td>1</td>
<td></td>
<td></td>
<td></td>
</tr>
<tr class="even">
<td>d</td>
<td>0</td>
<td>0</td>
<td>1</td>
<td></td>
<td></td>
</tr>
<tr class="odd">
<td>e</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>1</td>
<td></td>
</tr>
</tbody>
</table>
</div>
</div>
<div class="section level2">
<h2 id="codon-substitution-models">Codon substitution models<a class="anchor" aria-label="anchor" href="#codon-substitution-models"></a>
</h2>
<p>A special case of the transition rates are codon models.
<em>phangorn</em> now offers the possibility to estimate the <span class="math inline">\(d_N/d_S\)</span> ratio (sometimes called ka/ks),
for an overview see <span class="citation">(Yang 2014)</span>. These
functions extend the option to estimate the <span class="math inline">\(d_N/d_S\)</span> ratio for pairwise sequence
comparison as it is available through the function <code>kaks</code> in
<em>seqinr</em>. The transition rate between between codon <span class="math inline">\(i\)</span> and <span class="math inline">\(j\)</span> is defined as follows: <span class="math display">\[\begin{eqnarray}
q_{ij}=\left\{
    \begin{array}{l@{\quad}l}
         0 &amp; \textrm{if i and j differ in more than one position} \\
         \pi_j &amp; \textrm{for synonymous transversion} \\
         \pi_j\kappa &amp; \textrm{for synonymous transition} \\
         \pi_j\omega &amp; \textrm{for non-synonymous transversion} \\
         \pi_j\omega\kappa &amp; \textrm{for non-synonymous transition}
    \end{array}
    \right. \nonumber
\end{eqnarray}\]</span></p>
<p>where <span class="math inline">\(\omega\)</span> is the <span class="math inline">\(d_N/d_S\)</span> ratio, <span class="math inline">\(\kappa\)</span> the transition transversion ratio
and <span class="math inline">\(\pi_j\)</span> is the the equilibrium
frequency of codon <span class="math inline">\(j\)</span>. For <span class="math inline">\(\omega\sim1\)</span> the amino acid change is
neutral, for <span class="math inline">\(\omega &lt; 1\)</span>
purifying selection and <span class="math inline">\(\omega &gt;
1\)</span> positive selection.</p>
<p>Here we use a data set from and follow loosely the example in <span class="citation">Bielawski and Yang (2005)</span>. We first read in an
alignment and phylogenetic tree for 45 sequences of the nef gene in the
Human HIV-2 Genome using <code>read.phyDat</code> function.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fdir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/trees"</span>, package <span class="op">=</span> <span class="st">"phangorn"</span><span class="op">)</span></span>
<span><span class="va">hiv_2_nef</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/read.phyDat.html">read.phyDat</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">fdir</span>, <span class="st">"seqfile.txt"</span><span class="op">)</span>, format<span class="op">=</span><span class="st">"sequential"</span><span class="op">)</span></span>
<span><span class="va">tree</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ape/man/read.tree.html" class="external-link">read.tree</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">fdir</span>, <span class="st">"tree.txt"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>With the tree and data set we can estimate currently 3 different site
models:</p>
<ol style="list-style-type: decimal">
<li>The M0 model with a constant <span class="math inline">\(\omega\)</span>, where <span class="math inline">\(\omega\)</span> estimates the average over all
sites of the alignment. M0 does not allow for distinct <span class="math inline">\(\omega\)</span> and identifies classes, therefore
we will not retrieve any information regarding positive selection.<br>
</li>
<li>The M1a or nearly neutral model estimates two different <span class="math inline">\(\omega\)</span> value classes (<span class="math inline">\(\omega=1\)</span> &amp; <span class="math inline">\(\omega&lt;1\)</span>).</li>
<li>The M2a or positive selection model estimates three different
classes of <span class="math inline">\(\omega\)</span> (negative
selection <span class="math inline">\(\omega&lt;1\)</span>, neutral
selection <span class="math inline">\(\omega=1\)</span>, positive
selection <span class="math inline">\(\omega&gt;1\)</span>). One can use
a likelihood ratio test to compare the M1a and M2a to for positive
selection.</li>
</ol>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cdn</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/codonTest.html">codonTest</a></span><span class="op">(</span><span class="va">tree</span>, <span class="va">hiv_2_nef</span><span class="op">)</span></span>
<span><span class="va">cdn</span></span></code></pre></div>
<pre><code><span><span class="co">##   model Frequencies  estimate logLik  df   AIC   BIC  dnds_0 dnds_1 dnds_2</span></span>
<span><span class="co">## 1    M0        F3x4 empirical  -9773  98 19741 20087 0.50486     NA     NA</span></span>
<span><span class="co">## 2   M1a        F3x4 empirical  -9313  99 18824 19168 0.06281      1     NA</span></span>
<span><span class="co">## 3   M2a        F3x4 empirical  -9244 101 18689 19040 0.05551      1  2.469</span></span>
<span><span class="co">##      p_0    p_1    p_2  tstv</span></span>
<span><span class="co">## 1 1.0000     NA     NA 4.418</span></span>
<span><span class="co">## 2 0.5563 0.4437     NA 4.364</span></span>
<span><span class="co">## 3 0.5227 0.3617 0.1156 4.849</span></span></code></pre>
<p>Currently the choice of site models is limited to the three models
mentioned above and are no branch models implemented so far.</p>
<p>We can identify sites under positive selection using the Na"ive
empirical Bayes (NEB) method of <span class="citation">Yang and Nielsen
(1998)</span>: <span class="math display">\[
P(\omega|x_h) = \frac{P(X_h|\omega_i)p_i}{P(X_h)} =
\frac{P(X_h|\omega_i)p_i}{\sum_j P(X_h|\omega_j)p_j}
\]</span></p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">cdn</span>, <span class="st">"M1a"</span><span class="op">)</span></span></code></pre></div>
<p><img src="AdvancedFeatures_files/figure-html/plot_codon-1.png" width="576"></p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">cdn</span>, <span class="st">"M2a"</span><span class="op">)</span></span></code></pre></div>
<p><img src="AdvancedFeatures_files/figure-html/plot_codon-2.png" width="576"></p>
<p>A lot if implementations differ in the way the codon frequencies are
derived. The M0 model can be also estimated using <code>pml</code> and
<code>optim.pml</code> functions. There are several ways to estimate the
codon frequencies <span class="math inline">\(\pi_j\)</span>. The
simplest model is to assume they have equal frequencies (=1/61). A
second is to use the empirical codon frequencies, either computed using
<code>baseFreq</code> or using the argument <code>bf="empirical"</code>
in <code>pml</code>. This is usually not really good as some codons are
rare and have a high variance. One can estimate the frequencies from
nucleotide frequencies with the F1x4 model. Last but not least the
frequencies can be derived from the base frequencies at each codon
position, the F3x4 model is set by the argument
<code>bf="F3x4"</code>.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">treeM0</span> <span class="op">&lt;-</span> <span class="va">cdn</span><span class="op">$</span><span class="va">estimates</span><span class="op">[[</span><span class="st">"M0"</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">tree</span> <span class="co"># tree with edge lengths</span></span>
<span><span class="va">M0</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/pml.html">pml</a></span><span class="op">(</span><span class="va">treeM0</span>, <span class="fu"><a href="../reference/dna2codon.html">dna2codon</a></span><span class="op">(</span><span class="va">hiv_2_nef</span><span class="op">)</span>, bf<span class="op">=</span><span class="st">"F3x4"</span><span class="op">)</span></span>
<span><span class="va">M0</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/pml.html">optim.pml</a></span><span class="op">(</span><span class="va">M0</span>, model<span class="op">=</span><span class="st">"codon1"</span>, control<span class="op">=</span><span class="fu"><a href="../reference/pml.control.html">pml.control</a></span><span class="op">(</span>trace<span class="op">=</span><span class="fl">0</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">M0</span></span></code></pre></div>
<pre><code><span><span class="co">## model: codon1 </span></span>
<span><span class="co">## loglikelihood: -9773 </span></span>
<span><span class="co">## unconstrained loglikelihood: -1372 </span></span>
<span><span class="co">## dn/ds: 0.5049 </span></span>
<span><span class="co">## ts/tv: 4.418 </span></span>
<span><span class="co">## Freq: F3x4</span></span></code></pre>
<p>For the F3x4 model can optimize the codon frequencies setting the
option to <code>optBf=TRUE</code> in <code>optim.pml</code>.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">M0_opt</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/pml.html">optim.pml</a></span><span class="op">(</span><span class="va">M0</span>, model<span class="op">=</span><span class="st">"codon1"</span>, optBf<span class="op">=</span><span class="cn">TRUE</span>, control<span class="op">=</span><span class="fu"><a href="../reference/pml.control.html">pml.control</a></span><span class="op">(</span>trace<span class="op">=</span><span class="fl">0</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">M0_opt</span></span></code></pre></div>
<pre><code><span><span class="co">## model: codon1 </span></span>
<span><span class="co">## loglikelihood: -9668 </span></span>
<span><span class="co">## unconstrained loglikelihood: -1372 </span></span>
<span><span class="co">## dn/ds: 0.51 </span></span>
<span><span class="co">## ts/tv: 4.581 </span></span>
<span><span class="co">## Freq: F3x4</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<pre><code><span><span class="co">## R version 4.2.2 (2022-10-31)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu (64-bit)</span></span>
<span><span class="co">## Running under: Ubuntu 22.04.1 LTS</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3</span></span>
<span><span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">##  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">## [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">## [1] phangorn_2.10.0.900 ape_5.6-2          </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##  [1] Rcpp_1.0.9        highr_0.10        bslib_0.4.2       compiler_4.2.2   </span></span>
<span><span class="co">##  [5] jquerylib_0.1.4   tools_4.2.2       digest_0.6.31     jsonlite_1.8.4   </span></span>
<span><span class="co">##  [9] evaluate_0.19     memoise_2.0.1     lifecycle_1.0.3   nlme_3.1-160     </span></span>
<span><span class="co">## [13] lattice_0.20-45   pkgconfig_2.0.3   rlang_1.0.6       Matrix_1.5-1     </span></span>
<span><span class="co">## [17] fastmatch_1.1-3   igraph_1.3.5      cli_3.6.0         yaml_2.3.6       </span></span>
<span><span class="co">## [21] parallel_4.2.2    pkgdown_2.0.7     xfun_0.36         fastmap_1.1.0    </span></span>
<span><span class="co">## [25] stringr_1.5.0     knitr_1.41        desc_1.4.2        generics_0.1.3   </span></span>
<span><span class="co">## [29] fs_1.5.2          vctrs_0.5.1       sass_0.4.4        systemfonts_1.0.4</span></span>
<span><span class="co">## [33] rprojroot_2.0.3   grid_4.2.2        glue_1.6.2        R6_2.5.1         </span></span>
<span><span class="co">## [37] textshaping_0.3.6 rmarkdown_2.19    purrr_1.0.1       magrittr_2.0.3   </span></span>
<span><span class="co">## [41] codetools_0.2-18  htmltools_0.5.4   quadprog_1.5-8    ragg_1.2.5       </span></span>
<span><span class="co">## [45] stringi_1.7.12    cachem_1.0.6</span></span></code></pre>
</div>
<div class="section level2">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-Bielawski2005" class="csl-entry">
Bielawski, Joseph P., and Ziheng Yang. 2005. <span>“Maximum Likelihood
Methods for Detecting Adaptive Protein Evolution.”</span> In
<em>Statistical Methods in Molecular Evolution</em>, 103–24. New York,
NY: Springer New York. <a href="https://doi.org/10.1007/0-387-27733-1_5" class="external-link">https://doi.org/10.1007/0-387-27733-1_5</a>.
</div>
<div id="ref-Jukes1969" class="csl-entry">
Jukes, Thomas H., and Charles R. Cantor. 1969. <span>“{CHAPTER} 24 -
Evolution of Protein Molecules.”</span> In <em>Mammalian Protein
Metabolism</em>, edited by H. N. Munro, 21–132. Academic Press.
</div>
<div id="ref-Mathews2010" class="csl-entry">
Mathews, S., M. D. Clements, and M. A. Beilstein. 2010. <span>“A
Duplicate Gene Rooting of Seed Plants and the Phylogenetic Position of
Flowering Plants.”</span> <em>Phil. Trans. R. Soc. B</em> 365: 383–95.
</div>
<div id="ref-Schliep2011" class="csl-entry">
Schliep, Klaus Peter. 2011. <span>“Phangorn: Phylogenetic Analysis in
<span>R</span>.”</span> <em>Bioinformatics</em> 27 (4): 592–93. <a href="https://doi.org/10.1093/bioinformatics/btq706" class="external-link">https://doi.org/10.1093/bioinformatics/btq706</a>.
</div>
<div id="ref-Tavare1986" class="csl-entry">
Tavaré, Simon. 1986. <span>“Some Probabilistic and Statistical Problems
in the Analysis of DNA Sequences.”</span> <em>Lectures on Mathematics in
the Life Sciences</em>, no. 17: 57–86.
</div>
<div id="ref-Warnow2012" class="csl-entry">
Warnow, Tandy. 2012. <span>“Standard Maximum Likelihood Analyses of
Alignments with Gaps Can Be Statistically Inconsistent.”</span> <em>PLOS
Currents Tree of Life</em>.
</div>
<div id="ref-Yang2014" class="csl-entry">
Yang, Ziheng. 2014. <em>Molecular Evolution: A Statistical
Approach</em>. Oxford: Oxford University Press.
</div>
<div id="ref-Yang1998" class="csl-entry">
Yang, Ziheng, and Rasmus Nielsen. 1998. <span>“Synonymous and
Nonsynonymous Rate Variation in Nuclear Genes of Mammals.”</span>
<em>Journal of Molecular Evolution</em> 46 (4): 409–18. <a href="https://doi.org/10.1007/PL00006320" class="external-link">https://doi.org/10.1007/PL00006320</a>.
</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Klaus Schliep, Emmanuel Paradis, Leonardo de Oliveira Martins, Alastair Potts, Iris Bardel-Kahr.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
