<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Ancestral Sequence Reconstruction • phangorn</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Ancestral Sequence Reconstruction">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">phangorn</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.12.1.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/Trees.html">Estimating phylogenetic trees with phangorn</a></li>
    <li><a class="dropdown-item" href="../articles/Morphological.html">Phylogenetic trees from morphological data</a></li>
    <li><a class="dropdown-item" href="../articles/AdvancedFeatures.html">Markov models and transition rate matrices</a></li>
    <li><a class="dropdown-item" href="../articles/Ancestral.html">Ancestral Sequence Reconstruction</a></li>
    <li><a class="dropdown-item" href="../articles/Networx.html">Splits and Networx</a></li>
    <li><a class="dropdown-item" href="../articles/MLbyHand.html">Maximum likelihood by hand</a></li>
    <li><a class="dropdown-item" href="../articles/IntertwiningTreesAndNetworks.html">Intertwining phylogenetic trees and networks: R Example Script</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/KlausVigo/phangorn/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Ancestral Sequence Reconstruction</h1>
                        <h4 data-toc-skip class="author">Klaus
Schliep</h4>
            <address class="author_afil">
      Graz University of
Technology<br><a class="author_email" href="mailto:#"></a><a href="mailto:klaus.schliep@gmail.com" class="email">klaus.schliep@gmail.com</a>
      </address>
                  
            <h4 data-toc-skip class="date">2025-04-23</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/KlausVigo/phangorn/blob/master/vignettes/Ancestral.Rmd" class="external-link"><code>vignettes/Ancestral.Rmd</code></a></small>
      <div class="d-none name"><code>Ancestral.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>These notes describe the ancestral sequence reconstruction using the
<em>phangorn</em> package <span class="citation">(Schliep 2011)</span>.
<em>phangorn</em> provides several methods to estimate ancestral
character states with either Maximum Parsimony (MP) or Maximum
Likelihood (ML). For more background on all the methods see e.g. <span class="citation">(Felsenstein 2004)</span> or <span class="citation">(Yang 2006)</span>.</p>
</div>
<div class="section level2">
<h2 id="parsimony-reconstructions">Parsimony reconstructions<a class="anchor" aria-label="anchor" href="#parsimony-reconstructions"></a>
</h2>
<p>To reconstruct ancestral sequences we first load some data and
reconstruct a tree:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/KlausVigo/phangorn" class="external-link">phangorn</a></span><span class="op">)</span></span>
<span><span class="va">fdir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/trees"</span>, package <span class="op">=</span> <span class="st">"phangorn"</span><span class="op">)</span></span>
<span><span class="va">primates</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/read.phyDat.html">read.phyDat</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">fdir</span>, <span class="st">"primates.dna"</span><span class="op">)</span>,</span>
<span>                        format <span class="op">=</span> <span class="st">"interleaved"</span><span class="op">)</span></span>
<span><span class="va">tree</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/parsimony.html">pratchet</a></span><span class="op">(</span><span class="va">primates</span>, trace<span class="op">=</span><span class="fl">0</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="../reference/acctran.html">acctran</a></span><span class="op">(</span><span class="va">primates</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/pkg/ape/man/makeNodeLabel.html" class="external-link">makeNodeLabel</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/parsimony.html">parsimony</a></span><span class="op">(</span><span class="va">tree</span>, <span class="va">primates</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 746</span></span></code></pre>
<p>For parsimony analysis of the edge length represent the observed
number of changes. Reconstructing ancestral states therefore defines
also the edge lengths of a tree. However there can exist several equally
parsimonious reconstructions or states can be ambiguous and therefore
edge length can differ. In <em>phangorn</em> all the ancestral
reconstructions for parsimony used to be based on the fitch algorithm
(below version 3.0) and needed bifurcating trees. However trees can get
pruned afterwards using the function <code>multi2di</code> from
<em>ape</em>. Recently we replaced the acctran routine with a method
based on the sankoff algorithm adopting the algorithm for joint
reconstruction <span class="citation">(Pupko et al. 2000)</span> and
breaking ties at random. This has the additional benefit that it allows
us to infer phylogenies with multifurcations.</p>
<p>“MPR” reconstructs the ancestral states for each (internal) node as
if the tree would be rooted in that node. However the nodes are not
independent of each other. If one chooses one state for a specific node,
this can restrict the choice of neighboring nodes (figures 2 and 3).
There is also an option “POSTORDER” which is only a one pass algorithm,
which is useful for teaching purposes. The function acctran (accelerated
transformation) assigns edge length and internal nodes to the tree <span class="citation">(Swofford and Maddison 1987)</span>.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">anc.pars</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ancestral.pml.html">anc_pars</a></span><span class="op">(</span><span class="va">tree</span>, <span class="va">primates</span><span class="op">)</span></span></code></pre></div>
<p>The <code>plotSeqLogo</code> function is a wrapper around the from
the <em>ggseqlogo</em> function in the <em>ggseqlogo</em> package <span class="citation">(Wagih 2024)</span> and provides a simple way to show
proportions of a nucleotides of ancestral states (see figure 1).</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot.ancestral.html">plotSeqLogo</a></span><span class="op">(</span><span class="va">anc.pars</span>, node<span class="op">=</span><span class="fu"><a href="../reference/midpoint.html">getRoot</a></span><span class="op">(</span><span class="va">tree</span><span class="op">)</span>, <span class="fl">1</span>, <span class="fl">20</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="Ancestral_files/figure-html/seqLogo-1.png" alt="Fig 1. Ancestral reconstruction for a node." width="576"><p class="caption">
Fig 1. Ancestral reconstruction for a node.
</p>
</div>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot.ancestral.html">plotAnc</a></span><span class="op">(</span><span class="va">anc.pars</span>, <span class="fl">17</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/title.html" class="external-link">title</a></span><span class="op">(</span><span class="st">"MPR"</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="Ancestral_files/figure-html/MPR-1.png" alt="Fig 2. Ancestral reconstruction using MPR." width="576"><p class="caption">
Fig 2. Ancestral reconstruction using MPR.
</p>
</div>
</div>
<div class="section level2">
<h2 id="likelihood-reconstructions">Likelihood reconstructions<a class="anchor" aria-label="anchor" href="#likelihood-reconstructions"></a>
</h2>
<p><em>phangorn</em> also offers the possibility to estimate ancestral
states using ML. The advantages of ML over parsimony is that the
reconstruction accounts for different edge lengths. Currently a marginal
construction (see [<span class="citation">Yang (2006)</span>]<span class="citation">(Koshi and Goldstein 1996)</span>) and the joint
reconstruction <span class="citation">(Pupko et al. 2000)</span> is
implemented. Joint reconstructions is only for models without rate
variation (e.g. gamma models) or invariant sites.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/pml.html">pml</a></span><span class="op">(</span><span class="va">tree</span>, <span class="va">primates</span><span class="op">)</span></span>
<span><span class="va">fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/pml.html">optim.pml</a></span><span class="op">(</span><span class="va">fit</span>, model<span class="op">=</span><span class="st">"F81"</span>, control <span class="op">=</span> <span class="fu"><a href="../reference/pml.control.html">pml.control</a></span><span class="op">(</span>trace<span class="op">=</span><span class="fl">0</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>We can assign the ancestral states according to the highest
likelihood (“ml”):
<math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>P</mi><mrow><mo stretchy="true" form="prefix">(</mo><msub><mi>x</mi><mi>r</mi></msub><mo>=</mo><mi>A</mi><mo stretchy="true" form="postfix">)</mo></mrow><mo>=</mo><mfrac><mrow><mi>L</mi><mrow><mo stretchy="true" form="prefix">(</mo><msub><mi>x</mi><mi>r</mi></msub><mo>=</mo><mi>A</mi><mo stretchy="true" form="postfix">)</mo></mrow></mrow><mrow><munder><mo>∑</mo><mrow><mi>k</mi><mo>∈</mo><mo stretchy="false" form="prefix">{</mo><mi>A</mi><mo>,</mo><mi>C</mi><mo>,</mo><mi>G</mi><mo>,</mo><mi>T</mi><mo stretchy="false" form="postfix">}</mo></mrow></munder><mi>L</mi><mrow><mo stretchy="true" form="prefix">(</mo><msub><mi>x</mi><mi>r</mi></msub><mo>=</mo><mi>k</mi><mo stretchy="true" form="postfix">)</mo></mrow></mrow></mfrac></mrow><annotation encoding="application/x-tex">
P(x_r = A) =  \frac{L(x_r=A)}{\sum_{k \in \{A,C,G,T\}}L(x_r=k)}
</annotation></semantics></math> and the highest posterior probability
(“bayes”) criterion:
<math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>P</mi><mrow><mo stretchy="true" form="prefix">(</mo><msub><mi>x</mi><mi>r</mi></msub><mo>=</mo><mi>A</mi><mo stretchy="true" form="postfix">)</mo></mrow><mo>=</mo><mfrac><mrow><msub><mi>π</mi><mi>A</mi></msub><mi>L</mi><mrow><mo stretchy="true" form="prefix">(</mo><msub><mi>x</mi><mi>r</mi></msub><mo>=</mo><mi>A</mi><mo stretchy="true" form="postfix">)</mo></mrow></mrow><mrow><munder><mo>∑</mo><mrow><mi>k</mi><mo>∈</mo><mo stretchy="false" form="prefix">{</mo><mi>A</mi><mo>,</mo><mi>C</mi><mo>,</mo><mi>G</mi><mo>,</mo><mi>T</mi><mo stretchy="false" form="postfix">}</mo></mrow></munder><msub><mi>π</mi><mi>k</mi></msub><mi>L</mi><mrow><mo stretchy="true" form="prefix">(</mo><msub><mi>x</mi><mi>r</mi></msub><mo>=</mo><mi>k</mi><mo stretchy="true" form="postfix">)</mo></mrow></mrow></mfrac><mo>,</mo></mrow><annotation encoding="application/x-tex">
P(x_r=A) =  \frac{\pi_A L(x_r=A)}{\sum_{k \in \{A,C,G,T\}}\pi_k L(x_r=k)},
</annotation></semantics></math> where
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>L</mi><mrow><mo stretchy="true" form="prefix">(</mo><msub><mi>x</mi><mi>r</mi></msub><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">L(x_r)</annotation></semantics></math>
is the joint probability of states at the tips and the state at the root
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>x</mi><mi>r</mi></msub><annotation encoding="application/x-tex">x_r</annotation></semantics></math>
and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>π</mi><mi>i</mi></msub><annotation encoding="application/x-tex">\pi_i</annotation></semantics></math>
are the estimated base frequencies of state
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>i</mi><annotation encoding="application/x-tex">i</annotation></semantics></math>.
Both methods agree if all states (base frequencies) have equal
probabilities.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">anc.ml</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ancestral.pml.html">anc_pml</a></span><span class="op">(</span><span class="va">fit</span><span class="op">)</span></span></code></pre></div>
<p>The differences of the two approaches for a specific site (17) are
represented in the following figures.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot.ancestral.html">plotAnc</a></span><span class="op">(</span><span class="va">anc.ml</span>, <span class="fl">17</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/title.html" class="external-link">title</a></span><span class="op">(</span><span class="st">"ML"</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="Ancestral_files/figure-html/plotML-1.png" alt="Fig 4. Ancestral reconstruction the using the maximum likelihood." width="576"><p class="caption">
Fig 4. Ancestral reconstruction the using the maximum likelihood.
</p>
</div>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#plotAnc(anc.bayes, 17)</span></span>
<span><span class="co">#title("Bayes")</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="fitting-for-discrete-comparative-data">Fitting for discrete comparative data<a class="anchor" aria-label="anchor" href="#fitting-for-discrete-comparative-data"></a>
</h2>
<p>Often have already a phylogeny and only want estimate the ancestral
reconstruction for this tree. This is a common problem in phylogentic
comparative methods and we can use the function <em>ace</em> in the ape
<span class="citation">(Paradis and Schliep 2019)</span>,
<em>fitDiscrete</em> in the geiger <span class="citation">(Pennell et
al. 2014)</span> or <em>fitMK</em> in the phytools <span class="citation">(Revell 2012)</span> package. Here we want to show how
to fit these models using <em>optim.pml</em>.</p>
<p>First we load a tree and create some data.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"bird.orders"</span><span class="op">)</span></span>
<span><span class="va">x</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">5</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">18</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">x</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">20</span>,<span class="fl">22</span>,<span class="fl">23</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fl">2</span></span>
<span><span class="va">x</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">bird.orders</span><span class="op">$</span><span class="va">tip.label</span></span>
<span><span class="va">dat</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/as.phyDat.html">phyDat</a></span><span class="op">(</span><span class="va">x</span>, <span class="st">"USER"</span>, levels<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">1</span>,<span class="fl">2</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>We than set up the <em>pml</em> object and optimize the model.
Instead of optimizing the edge length we only optimize the rate.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/pml.html">pml</a></span><span class="op">(</span><span class="va">bird.orders</span>, <span class="va">dat</span><span class="op">)</span></span>
<span><span class="va">fit_ER</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/pml.html">optim.pml</a></span><span class="op">(</span><span class="va">fit</span>, optEdge <span class="op">=</span> <span class="cn">FALSE</span>, optRate<span class="op">=</span><span class="cn">TRUE</span>, </span>
<span>                    control <span class="op">=</span> <span class="fu"><a href="../reference/pml.control.html">pml.control</a></span><span class="op">(</span>trace<span class="op">=</span><span class="fl">0</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">fit_ER</span></span></code></pre></div>
<pre><code><span><span class="co">## model: Mk </span></span>
<span><span class="co">## loglikelihood: -16.47 </span></span>
<span><span class="co">## unconstrained loglikelihood: 0 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Rate matrix:</span></span>
<span><span class="co">##   0 1 2</span></span>
<span><span class="co">## 0 0 1 1</span></span>
<span><span class="co">## 1 1 0 1</span></span>
<span><span class="co">## 2 1 1 0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Base frequencies:  </span></span>
<span><span class="co">##      0      1      2 </span></span>
<span><span class="co">## 0.3333 0.3333 0.3333 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Rate: 0.007846</span></span></code></pre>
<p>We can also fit the symmetric (<em>model=“SYM”</em>) or ordered
metristic model (<em>model=“ORDERED”</em>).</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fit_SYM</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/pml.html">optim.pml</a></span><span class="op">(</span><span class="va">fit</span>, optEdge <span class="op">=</span> <span class="cn">FALSE</span>, optRate<span class="op">=</span><span class="cn">TRUE</span>, model<span class="op">=</span><span class="st">"SYM"</span>, </span>
<span>                    control <span class="op">=</span> <span class="fu"><a href="../reference/pml.control.html">pml.control</a></span><span class="op">(</span>trace<span class="op">=</span><span class="fl">0</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">fit_SYM</span></span></code></pre></div>
<pre><code><span><span class="co">## model: SYM </span></span>
<span><span class="co">## loglikelihood: -15.31 </span></span>
<span><span class="co">## unconstrained loglikelihood: 0 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Rate matrix:</span></span>
<span><span class="co">##           0      1         2</span></span>
<span><span class="co">## 0 0.000e+00 0.2747 1.604e-06</span></span>
<span><span class="co">## 1 2.747e-01 0.0000 1.000e+00</span></span>
<span><span class="co">## 2 1.604e-06 1.0000 0.000e+00</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Base frequencies:  </span></span>
<span><span class="co">##      0      1      2 </span></span>
<span><span class="co">## 0.3333 0.3333 0.3333 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Rate: 0.00678</span></span></code></pre>
<p>We can compare the estimate with the one from <em>ace</em> from
<em>ape</em>.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fit_ace</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ape/man/ace.html" class="external-link">ace</a></span><span class="op">(</span><span class="va">x</span>, <span class="va">bird.orders</span>, model<span class="op">=</span><span class="st">"SYM"</span>, type <span class="op">=</span> <span class="st">"d"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Warning in sqrt(diag(solve(h))): NaNs produced</span></span></code></pre>
<p>The log-likelihood values differ slightly as in phangorn the values
get multiplied by the state frequencies. Thus if we add
<code>log(1/3)</code> as we have three states to ace estimate the two
estimates are almost identical.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fit_SYM</span><span class="op">$</span><span class="va">logLik</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] -15.31</span></span></code></pre>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fit_ace</span><span class="op">$</span><span class="va">loglik</span><span class="op">+</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log</a></span><span class="op">(</span><span class="fl">1</span><span class="op">/</span><span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] -15.31</span></span></code></pre>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/all.equal.html" class="external-link">all.equal</a></span><span class="op">(</span><span class="va">fit_SYM</span><span class="op">$</span><span class="va">logLik</span>, <span class="va">fit_ace</span><span class="op">$</span><span class="va">loglik</span><span class="op">+</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log</a></span><span class="op">(</span><span class="fl">1</span><span class="op">/</span><span class="fl">3</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "Mean relative difference: 1.229e-07"</span></span></code></pre>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">anc_SYM</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ancestral.pml.html">anc_pml</a></span><span class="op">(</span><span class="va">fit_SYM</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/plot.ancestral.html">plotAnc</a></span><span class="op">(</span><span class="va">anc_SYM</span><span class="op">)</span></span></code></pre></div>
<p><img src="Ancestral_files/figure-html/SYM_reconstruction-1.png" width="576"></p>
<p>More complicated models can be applied using defining the rate matrix
as shown in the vignette <em>Markov models and transition rate
matrices</em>. The “ARD” model is currently not available as
<em>phangorn</em> only fits reversible models.</p>
</div>
<div class="section level2">
<h2 id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<pre><code><span><span class="co">## R version 4.5.0 (2025-04-11)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">## Running under: Ubuntu 24.04.2 LTS</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.26.so;  LAPACK version 3.12.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">##  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">## [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: UTC</span></span>
<span><span class="co">## tzcode source: system (glibc)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">## [1] phangorn_2.12.1.1 ape_5.8-1.3      </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##  [1] Matrix_1.7-3      gtable_0.3.6      jsonlite_2.0.0    dplyr_1.1.4      </span></span>
<span><span class="co">##  [5] compiler_4.5.0    tidyselect_1.2.1  Rcpp_1.0.14       parallel_4.5.0   </span></span>
<span><span class="co">##  [9] jquerylib_0.1.4   scales_1.3.0      systemfonts_1.2.2 textshaping_1.0.0</span></span>
<span><span class="co">## [13] yaml_2.3.10       fastmap_1.2.0     lattice_0.22-6    ggplot2_3.5.2    </span></span>
<span><span class="co">## [17] R6_2.6.1          labeling_0.4.3    generics_0.1.3    igraph_2.1.4     </span></span>
<span><span class="co">## [21] knitr_1.50        htmlwidgets_1.6.4 backports_1.5.0   tibble_3.2.1     </span></span>
<span><span class="co">## [25] checkmate_2.3.2   desc_1.4.3        munsell_0.5.1     pillar_1.10.2    </span></span>
<span><span class="co">## [29] bslib_0.9.0       rlang_1.1.6       fastmatch_1.1-6   cachem_1.1.0     </span></span>
<span><span class="co">## [33] xfun_0.52         fs_1.6.6          quadprog_1.5-8    sass_0.4.10      </span></span>
<span><span class="co">## [37] cli_3.6.4         withr_3.0.2       pkgdown_2.1.1     magrittr_2.0.3   </span></span>
<span><span class="co">## [41] digest_0.6.37     grid_4.5.0        lifecycle_1.0.4   nlme_3.1-168     </span></span>
<span><span class="co">## [45] vctrs_0.6.5       evaluate_1.0.3    glue_1.8.0        farver_2.1.2     </span></span>
<span><span class="co">## [49] codetools_0.2-20  ragg_1.4.0        ggseqlogo_0.2     colorspace_2.1-1 </span></span>
<span><span class="co">## [53] rmarkdown_2.29    tools_4.5.0       pkgconfig_2.0.3   htmltools_0.5.8.1</span></span></code></pre>
</div>
<div class="section level2">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent" entry-spacing="0">
<div id="ref-Felsenstein2004" class="csl-entry">
Felsenstein, Joseph. 2004. <em>Inferring Phylogenies</em>. Sunderland:
Sinauer Associates.
</div>
<div id="ref-Koshi1996" class="csl-entry">
Koshi, Jeffrey M., and Richard A. Goldstein. 1996. <span>“<span class="nocase">Probabilistic reconstruction of ancestral protein
sequences</span>.”</span> <em>Journal of Molecular Evolution</em> 42
(2): 313–20. <a href="https://doi.org/10.1007/BF02198858" class="external-link">https://doi.org/10.1007/BF02198858</a>.
</div>
<div id="ref-Paradis2018" class="csl-entry">
Paradis, Emmanuel, and Klaus Schliep. 2019. <span>“Ape 5.0: An
Environment for Modern Phylogenetics and Evolutionary Analyses in
r.”</span> <em>Bioinformatics</em> 35 (3): 526–28. <a href="https://doi.org/10.1093/bioinformatics/bty633" class="external-link">https://doi.org/10.1093/bioinformatics/bty633</a>.
</div>
<div id="ref-Pennell2014" class="csl-entry">
Pennell, M. W., J. M. Eastman, G. J. Slater, J. W. Brown, J. C. Uyeda,
R. G. Fitzjohn, M. E. Alfaro, and L. J. Harmon. 2014. <span>“Geiger
V2.0: An Expanded Suite of Methods for Fitting Macroevolutionary Models
to Phylogenetic Trees.”</span> <em>Bioinformatics</em> 30: 2216–18. <a href="https://doi.org/10.1093/bioinformatics/btu181" class="external-link">https://doi.org/10.1093/bioinformatics/btu181</a>.
</div>
<div id="ref-Pupko2000" class="csl-entry">
Pupko, Tal, Itsik Pe, Ron Shamir, and Dan Graur. 2000. <span>“<span class="nocase">A Fast Algorithm for Joint Reconstruction of Ancestral
Amino Acid Sequences</span>.”</span> <em>Molecular Biology and
Evolution</em> 17 (6): 890–96. <a href="https://doi.org/10.1093/oxfordjournals.molbev.a026369" class="external-link">https://doi.org/10.1093/oxfordjournals.molbev.a026369</a>.
</div>
<div id="ref-Revell2012" class="csl-entry">
Revell, Liam J. 2012. <span>“Phytools: An r Package for Phylogenetic
Comparative Biology (and Other Things).”</span> <em>Methods in Ecology
and Evolution</em> 3: 217–23.
</div>
<div id="ref-Schliep2011" class="csl-entry">
Schliep, Klaus Peter. 2011. <span>“Phangorn: Phylogenetic Analysis in
<span>R</span>.”</span> <em>Bioinformatics</em> 27 (4): 592–93. <a href="https://doi.org/10.1093/bioinformatics/btq706" class="external-link">https://doi.org/10.1093/bioinformatics/btq706</a>.
</div>
<div id="ref-Swofford1987" class="csl-entry">
Swofford, D. L., and W. P. Maddison. 1987. <span>“Reconstructing
Ancestral Character States Under Wagner Parsimony.”</span> <em>Math.
Biosci.</em> 87: 199–229.
</div>
<div id="ref-ggseqlogo" class="csl-entry">
Wagih, Omar. 2024. <em>Ggseqlogo: A ’Ggplot2’ Extension for Drawing
Publication-Ready Sequence Logos</em>. <a href="https://CRAN.R-project.org/package=ggseqlogo" class="external-link">https://CRAN.R-project.org/package=ggseqlogo</a>.
</div>
<div id="ref-Yang2006" class="csl-entry">
Yang, Ziheng. 2006. <em>Computational Molecular Evolution</em>. Oxford:
Oxford University Press.
</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Klaus Schliep, Emmanuel Paradis, Leonardo de Oliveira Martins, Alastair Potts, Iris Bardel-Kahr.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
