<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Intertwining phylogenetic trees and networks: R Example Script • phangorn</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Intertwining phylogenetic trees and networks: R Example Script">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">phangorn</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.12.1.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/Trees.html">Estimating phylogenetic trees with phangorn</a></li>
    <li><a class="dropdown-item" href="../articles/Morphological.html">Phylogenetic trees from morphological data</a></li>
    <li><a class="dropdown-item" href="../articles/AdvancedFeatures.html">Markov models and transition rate matrices</a></li>
    <li><a class="dropdown-item" href="../articles/Ancestral.html">Ancestral Sequence Reconstruction</a></li>
    <li><a class="dropdown-item" href="../articles/Networx.html">Splits and Networx</a></li>
    <li><a class="dropdown-item" href="../articles/MLbyHand.html">Maximum likelihood by hand</a></li>
    <li><a class="dropdown-item" href="../articles/IntertwiningTreesAndNetworks.html">Intertwining phylogenetic trees and networks: R Example Script</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/KlausVigo/phangorn/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Intertwining phylogenetic trees and networks: R Example Script</h1>
                        <h4 data-toc-skip class="author">Klaus Schliep,
Alastair Potts, David Morrison and Guido Grimm</h4>
            
            <h4 data-toc-skip class="date">October 28, 2024</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/KlausVigo/phangorn/blob/master/vignettes/IntertwiningTreesAndNetworks.Rmd" class="external-link"><code>vignettes/IntertwiningTreesAndNetworks.Rmd</code></a></small>
      <div class="d-none name"><code>IntertwiningTreesAndNetworks.Rmd</code></div>
    </div>

    
    
<p><em>Description:</em> This script provides examples of the new
functions available in the phangorn library to ‘intertwine’ trees and
networks, i.e. compare trees and networks and data transference. It also
provides a step-by-step guide for users new to R.</p>
<p><em>Methodological advancement:</em> The major advancement in this
phangorn update is the introduction of a generic network object with a
wide range of related transfer and analysis functions. These new
functions provide the first means to directly transfer information
amongst a wide range of phylogenetic trees and networks, as well as
means to visualize and further analyze this information. This should
provide a platform for individuals to easily conduct tree-network
comparisons and stimulate further function development by the
community.</p>
<p><em>What next?:</em> By implementing full network handling
compatibility in R, and providing exemplar scripts (such as this) and <a href="https://github.com/KlausVigo/phangorn" class="external-link">support</a>, the scientific
community now has an easy means to analyse and compare the results of
phylogenetic trees vs. network approaches. We hope this will open a
largely unexplored world to the general phylogenetic audience.</p>
<div class="section level3">
<h3 id="installing-r">Installing R<a class="anchor" aria-label="anchor" href="#installing-r"></a>
</h3>
<ol style="list-style-type: decimal">
<li>Download R<br>
Select the nearest mirror to your location at <a href="https://cran.r-project.org/mirrors.html" class="external-link uri">https://cran.r-project.org/mirrors.html</a><br>
</li>
<li>Select your operating system and download the relevant installation
file.<br>
</li>
<li>Install R following the instructions.</li>
</ol>
</div>
<div class="section level3">
<h3 id="installing-the-phangorn-library">Installing the phangorn library<a class="anchor" aria-label="anchor" href="#installing-the-phangorn-library"></a>
</h3>
<p>Open R and run the two lines of code below in the command line.<br>
(You will need to select a region from which to download the
library)</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"phangorn"</span>, dependencies<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co"># install latest development version needs devtools</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"devtools"</span>, dependencies<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://devtools.r-lib.org/" class="external-link">devtools</a></span><span class="op">)</span></span>
<span><span class="fu">install_github</span><span class="op">(</span><span class="st">"KlausVigo/phangorn"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="getting-started">Getting started<a class="anchor" aria-label="anchor" href="#getting-started"></a>
</h3>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/KlausVigo/phangorn" class="external-link">phangorn</a></span><span class="op">)</span>    <span class="co"># load the phangorn library</span></span></code></pre></div>
<pre><code><span><span class="co">## Loading required package: ape</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="set-the-working-directory">Set the working directory<a class="anchor" aria-label="anchor" href="#set-the-working-directory"></a>
</h3>
<p>This is often a major stumbling block for new R users. You need to
specify where in your folder structure you wish to work. i.e, where are
the files stored that you wish to input?</p>
<p>This is done using the <code><a href="https://rdrr.io/r/base/getwd.html" class="external-link">setwd()</a></code> function,
e.g. <code>setwd("C:/TreesNetworks/Example Files")</code>.</p>
<p>We now set it to the folder of the phangorn package, which contains
the files we want to load for this example.</p>
</div>
<div class="section level3">
<h3 id="read-in-the-example-file-datasets">Read in the example file datasets:<a class="anchor" aria-label="anchor" href="#read-in-the-example-file-datasets"></a>
</h3>
<p>This example files are based on the woodmouse dataset available in
the ape library. Ultimately, this dataset is from this study: Michaux,
J. R., Magnanou, E., Paradis, E., Nieberding, C. and Libois, R. (2003)
Mitochondrial phylogeography of the Woodmouse (<em>Apodemus
sylvaticus</em>) in the Western Palearctic region. Molecular Ecology,
12, 685-697.)</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## automatically set the correct working directory for the examples below</span></span>
<span><span class="co"># setwd(system.file("extdata/trees", package = "phangorn"))</span></span>
<span><span class="co"># for this vignette we create a path to the files we want to load </span></span>
<span><span class="va">fdir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/trees"</span>, package <span class="op">=</span> <span class="st">"phangorn"</span><span class="op">)</span></span>
<span><span class="co">## in your case it may look something like this...</span></span>
<span><span class="co"># setwd("C:/TreesNetworks/Example Files")</span></span>
<span></span>
<span><span class="co">##DNA Matrix, maybe not needed </span></span>
<span><span class="va">woodmouse</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/read.phyDat.html">read.phyDat</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">fdir</span>, <span class="st">"woodmouse.fasta"</span><span class="op">)</span>,format<span class="op">=</span><span class="st">"fasta"</span><span class="op">)</span> </span>
<span><span class="co">## RAxML best-known tree with bipartition support (from previous analysis)</span></span>
<span><span class="va">raxml.tree</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ape/man/read.tree.html" class="external-link">read.tree</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">fdir</span>,<span class="st">"RAxML_bipartitions.woodmouse"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">## RAxML bootstrap trees (from previous analysis)</span></span>
<span><span class="va">raxml.bootstrap</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ape/man/read.tree.html" class="external-link">read.tree</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">fdir</span>,<span class="st">"RAxML_bootstrap.woodmouse"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">## MrBayes consensus tree (50% majority rule) (from previous analysis)</span></span>
<span><span class="va">mrbayes.tree</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ape/man/read.nexus.html" class="external-link">read.nexus</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">fdir</span>,<span class="st">"woodmouse.mrbayes.nex.con"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">## MrBayes sample runs 1 and 2 (from previous analysis)</span></span>
<span><span class="va">run1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ape/man/read.nexus.html" class="external-link">read.nexus</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">fdir</span>,<span class="st">"woodmouse.mrbayes.nex.run1.t"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">run2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ape/man/read.nexus.html" class="external-link">read.nexus</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">fdir</span>,<span class="st">"woodmouse.mrbayes.nex.run2.t"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">## How many trees are in the MrBayes tree sample?</span></span>
<span><span class="va">run1</span></span></code></pre></div>
<pre><code><span><span class="co">## 1001 phylogenetic trees</span></span></code></pre>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">run2</span></span></code></pre></div>
<pre><code><span><span class="co">## 1001 phylogenetic trees</span></span></code></pre>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Combining the two runs and removing 25% burn-in</span></span>
<span><span class="va">mrbayes.trees</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">run1</span><span class="op">[</span><span class="fl">251</span><span class="op">:</span><span class="fl">1001</span><span class="op">]</span>,<span class="va">run2</span><span class="op">[</span><span class="fl">251</span><span class="op">:</span><span class="fl">1001</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">## NeigbourNet Nexus file generated by SplitsTree (from previous analysis)</span></span>
<span><span class="va">Nnet</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/read.nexus.splits.html">read.nexus.networx</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">fdir</span>,<span class="st">"woodmouse.nxs"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>All example files read into R.</p>
</div>
<div class="section level3">
<h3 id="viewing-the-data">Viewing the data<a class="anchor" aria-label="anchor" href="#viewing-the-data"></a>
</h3>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/par.html" class="external-link">par</a></span><span class="op">(</span>mfrow<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span><span class="op">)</span>, mar<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">1</span>,<span class="fl">1</span>,<span class="fl">1</span><span class="op">)</span><span class="op">)</span> <span class="co"># Setting plot parameters</span></span>
<span><span class="co">### Plotting trees with support values:</span></span>
<span><span class="co">##  RAxML</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">raxml.tree</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/ape/man/nodelabels.html" class="external-link">nodelabels</a></span><span class="op">(</span><span class="va">raxml.tree</span><span class="op">$</span><span class="va">node.label</span>, adj <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">0</span><span class="op">)</span>, frame <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span></span>
<span><span class="co">##  MrBayes</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">mrbayes.tree</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/ape/man/nodelabels.html" class="external-link">nodelabels</a></span><span class="op">(</span><span class="va">mrbayes.tree</span><span class="op">$</span><span class="va">node.label</span>, adj <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">0</span><span class="op">)</span>, frame <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span></span></code></pre></div>
<p><img src="IntertwiningTreesAndNetworks_files/figure-html/unnamed-chunk-4-1.png" width="576"></p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/par.html" class="external-link">par</a></span><span class="op">(</span>mfrow<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">1</span><span class="op">)</span><span class="op">)</span> <span class="co"># Setting plot parameters</span></span>
<span><span class="co"># NeighborNet</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">Nnet</span>,<span class="st">"2D"</span><span class="op">)</span></span></code></pre></div>
<p><img src="IntertwiningTreesAndNetworks_files/figure-html/unnamed-chunk-4-2.png" width="576"></p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## alternatively,</span></span>
<span><span class="co"># plot(Nnet,"2D")</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="intertwining-trees-and-network-functions">Intertwining trees and network functions<a class="anchor" aria-label="anchor" href="#intertwining-trees-and-network-functions"></a>
</h3>
</div>
<div class="section level3">
<h3 id="a">1A:<a class="anchor" aria-label="anchor" href="#a"></a>
</h3>
<p>Identification of edge bundles (in black) in a neighbor-net (NN)
network that correspond to branches (labelled 1-12) in a tree (a maximum
likelihood tree, in this case).</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># create a vector of labels for the network corresponding to edges in the tree</span></span>
<span><span class="va">edge.lab</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/addConfidences.html">createLabel</a></span><span class="op">(</span><span class="va">Nnet</span>, <span class="va">raxml.tree</span>, <span class="va">raxml.tree</span><span class="op">$</span><span class="va">edge</span><span class="op">[</span>,<span class="fl">2</span><span class="op">]</span>, <span class="st">"edge"</span><span class="op">)</span></span>
<span><span class="co"># could be also 1:27 instead of raxml.tree$edge[,2]</span></span>
<span></span>
<span><span class="co"># Show the correspondingly labelled tree and network in R</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/par.html" class="external-link">par</a></span><span class="op">(</span>mfrow<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span><span class="op">)</span><span class="op">)</span>  </span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">raxml.tree</span>, <span class="st">"u"</span>, rotate.tree <span class="op">=</span> <span class="fl">180</span>, cex<span class="op">=</span><span class="fl">.7</span><span class="op">)</span> </span>
<span><span class="fu"><a href="https://rdrr.io/pkg/ape/man/nodelabels.html" class="external-link">edgelabels</a></span><span class="op">(</span><span class="va">raxml.tree</span><span class="op">$</span><span class="va">edge</span><span class="op">[</span>,<span class="fl">2</span><span class="op">]</span>,col<span class="op">=</span><span class="st">"blue"</span>, frame<span class="op">=</span><span class="st">"none"</span>, cex<span class="op">=</span><span class="fl">.7</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># find edges that are in the network but not in the tree</span></span>
<span><span class="va">edge.col</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"black"</span>, <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">Nnet</span><span class="op">$</span><span class="va">edge</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">edge.col</span><span class="op">[</span> <span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">edge.lab</span><span class="op">)</span> <span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"red"</span></span>
<span><span class="co"># or a simpler alternative...</span></span>
<span><span class="va">edge.col</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/addConfidences.html">createLabel</a></span><span class="op">(</span><span class="va">Nnet</span>, <span class="va">raxml.tree</span>, <span class="st">"black"</span>, nomatch<span class="op">=</span><span class="st">"red"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">x</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">Nnet</span>, edge.label <span class="op">=</span> <span class="va">edge.lab</span>, show.edge.label <span class="op">=</span> <span class="cn">T</span>, <span class="st">"2D"</span>, </span>
<span>          edge.color <span class="op">=</span> <span class="va">edge.col</span>, col.edge.label <span class="op">=</span> <span class="st">"blue"</span>, cex<span class="op">=</span><span class="fl">.7</span><span class="op">)</span></span></code></pre></div>
<p><img src="IntertwiningTreesAndNetworks_files/figure-html/unnamed-chunk-5-1.png" width="672"></p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># the above plot function returns an invisible networx object and this object </span></span>
<span><span class="co"># also contains the colors for the edges.</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="b">1B:<a class="anchor" aria-label="anchor" href="#b"></a>
</h3>
<p>Bootstrap support for all branches (branch labels) in the ML tree
mapped on the corresponding edge bundles of the NN network. The edges in
the network which are not found as ML tree branches are highlighted in
red.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># the scaler argument multiplies the confidence values. This is useful to switch</span></span>
<span><span class="co"># confidences values between total, percentage or ratios.   </span></span>
<span><span class="va">x</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/addConfidences.html">addConfidences</a></span><span class="op">(</span><span class="va">Nnet</span>,<span class="va">raxml.tree</span>, scaler <span class="op">=</span> <span class="fl">.01</span><span class="op">)</span></span>
<span><span class="co"># find splits that are in the network but not in the tree</span></span>
<span><span class="va">split.col</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"black"</span>, <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">x</span><span class="op">$</span><span class="va">splits</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">split.col</span><span class="op">[</span> <span class="op">!</span><span class="fu"><a href="../reference/as.splits.html">matchSplits</a></span><span class="op">(</span><span class="fu"><a href="../reference/as.splits.html">as.splits</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span>, <span class="fu"><a href="../reference/as.splits.html">as.splits</a></span><span class="op">(</span><span class="va">raxml.tree</span><span class="op">)</span><span class="op">)</span> <span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"red"</span></span>
<span></span>
<span><span class="co"># simpler alternative...</span></span>
<span><span class="va">split.col2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/addConfidences.html">createLabel</a></span><span class="op">(</span><span class="va">x</span>, <span class="va">raxml.tree</span>, label<span class="op">=</span><span class="st">"black"</span>, <span class="st">"split"</span>, nomatch<span class="op">=</span><span class="st">"red"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Plotting in R</span></span>
<span><span class="va">out.x</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">x</span>,<span class="st">"2D"</span>,show.edge.label<span class="op">=</span><span class="cn">TRUE</span>, split.color<span class="op">=</span><span class="va">split.col</span>, </span>
<span>              col.edge.label <span class="op">=</span> <span class="st">"blue"</span><span class="op">)</span></span></code></pre></div>
<p><img src="IntertwiningTreesAndNetworks_files/figure-html/unnamed-chunk-6-1.png" width="576"></p>
<p>Again we can write to SplitsTree for viewing…</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># write.nexus.networx(out.x,"woodmouse.tree.support.nxs")</span></span>
<span><span class="co">## or we can also export the splits alone (for usage in software other than SplitsTree)</span></span>
<span><span class="co"># write.nexus.splits(as.splits(out.x),"woodmouse.splits.support.nxs")</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="c">1C:<a class="anchor" aria-label="anchor" href="#c"></a>
</h3>
<p>Frequencies of bipartitions found in the bootstrap pseudoreplicates
mapped on the corresponding edge bundles of the NN network using a
threshold of 10% (i.e. any edge is labelled that occurs in at least 100
of the 1000 ML-BS pseudoreplicates). Edge bundles not found in the ML
tree are labelled using grey edges.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">y</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/addConfidences.html">addConfidences</a></span><span class="op">(</span><span class="va">Nnet</span>, <span class="fu"><a href="../reference/as.splits.html">as.splits</a></span><span class="op">(</span><span class="va">raxml.bootstrap</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">edge.col</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/addConfidences.html">createLabel</a></span><span class="op">(</span><span class="va">y</span>, <span class="va">raxml.tree</span>, label<span class="op">=</span><span class="st">"black"</span>, <span class="st">"edge"</span>, nomatch<span class="op">=</span><span class="st">"grey"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">y</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">y</span>,<span class="st">"2D"</span>,show.edge.label<span class="op">=</span><span class="cn">TRUE</span>, edge.color<span class="op">=</span><span class="va">edge.col</span><span class="op">)</span></span></code></pre></div>
<p><img src="IntertwiningTreesAndNetworks_files/figure-html/unnamed-chunk-8-1.png" width="576"></p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Write to SplitsTree for viewing</span></span>
<span><span class="co"># write.nexus.networx(y,"NN.with.bs.support.nxs")</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="extras">Extras…<a class="anchor" aria-label="anchor" href="#extras"></a>
</h3>
<p>We can also compare the neighborNet with a consensusNet (Holland BR,
Huber KT, Moulton V, Lockhart PJ,2004, Using consensus networks to
visualize contradictory evidence for species phylogeny. Molecular
Biology and Evolution, 21, 1459-1461). Furthermore, we can extract the
support values from the consensus network, and place these on to the
NeighborNet (this is similar to the process explained in 1C above).</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cnet</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/consensusNet.html">consensusNet</a></span><span class="op">(</span><span class="va">raxml.bootstrap</span>,prob<span class="op">=</span><span class="fl">0.10</span><span class="op">)</span></span>
<span><span class="va">edge.col</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/addConfidences.html">createLabel</a></span><span class="op">(</span><span class="va">cnet</span>, <span class="va">Nnet</span>, label<span class="op">=</span><span class="st">"black"</span>, <span class="st">"edge"</span>, nomatch<span class="op">=</span><span class="st">"grey"</span><span class="op">)</span></span>
<span><span class="va">cnet</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">cnet</span>, <span class="st">"2D"</span>, show.edge.label <span class="op">=</span> <span class="cn">TRUE</span>, edge.color<span class="op">=</span><span class="va">edge.col</span><span class="op">)</span></span></code></pre></div>
<p><img src="IntertwiningTreesAndNetworks_files/figure-html/unnamed-chunk-9-1.png" width="576"></p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">edge.col</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/addConfidences.html">createLabel</a></span><span class="op">(</span><span class="va">Nnet</span>, <span class="va">cnet</span>, label<span class="op">=</span><span class="st">"black"</span>, <span class="st">"edge"</span>, nomatch<span class="op">=</span><span class="st">"grey"</span><span class="op">)</span></span>
<span><span class="va">z</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">Nnet</span>, <span class="st">"2D"</span>, show.edge.label <span class="op">=</span> <span class="cn">TRUE</span>, edge.color<span class="op">=</span><span class="va">edge.col</span><span class="op">)</span></span></code></pre></div>
<p><img src="IntertwiningTreesAndNetworks_files/figure-html/unnamed-chunk-9-2.png" width="576"></p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/addConfidences.html">addConfidences</a></span><span class="op">(</span><span class="va">Nnet</span>,<span class="va">cnet</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">obj</span>,<span class="st">"2D"</span>,show.edge.label<span class="op">=</span><span class="cn">T</span>, edge.color<span class="op">=</span><span class="va">edge.col</span>, col.edge.label <span class="op">=</span> <span class="st">"blue"</span><span class="op">)</span></span></code></pre></div>
<p><img src="IntertwiningTreesAndNetworks_files/figure-html/unnamed-chunk-9-3.png" width="576"></p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Write to SplitsTree for viewing</span></span>
<span><span class="co"># write.nexus.networx(obj,"Nnet.with.ML.Cnet.Bootstrap.nxs")</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="section"></h3>
<p>There are four possible data patterns in phylogenetic reconstruction:
(1) patterns that are well supported in the network and appear in the
bootstrapped trees; (2) patterns that are well supported by (part of)
the data/network but do not appear in the optimized trees, i.e. they are
incompatible with the tree; (3) patterns that are weakly supported by
the network but appear in the optimized trees anyway, i.e. they are
compatible with the tree.</p>
<p>Here we demonstrate these patterns by showing the relationships
between splits weights, from a NeighborNet splits graph, bootstrap
bipartitions support and bootstrap percentages plotted on the optimized
tree for a dataset of <span class="citation">Wang, Braun, and Kimball
(2012)</span>.</p>
<p>(An advanced user figure…)</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">Nnet</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/read.nexus.splits.html">read.nexus.networx</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">fdir</span>,<span class="st">"RAxML_distances.Wang.nxs"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">raxml.tree</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ape/man/read.tree.html" class="external-link">read.tree</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">fdir</span>,<span class="st">"RAxML_bestTree.Wang.out"</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/pkg/ape/man/root.html" class="external-link">unroot</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">raxml.bootstrap</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ape/man/read.tree.html" class="external-link">read.tree</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">fdir</span>,<span class="st">"RAxML_bootstrap.Wang.out"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">bs_splits</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/as.splits.html">as.splits</a></span><span class="op">(</span><span class="va">raxml.bootstrap</span><span class="op">)</span></span>
<span><span class="va">tree_splits</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/as.splits.html">as.splits</a></span><span class="op">(</span><span class="va">raxml.tree</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="../reference/as.splits.html">removeTrivialSplits</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co"># we overwrite bootstrap values and set the weights </span></span>
<span><span class="co"># to 1e-6 (almost zero), as we plot them on a log scale later on</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/attr.html" class="external-link">attr</a></span><span class="op">(</span><span class="va">bs_splits</span>, <span class="st">"weights"</span><span class="op">)</span><span class="op">[</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fl">1e-6</span></span>
<span><span class="co"># combine the splits from the bootstrap and neighbor net</span></span>
<span><span class="co"># and delete duplicates and add the confidence values</span></span>
<span><span class="co"># we get rid of trivial splits</span></span>
<span><span class="va">all_splits</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">Nnet</span><span class="op">$</span><span class="va">splits</span>, <span class="va">bs_splits</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="../reference/as.splits.html">removeTrivialSplits</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="../reference/addConfidences.html">addConfidences</a></span><span class="op">(</span><span class="va">bs_splits</span>, scaler<span class="op">=</span><span class="fl">100</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># For easier plotting we create a matrix with the confidences and </span></span>
<span><span class="co"># weights as columns</span></span>
<span><span class="va">tab</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>SplitWeight <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/attr.html" class="external-link">attr</a></span><span class="op">(</span><span class="va">all_splits</span>, <span class="st">"weights"</span><span class="op">)</span>, </span>
<span>                  Bootstrap<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/attr.html" class="external-link">attr</a></span><span class="op">(</span><span class="va">all_splits</span>, <span class="st">"confidences"</span><span class="op">)</span>, Tree<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="co"># we add a logical variable pto indicate which splits are in the RAxML tree</span></span>
<span><span class="va">tab</span><span class="op">$</span><span class="va">Tree</span><span class="op">[</span><span class="fu"><a href="../reference/as.splits.html">matchSplits</a></span><span class="op">(</span><span class="va">tree_splits</span>, <span class="va">all_splits</span>, <span class="cn">FALSE</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="cn">TRUE</span></span>
<span></span>
<span><span class="va">tab</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">tab</span><span class="op">[</span>,<span class="st">"Bootstrap"</span><span class="op">]</span><span class="op">)</span>,<span class="st">"Bootstrap"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fl">0</span></span>
<span><span class="va">tab</span><span class="op">[</span>,<span class="st">"Bootstrap"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="va">tab</span><span class="op">[</span>,<span class="st">"Bootstrap"</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">tab</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="va">all_splits</span>, zero.print <span class="op">=</span> <span class="st">"."</span>, one.print <span class="op">=</span> <span class="st">"|"</span><span class="op">)</span>, </span>
<span>                       <span class="fl">1</span>, <span class="va">paste0</span>, collapse<span class="op">=</span><span class="st">""</span><span class="op">)</span></span>
<span><span class="va">tab</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span>,<span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">##                              SplitWeight Bootstrap  Tree</span></span>
<span><span class="co">## ..||........................   0.0171433       100  TRUE</span></span>
<span><span class="co">## ..||||......................   0.0013902        14 FALSE</span></span>
<span><span class="co">## ..||||......|||||...........   0.0001589         0 FALSE</span></span>
<span><span class="co">## ||.........................|   0.0027691         1 FALSE</span></span>
<span><span class="co">## ||..........................   0.0840367       100  TRUE</span></span>
<span><span class="co">## ...|||......................   0.0001773         0 FALSE</span></span>
<span><span class="co">## ...|||........|.|...........   0.0003663         0 FALSE</span></span>
<span><span class="co">## |.|.........................   0.0060907         0 FALSE</span></span>
<span><span class="co">## ....||......................   0.0385909       100  TRUE</span></span>
<span><span class="co">## ||||........................   0.0018195        34  TRUE</span></span></code></pre>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">col</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"blue"</span>, <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">tab</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">col</span><span class="op">[</span><span class="va">tab</span><span class="op">[</span>,<span class="st">"Bootstrap"</span><span class="op">]</span><span class="op">==</span><span class="fl">0</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"green"</span></span>
<span><span class="va">col</span><span class="op">[</span><span class="va">tab</span><span class="op">[</span>,<span class="st">"SplitWeight"</span><span class="op">]</span><span class="op">==</span><span class="fl">1e-6</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"red"</span></span>
<span></span>
<span><span class="va">pch</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">19</span>, <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">tab</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">pch</span><span class="op">[</span><span class="va">tab</span><span class="op">$</span><span class="va">Tree</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fl">17</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/par.html" class="external-link">par</a></span><span class="op">(</span>mar<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">5.1</span>, <span class="fl">4.1</span>, <span class="fl">4.1</span>, <span class="fl">8.1</span><span class="op">)</span>, xpd<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">tab</span><span class="op">[</span>,<span class="st">"SplitWeight"</span><span class="op">]</span>, <span class="va">tab</span><span class="op">[</span>,<span class="st">"Bootstrap"</span><span class="op">]</span>, log<span class="op">=</span><span class="st">"x"</span>, col<span class="op">=</span><span class="va">col</span>, pch<span class="op">=</span><span class="va">pch</span>,  </span>
<span>     xlab<span class="op">=</span><span class="st">"Split weight (log scale)"</span>, ylab<span class="op">=</span><span class="st">"Bootstrap (%)"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/legend.html" class="external-link">legend</a></span><span class="op">(</span><span class="st">"topright"</span>, inset<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">0.35</span>,<span class="fl">0</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Pattern 1"</span>, <span class="st">"Pattern 2"</span>, <span class="st">"Pattern 3"</span>, </span>
<span>                       <span class="st">"Pattern in the\nbest tree"</span><span class="op">)</span>, pch<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">19</span>,<span class="fl">19</span>,<span class="fl">19</span>,<span class="fl">17</span><span class="op">)</span>, </span>
<span>                       col<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"blue"</span>, <span class="st">"green"</span>, <span class="st">"red"</span>, <span class="st">"black"</span><span class="op">)</span>, bty<span class="op">=</span><span class="st">"n"</span><span class="op">)</span></span></code></pre></div>
<p><img src="IntertwiningTreesAndNetworks_files/figure-html/unnamed-chunk-10-1.png" width="672"></p>
</div>
<div class="section level3">
<h3 id="figure-4">Figure 4<a class="anchor" aria-label="anchor" href="#figure-4"></a>
</h3>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">YCh</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ape/man/read.tree.html" class="external-link">read.tree</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">fdir</span>, <span class="st">"RAxML_bestTree.YCh"</span><span class="op">)</span><span class="op">)</span> </span>
<span><span class="va">mtG</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ape/man/read.tree.html" class="external-link">read.tree</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">fdir</span>, <span class="st">"RAxML_bestTree.mtG"</span><span class="op">)</span><span class="op">)</span> </span>
<span><span class="va">ncAI</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ape/man/read.tree.html" class="external-link">read.tree</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">fdir</span>, <span class="st">"RAxML_bestTree.AIs"</span><span class="op">)</span><span class="op">)</span> </span>
<span><span class="va">all_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ape/man/read.tree.html" class="external-link">read.tree</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">fdir</span>, <span class="st">"RAxML_bestTree.3moles"</span><span class="op">)</span><span class="op">)</span> </span>
<span><span class="va">YCh_boot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ape/man/read.tree.html" class="external-link">read.tree</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">fdir</span>, <span class="st">"RAxML_bootstrap.YCh"</span><span class="op">)</span><span class="op">)</span> </span>
<span><span class="va">mtG_boot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ape/man/read.tree.html" class="external-link">read.tree</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">fdir</span>, <span class="st">"RAxML_bootstrap.mtG"</span><span class="op">)</span><span class="op">)</span> </span>
<span><span class="va">ncAI_boot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ape/man/read.tree.html" class="external-link">read.tree</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">fdir</span>, <span class="st">"RAxML_bootstrap.AIs"</span><span class="op">)</span><span class="op">)</span> </span>
<span><span class="va">all_data_boot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ape/man/read.tree.html" class="external-link">read.tree</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">fdir</span>, <span class="st">"RAxML_bootstrap.3moles"</span><span class="op">)</span><span class="op">)</span> </span></code></pre></div>
<p>There are several option plotting a co-phylogeny. In the following we
use the <code>cophylo</code> function of the <code>phytools</code>
package.</p>
<pre><code><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/liamrevell/phytools" class="external-link">phytools</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/par.html" class="external-link">par</a></span><span class="op">(</span>mfrow<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span>,<span class="fl">1</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">obj</span> <span class="op">&lt;-</span> <span class="fu">cophylo</span><span class="op">(</span><span class="va">YCh</span>, <span class="va">mtG</span><span class="op">)</span> </span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">obj</span>, mar<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">.1</span>,<span class="fl">.1</span>,<span class="fl">2</span>,<span class="fl">.1</span><span class="op">)</span>,scale.bar<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">.005</span>,<span class="fl">.05</span><span class="op">)</span>, ylim<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">.2</span>,<span class="fl">1</span><span class="op">)</span><span class="op">)</span>  </span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/title.html" class="external-link">title</a></span><span class="op">(</span><span class="st">"A. YCh                                    B. mtG"</span><span class="op">)</span></span>
<span><span class="va">obj</span> <span class="op">&lt;-</span>  <span class="fu">cophylo</span><span class="op">(</span><span class="va">ncAI</span>, <span class="va">all_data</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">obj</span>, mar<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">.1</span>,<span class="fl">.1</span>,<span class="fl">2</span>,<span class="fl">.1</span><span class="op">)</span>,scale.bar<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">.005</span>,<span class="fl">.05</span><span class="op">)</span>, ylim<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">.2</span>,<span class="fl">1</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/title.html" class="external-link">title</a></span><span class="op">(</span><span class="st">"C. ncAI                              D. All data"</span><span class="op">)</span></span></code></pre>
<p><img src="cophylo.png"> Unfortunately this function does not (yet)
offer to add confidences to some splits, but we can do this easily with
more basic plot functions.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/par.html" class="external-link">par</a></span><span class="op">(</span>mfrow<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span>,<span class="fl">2</span><span class="op">)</span>, mar <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span>,<span class="fl">2</span>,<span class="fl">4</span>,<span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">YCh</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plotBS.html">plotBS</a></span><span class="op">(</span><span class="fu"><a href="../reference/midpoint.html">midpoint</a></span><span class="op">(</span><span class="va">YCh</span><span class="op">)</span>, <span class="va">YCh_boot</span>, <span class="st">"phylogram"</span>, p<span class="op">=</span><span class="fl">0</span>, main <span class="op">=</span> <span class="st">"YCh"</span><span class="op">)</span></span>
<span><span class="va">mtG</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plotBS.html">plotBS</a></span><span class="op">(</span><span class="fu"><a href="../reference/midpoint.html">midpoint</a></span><span class="op">(</span><span class="va">mtG</span><span class="op">)</span>, <span class="va">mtG_boot</span>, <span class="st">"phylogram"</span>, p<span class="op">=</span><span class="fl">0</span>, main <span class="op">=</span> <span class="st">"mtG"</span><span class="op">)</span></span>
<span><span class="va">ncAI</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plotBS.html">plotBS</a></span><span class="op">(</span><span class="fu"><a href="../reference/midpoint.html">midpoint</a></span><span class="op">(</span><span class="va">ncAI</span><span class="op">)</span>, <span class="va">ncAI_boot</span>, <span class="st">"phylogram"</span>, p<span class="op">=</span><span class="fl">0</span>, main <span class="op">=</span> <span class="st">"ncAI"</span><span class="op">)</span></span>
<span><span class="va">all_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plotBS.html">plotBS</a></span><span class="op">(</span><span class="fu"><a href="../reference/midpoint.html">midpoint</a></span><span class="op">(</span><span class="va">all_data</span><span class="op">)</span>, <span class="va">all_data_boot</span>, <span class="st">"phylogram"</span>, p<span class="op">=</span><span class="fl">0</span>, </span>
<span>                   main <span class="op">=</span> <span class="st">"All data"</span><span class="op">)</span></span></code></pre></div>
<p><img src="IntertwiningTreesAndNetworks_files/figure-html/unnamed-chunk-12-1.png" width="576"></p>
<p>We can compare this with consensus network with the different
bootstrap values for the different genes.</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/par.html" class="external-link">par</a></span><span class="op">(</span>mfrow<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">1</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">cn</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/consensusNet.html">consensusNet</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">YCh</span>, <span class="va">mtG</span>, <span class="va">ncAI</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">cn</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/addConfidences.html">addConfidences</a></span><span class="op">(</span><span class="va">cn</span>, <span class="va">YCh_boot</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="../reference/addConfidences.html">addConfidences</a></span><span class="op">(</span><span class="va">mtG_boot</span>, add<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="../reference/addConfidences.html">addConfidences</a></span><span class="op">(</span><span class="va">ncAI_boot</span>, add<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="../reference/addConfidences.html">addConfidences</a></span><span class="op">(</span><span class="va">all_data_boot</span>, add<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">cn</span>, <span class="st">"2D"</span>, show.edge.label<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="IntertwiningTreesAndNetworks_files/figure-html/unnamed-chunk-13-1.png" width="576"></p>
<div id="refs" class="references csl-bib-body hanging-indent" entry-spacing="0">
<div id="ref-Schliep2017" class="csl-entry">
Schliep, Klaus, Alastair J. Potts, David A. Morrison, and Guido W.
Grimm. 2017. <span>“Intertwining Phylogenetic Trees and
Networks.”</span> <em>Methods in Ecology and Evolution</em> 8 (10):
1212–20.
</div>
<div id="ref-Wang2012" class="csl-entry">
Wang, Ning, Edward L. Braun, and Rebecca T. Kimball. 2012.
<span>“Testing Hypotheses about the Sister Group of the Passeriformes
Using an Independent 30-Locus Data Set.”</span> <em>Molecular Biology
and Evolution</em> 29 (2): 737–50. <a href="https://doi.org/10.1093/molbev/msr230" class="external-link">https://doi.org/10.1093/molbev/msr230</a>.
</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Klaus Schliep, Emmanuel Paradis, Leonardo de Oliveira Martins, Alastair Potts, Iris Bardel-Kahr.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
