<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en-US">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Estimating phylogenetic trees with phangorn • phangorn</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Estimating phylogenetic trees with phangorn">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">phangorn</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.12.1.3</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/Trees.html">Estimating phylogenetic trees with phangorn</a></li>
    <li><a class="dropdown-item" href="../articles/Morphological.html">Phylogenetic trees from morphological data</a></li>
    <li><a class="dropdown-item" href="../articles/AdvancedFeatures.html">Markov models and transition rate matrices</a></li>
    <li><a class="dropdown-item" href="../articles/Ancestral.html">Ancestral Sequence Reconstruction</a></li>
    <li><a class="dropdown-item" href="../articles/Networx.html">Splits and Networx</a></li>
    <li><a class="dropdown-item" href="../articles/MLbyHand.html">Maximum likelihood by hand</a></li>
    <li><a class="dropdown-item" href="../articles/IntertwiningTreesAndNetworks.html">Intertwining phylogenetic trees and networks: R Example Script</a></li>
    <li><a class="dropdown-item" href="../articles/Parallel.html">Parallel computing with phangorn</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/KlausVigo/phangorn/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Estimating phylogenetic trees with phangorn</h1>
                        <h4 data-toc-skip class="author">Klaus Schliep,
Iris Bardel-Kahr</h4>
            <address class="author_afil">
      Graz University of Technology, University of
Graz<br><a class="author_email" href="mailto:#"></a><a href="mailto:klaus.schliep@gmail.com" class="email">klaus.schliep@gmail.com</a>
      </address>
                  
            <h4 data-toc-skip class="date">2026-01-20</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/KlausVigo/phangorn/blob/main/vignettes/Trees.Rmd" class="external-link"><code>vignettes/Trees.Rmd</code></a></small>
      <div class="d-none name"><code>Trees.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>These notes should enable the user to estimate phylogenetic trees
from alignment data with different methods using the
<code>phangorn</code> package <span class="citation">(Schliep
2011)</span> . Several functions of this <em>package</em> are also
described in more detail in <span class="citation">(Paradis
2012)</span>. For more theoretical background on all the methods see
e.g. <span class="citation">(Felsenstein 2004; Yang 2006)</span>. This
document illustrates some of the <em>package’s</em> features to estimate
phylogenetic trees using different reconstruction methods.</p>
</div>
<div class="section level2">
<h2 id="getting-started">Getting started<a class="anchor" aria-label="anchor" href="#getting-started"></a>
</h2>
<p>The first thing we have to do is to read in an alignment.
Unfortunately there exist many different file formats that alignments
can be stored in. In most cases, the function <code>read.phyDat</code>
is used to read in an alignment. In the <em>ape</em> package <span class="citation">(Paradis and Schliep 2019)</span> and
<em>phangorn</em>, there are several functions to read in alignments,
depending on the format of the data set (“nexus”, “phylip”, “fasta”) and
the kind of data (amino acid, nucleotides, morphological data). The
function <code>read.phyDat</code> calls these other functions and
transforms them into a <code>phyDat</code> object. For the specific
parameter settings available look in the help files of the function
<code>read.dna</code> (for phylip, fasta, clustal format),
<code>read.nexus.data</code> for nexus files. For amino acid data
additional <code>read.aa</code> is called. Morphological data will be
shown later in the vignette <em>Phylogenetic trees from morphological
data</em>.</p>
<p>We start our analysis loading the <em>phangorn</em> package and then
reading in an alignment.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/emmanuelparadis/ape" class="external-link">ape</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/KlausVigo/phangorn" class="external-link">phangorn</a></span><span class="op">)</span></span>
<span><span class="va">fdir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/trees"</span>, package <span class="op">=</span> <span class="st">"phangorn"</span><span class="op">)</span></span>
<span><span class="va">primates</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/read.phyDat.html">read.phyDat</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">fdir</span>, <span class="st">"primates.dna"</span><span class="op">)</span>,</span>
<span>                        format <span class="op">=</span> <span class="st">"interleaved"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="distance-based-methods">Distance based methods<a class="anchor" aria-label="anchor" href="#distance-based-methods"></a>
</h2>
<p>After reading in the nucleotide alignment we can build a first tree
with distance based methods. The function <code>dist.dna</code> from the
<em>ape</em> package computes distances for many DNA substitution
models, but to use the function <code>dist.dna</code>, we have to
transform the data to class DNAbin. The function <code>dist.ml</code>
from <em>phangorn</em> offers the substitution models “JC69” and “F81”
for DNA, and also common substitution models for amino acids
(e.g. “WAG”, “JTT”, “LG”, “Dayhoff”, “cpREV”, “mtmam”, “mtArt”, “MtZoa”
or “mtREV24”).</p>
<p>After constructing a distance matrix, we reconstruct a rooted tree
with UPGMA and alternatively an unrooted tree using Neighbor Joining
<span class="citation">(Saitou and Nei 1987; Studier and Keppler
1988)</span>. More distance methods like <code>fastme</code> are
available in the <em>ape</em> package.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dm</span>  <span class="op">&lt;-</span> <span class="fu"><a href="../reference/dist.hamming.html">dist.ml</a></span><span class="op">(</span><span class="va">primates</span><span class="op">)</span></span>
<span><span class="va">treeUPGMA</span>  <span class="op">&lt;-</span> <span class="fu"><a href="../reference/upgma.html">upgma</a></span><span class="op">(</span><span class="va">dm</span><span class="op">)</span></span>
<span><span class="va">treeNJ</span>  <span class="op">&lt;-</span> <span class="fu"><a href="../reference/NJ.html">NJ</a></span><span class="op">(</span><span class="va">dm</span><span class="op">)</span></span></code></pre></div>
<p>We can plot the trees <code>treeUPGMA</code> and <code>treeNJ</code>
with the commands:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">treeUPGMA</span>, main<span class="op">=</span><span class="st">"UPGMA"</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="Trees_files/figure-html/plot1-1.png" class="r-plt" alt="Rooted UPGMA tree." width="768"><p class="caption">
Rooted UPGMA tree.
</p>
</div>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">treeNJ</span>, <span class="st">"unrooted"</span>, main<span class="op">=</span><span class="st">"NJ"</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="Trees_files/figure-html/plot2-1.png" class="r-plt" alt="Unrooted NJ tree." width="768"><p class="caption">
Unrooted NJ tree.
</p>
</div>
<div class="section level3">
<h3 id="bootstrap">Bootstrap<a class="anchor" aria-label="anchor" href="#bootstrap"></a>
</h3>
<p>To run the bootstrap we first need to write a function which computes
a tree from an alignment. So we first need to compute a distance matrix
and afterwards compute the tree. We can then give this function to the
<code>bootstrap.phyDat</code> function.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fun</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="../reference/upgma.html">upgma</a></span><span class="op">(</span><span class="fu"><a href="../reference/dist.hamming.html">dist.ml</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">bs_upgma</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/bootstrap.pml.html">bootstrap.phyDat</a></span><span class="op">(</span><span class="va">primates</span>,  <span class="va">fun</span><span class="op">)</span></span></code></pre></div>
<p>With the new syntax of R 4.1 this can be written a bit shorter:</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a>bs_upgma <span class="ot">&lt;-</span> <span class="fu">bootstrap.phyDat</span>(primates,  \(x){<span class="fu">dist.ml</span>(x) <span class="sc">|&gt;</span> upgma})</span></code></pre></div>
<p>Finally, we can plot the tree with bootstrap values added:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotBS.html">plotBS</a></span><span class="op">(</span><span class="va">treeUPGMA</span>, <span class="va">bs_upgma</span>, main<span class="op">=</span><span class="st">"UPGMA"</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="Trees_files/figure-html/plot_bs-1.png" class="r-plt" alt="Rooted UPGMA tree." width="768"><p class="caption">
Rooted UPGMA tree.
</p>
</div>
<p>Distance based methods are very fast and we will use the UPGMA and NJ
tree as starting trees for the maximum parsimony and maximum likelihood
analyses.</p>
</div>
</div>
<div class="section level2">
<h2 id="parsimony">Parsimony<a class="anchor" aria-label="anchor" href="#parsimony"></a>
</h2>
<p>The function parsimony returns the parsimony score, that is the
minimum number of changes necessary to describe the data for a given
tree. We can compare the parsimony score for the two trees we computed
so far:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/parsimony.html">parsimony</a></span><span class="op">(</span><span class="va">treeUPGMA</span>, <span class="va">primates</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 751</span></span></code></pre>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/parsimony.html">parsimony</a></span><span class="op">(</span><span class="va">treeNJ</span>, <span class="va">primates</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 746</span></span></code></pre>
<p>The function most users want to use to infer phylogenies with MP
(maximum parsimony) is <code>pratchet</code>, an implementation of the
parsimony ratchet <span class="citation">(Nixon 1999)</span>. This
allows to escape local optima and find better trees than only performing
NNI / SPR rearrangements.</p>
<p>The current implementation is</p>
<ol style="list-style-type: decimal">
<li>Create a bootstrap data set
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>D</mi><mi>b</mi></msub><annotation encoding="application/x-tex">D_b</annotation></semantics></math>
from the original data set.</li>
<li>Take the current best tree and perform tree rearrangements on
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>D</mi><mi>b</mi></msub><annotation encoding="application/x-tex">D_b</annotation></semantics></math>
and save bootstrap tree as
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>T</mi><mi>b</mi></msub><annotation encoding="application/x-tex">T_b</annotation></semantics></math>.</li>
<li>Use
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>T</mi><mi>b</mi></msub><annotation encoding="application/x-tex">T_b</annotation></semantics></math>
and perform tree rearrangements on the original data set. If this tree
has a lower parsimony score than the currently best tree, replace
it.</li>
<li>Iterate 1:3 until either a given number of iteration is reached
(<code>minit</code>) or no improvements have been recorded for a number
of iterations (<code>k</code>).</li>
</ol>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">treeRatchet</span>  <span class="op">&lt;-</span> <span class="fu"><a href="../reference/parsimony.html">pratchet</a></span><span class="op">(</span><span class="va">primates</span>, trace <span class="op">=</span> <span class="fl">0</span>, minit<span class="op">=</span><span class="fl">100</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/parsimony.html">parsimony</a></span><span class="op">(</span><span class="va">treeRatchet</span>, <span class="va">primates</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 746</span></span></code></pre>
<p>Here we set the minimum iteration of the parsimony ratchet
(<code>minit</code>) to 100 iterations, the default number for
<code>k</code> is 10. As the ratchet implicitly performs bootstrap
resampling, we already computed some branch support, in our case with at
least 100 bootstrap iterations. The parameter <code>trace=0</code> tells
the function not write the current status to the console. The function
may return several best trees, but these trees have no branch length
assigned to them yet. Now let’s do this:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">treeRatchet</span>  <span class="op">&lt;-</span> <span class="fu"><a href="../reference/acctran.html">acctran</a></span><span class="op">(</span><span class="va">treeRatchet</span>, <span class="va">primates</span><span class="op">)</span></span></code></pre></div>
<p>After assigning edge weights, we prune away internal edges of length
<code>tol</code> (default = 1e-08), so our trees may contain
multifurcations.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">treeRatchet</span>  <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ape/man/multi2di.html" class="external-link">di2multi</a></span><span class="op">(</span><span class="va">treeRatchet</span><span class="op">)</span></span></code></pre></div>
<p>Some trees might have differed only between edges of length 0.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">inherits</a></span><span class="op">(</span><span class="va">treeRatchet</span>, <span class="st">"multiPhylo"</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="va">treeRatchet</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">treeRatchet</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p>As mentioned above, the parsimony ratchet implicitly performs a
bootstrap analysis (step 1). We make use of this and store the trees
which where visited. This allows us to add bootstrap support values to
the tree.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotBS.html">plotBS</a></span><span class="op">(</span><span class="fu"><a href="../reference/midpoint.html">midpoint</a></span><span class="op">(</span><span class="va">treeRatchet</span><span class="op">)</span>, type<span class="op">=</span><span class="st">"phylogram"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/ape/man/add.scale.bar.html" class="external-link">add.scale.bar</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="Trees_files/figure-html/midpoint-1.png" class="r-plt" alt="MP phylogeny with bootstrap values." width="768"><p class="caption">
MP phylogeny with bootstrap values.
</p>
</div>
<p>If <code>treeRatchet</code> is a list of trees, i.e. an object of
class <code>multiPhylo</code>, we can subset the i-th trees with
<code>treeRatchet[[i]]</code>.</p>
<p>While in most cases <code>pratchet</code> will be enough to use,
<code>phangorn</code> exports some function which might be useful.
<code>random.addition</code> computes random addition and can be used to
generate starting trees. The function <code>optim.parsimony</code>
performs tree rearrangements to find trees with a lower parsimony score.
The tree rearrangements implemented are nearest-neighbor interchanges
(NNI) and subtree pruning and regrafting (SPR). The latter so far only
works with the fitch algorithm.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">treeRA</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/parsimony.html">random.addition</a></span><span class="op">(</span><span class="va">primates</span><span class="op">)</span></span>
<span><span class="va">treeSPR</span>  <span class="op">&lt;-</span> <span class="fu"><a href="../reference/parsimony.html">optim.parsimony</a></span><span class="op">(</span><span class="va">treeRA</span>, <span class="va">primates</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Final p-score 750 after  2 nni operations</span></span></code></pre>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/parsimony.html">parsimony</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">treeRA</span>, <span class="va">treeSPR</span><span class="op">)</span>, <span class="va">primates</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 753 750</span></span></code></pre>
<div class="section level3">
<h3 id="branch-and-bound">Branch and bound<a class="anchor" aria-label="anchor" href="#branch-and-bound"></a>
</h3>
<p>For data sets with few species it is also possible to find all most
parsimonious trees using a branch and bound algorithm <span class="citation">(Hendy and Penny 1982)</span>. For data sets with more
than 10 taxa this can take a long time and depends strongly on how
“tree-like” the data is. And for more than 20-30 taxa this will take
almost forever.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">trees</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/bab.html">bab</a></span><span class="op">(</span><span class="va">primates</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span>,<span class="op">]</span>, trace<span class="op">=</span><span class="fl">0</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## 1 phylogenetic tree</span></span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="maximum-likelihood">Maximum likelihood<a class="anchor" aria-label="anchor" href="#maximum-likelihood"></a>
</h2>
<p>The last method we will describe in this vignette is Maximum
Likelihood (ML) as introduced by Felsenstein <span class="citation">(Felsenstein 1981)</span>.</p>
<div class="section level3">
<h3 id="model-selection">Model selection<a class="anchor" aria-label="anchor" href="#model-selection"></a>
</h3>
<p>Usually, as a first step, we will try to find the best fitting model.
For this we use the function <code>modelTest</code> to compare different
nucleotide or protein models with the AIC, AICc or BIC, similar to
popular programs ModelTest and ProtTest <span class="citation">(D.
Posada and Crandall 1998; David Posada 2008; Abascal, Zardoya, and
Posada 2005)</span>. By default available nucleotide or amino acid
models are compared.</p>
<p>The Vignette <em>Markov models and transition rate matrices</em>
gives further background on those models, how they are estimated and how
you can work with them.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mt</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/modelTest.html">modelTest</a></span><span class="op">(</span><span class="va">primates</span><span class="op">)</span></span></code></pre></div>
<p>It’s also possible to only select some common models:</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mt</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/modelTest.html">modelTest</a></span><span class="op">(</span><span class="va">primates</span>, model<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"JC"</span>, <span class="st">"F81"</span>, <span class="st">"K80"</span>, <span class="st">"HKY"</span>, <span class="st">"SYM"</span>, <span class="st">"GTR"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>The results of <code>modelTest</code> is illustrated in following
table:</p>
<table class="table">
<thead><tr class="header">
<th align="left">Model</th>
<th align="right">df</th>
<th align="right">logLik</th>
<th align="right">AIC</th>
<th align="right">AICw</th>
<th align="right">AICc</th>
<th align="right">AICcw</th>
<th align="right">BIC</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">JC</td>
<td align="right">25</td>
<td align="right">-3068.42</td>
<td align="right">6186.83</td>
<td align="right">0.00</td>
<td align="right">6193.15</td>
<td align="right">0.00</td>
<td align="right">6273.00</td>
</tr>
<tr class="even">
<td align="left">JC+I</td>
<td align="right">26</td>
<td align="right">-3062.63</td>
<td align="right">6177.26</td>
<td align="right">0.00</td>
<td align="right">6184.10</td>
<td align="right">0.00</td>
<td align="right">6266.87</td>
</tr>
<tr class="odd">
<td align="left">JC+G(4)</td>
<td align="right">26</td>
<td align="right">-3066.92</td>
<td align="right">6185.83</td>
<td align="right">0.00</td>
<td align="right">6192.68</td>
<td align="right">0.00</td>
<td align="right">6275.45</td>
</tr>
<tr class="even">
<td align="left">JC+G(4)+I</td>
<td align="right">27</td>
<td align="right">-3062.63</td>
<td align="right">6179.27</td>
<td align="right">0.00</td>
<td align="right">6186.68</td>
<td align="right">0.00</td>
<td align="right">6272.33</td>
</tr>
<tr class="odd">
<td align="left">F81</td>
<td align="right">28</td>
<td align="right">-2918.17</td>
<td align="right">5892.33</td>
<td align="right">0.00</td>
<td align="right">5900.33</td>
<td align="right">0.00</td>
<td align="right">5988.84</td>
</tr>
<tr class="even">
<td align="left">F81+I</td>
<td align="right">29</td>
<td align="right">-2909.12</td>
<td align="right">5876.24</td>
<td align="right">0.00</td>
<td align="right">5884.85</td>
<td align="right">0.00</td>
<td align="right">5976.20</td>
</tr>
<tr class="odd">
<td align="left">F81+G(4)</td>
<td align="right">29</td>
<td align="right">-2912.56</td>
<td align="right">5883.12</td>
<td align="right">0.00</td>
<td align="right">5891.73</td>
<td align="right">0.00</td>
<td align="right">5983.07</td>
</tr>
<tr class="even">
<td align="left">F81+G(4)+I</td>
<td align="right">30</td>
<td align="right">-2908.52</td>
<td align="right">5877.04</td>
<td align="right">0.00</td>
<td align="right">5886.29</td>
<td align="right">0.00</td>
<td align="right">5980.44</td>
</tr>
<tr class="odd">
<td align="left">K80</td>
<td align="right">26</td>
<td align="right">-2952.94</td>
<td align="right">5957.89</td>
<td align="right">0.00</td>
<td align="right">5964.73</td>
<td align="right">0.00</td>
<td align="right">6047.50</td>
</tr>
<tr class="even">
<td align="left">K80+I</td>
<td align="right">27</td>
<td align="right">-2944.51</td>
<td align="right">5943.02</td>
<td align="right">0.00</td>
<td align="right">5950.43</td>
<td align="right">0.00</td>
<td align="right">6036.08</td>
</tr>
<tr class="odd">
<td align="left">K80+G(4)</td>
<td align="right">27</td>
<td align="right">-2944.76</td>
<td align="right">5943.53</td>
<td align="right">0.00</td>
<td align="right">5950.94</td>
<td align="right">0.00</td>
<td align="right">6036.59</td>
</tr>
<tr class="even">
<td align="left">K80+G(4)+I</td>
<td align="right">28</td>
<td align="right">-2942.34</td>
<td align="right">5940.68</td>
<td align="right">0.00</td>
<td align="right">5948.68</td>
<td align="right">0.00</td>
<td align="right">6037.19</td>
</tr>
<tr class="odd">
<td align="left">HKY</td>
<td align="right">29</td>
<td align="right">-2624.93</td>
<td align="right">5307.86</td>
<td align="right">0.00</td>
<td align="right">5316.47</td>
<td align="right">0.00</td>
<td align="right">5407.82</td>
</tr>
<tr class="even">
<td align="left">HKY+I</td>
<td align="right">30</td>
<td align="right">-2621.11</td>
<td align="right">5302.22</td>
<td align="right">0.00</td>
<td align="right">5311.48</td>
<td align="right">0.00</td>
<td align="right">5405.63</td>
</tr>
<tr class="odd">
<td align="left">HKY+G(4)</td>
<td align="right">30</td>
<td align="right">-2612.60</td>
<td align="right">5285.19</td>
<td align="right">0.18</td>
<td align="right">5294.45</td>
<td align="right">0.46</td>
<td align="right">5388.60</td>
</tr>
<tr class="even">
<td align="left">HKY+G(4)+I</td>
<td align="right">31</td>
<td align="right">-2612.44</td>
<td align="right">5286.89</td>
<td align="right">0.08</td>
<td align="right">5296.81</td>
<td align="right">0.14</td>
<td align="right">5393.74</td>
</tr>
<tr class="odd">
<td align="left">SYM</td>
<td align="right">30</td>
<td align="right">-2813.90</td>
<td align="right">5687.79</td>
<td align="right">0.00</td>
<td align="right">5697.05</td>
<td align="right">0.00</td>
<td align="right">5791.19</td>
</tr>
<tr class="even">
<td align="left">SYM+I</td>
<td align="right">31</td>
<td align="right">-2811.73</td>
<td align="right">5685.46</td>
<td align="right">0.00</td>
<td align="right">5695.38</td>
<td align="right">0.00</td>
<td align="right">5792.31</td>
</tr>
<tr class="odd">
<td align="left">SYM+G(4)</td>
<td align="right">31</td>
<td align="right">-2804.68</td>
<td align="right">5671.36</td>
<td align="right">0.00</td>
<td align="right">5681.28</td>
<td align="right">0.00</td>
<td align="right">5778.20</td>
</tr>
<tr class="even">
<td align="left">SYM+G(4)+I</td>
<td align="right">32</td>
<td align="right">-2804.67</td>
<td align="right">5673.34</td>
<td align="right">0.00</td>
<td align="right">5683.95</td>
<td align="right">0.00</td>
<td align="right">5783.63</td>
</tr>
<tr class="odd">
<td align="left">GTR</td>
<td align="right">33</td>
<td align="right">-2618.46</td>
<td align="right">5302.92</td>
<td align="right">0.00</td>
<td align="right">5314.25</td>
<td align="right">0.00</td>
<td align="right">5416.66</td>
</tr>
<tr class="even">
<td align="left">GTR+I</td>
<td align="right">34</td>
<td align="right">-2613.53</td>
<td align="right">5295.06</td>
<td align="right">0.00</td>
<td align="right">5307.14</td>
<td align="right">0.00</td>
<td align="right">5412.24</td>
</tr>
<tr class="odd">
<td align="left">GTR+G(4)</td>
<td align="right">34</td>
<td align="right">-2607.65</td>
<td align="right">5283.30</td>
<td align="right">0.47</td>
<td align="right">5295.38</td>
<td align="right">0.29</td>
<td align="right">5400.48</td>
</tr>
<tr class="even">
<td align="left">GTR+G(4)+I</td>
<td align="right">35</td>
<td align="right">-2607.21</td>
<td align="right">5284.43</td>
<td align="right">0.27</td>
<td align="right">5297.28</td>
<td align="right">0.11</td>
<td align="right">5405.06</td>
</tr>
</tbody>
</table>
<p>To speed computations up the thresholds for the optimizations in
<code>modelTest</code> are not as strict as for <code>optim.pml</code>
(shown in the coming vignettes) and no tree rearrangements are
performed, which is the most time consuming part of the optimizing
process. As <code>modelTest</code> computes and optimizes a lot of
models it would be a waste of computer time not to save these results.
The results are saved as call together with the optimized trees in an
environment and the function <code>as.pml</code> evaluates this call to
get a <code>pml</code> object back to use for further optimization or
analysis. This can either be done for a specific model, or for a
specific criterion.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/pml.html">as.pml</a></span><span class="op">(</span><span class="va">mt</span>, <span class="st">"HKY+G(4)+I"</span><span class="op">)</span></span>
<span><span class="va">fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/pml.html">as.pml</a></span><span class="op">(</span><span class="va">mt</span>, <span class="st">"BIC"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="conducting-a-ml-tree">Conducting a ML tree<a class="anchor" aria-label="anchor" href="#conducting-a-ml-tree"></a>
</h3>
<p>To simplify the workflow, we can give the result of
<code>modelTest</code> to the function <code>pml_bb</code> and optimize
the parameters taking the best model according to BIC. Ultrafast
bootstrapping <span class="citation">(Minh, Nguyen, and Haeseler
2013)</span> is conducted automatically if the default
<code>rearrangements="stochastic"</code> is used. If
<code>rearrangements="NNI"</code> is used, no bootstrapping is
conducted.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fit_mt</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/pml_bb.html">pml_bb</a></span><span class="op">(</span><span class="va">mt</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fit_mt</span></span></code></pre></div>
<pre><code><span><span class="co">## model: HKY+G(4) </span></span>
<span><span class="co">## loglikelihood: -2615.014 </span></span>
<span><span class="co">## unconstrained loglikelihood: -1230.335 </span></span>
<span><span class="co">## Total tree length: 43.77113 </span></span>
<span><span class="co">## Model of rate heterogeneity: Discrete gamma model</span></span>
<span><span class="co">## Number of rate categories: 4 </span></span>
<span><span class="co">## Shape parameter: 2.272865 </span></span>
<span><span class="co">##        Rate Proportion</span></span>
<span><span class="co">## 1 0.3238072       0.25</span></span>
<span><span class="co">## 2 0.6812111       0.25</span></span>
<span><span class="co">## 3 1.0755269       0.25</span></span>
<span><span class="co">## 4 1.9194548       0.25</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Rates:</span></span>
<span><span class="co">## a &lt;-&gt; c : 1 </span></span>
<span><span class="co">## a &lt;-&gt; g : 54.5506 </span></span>
<span><span class="co">## a &lt;-&gt; t : 1 </span></span>
<span><span class="co">## c &lt;-&gt; g : 1 </span></span>
<span><span class="co">## c &lt;-&gt; t : 54.5506 </span></span>
<span><span class="co">## g &lt;-&gt; t : 1 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Base frequencies:  </span></span>
<span><span class="co">##          a          c          g          t </span></span>
<span><span class="co">## 0.37480751 0.40160148 0.03911303 0.18447798</span></span></code></pre>
<p>We can also use <code>pml_bb</code> with a defined model to infer a
phylogenetic tree.</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fitGTR</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/pml_bb.html">pml_bb</a></span><span class="op">(</span><span class="va">primates</span>, model<span class="op">=</span><span class="st">"GTR+G(4)+I"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="bootstrap-1">Bootstrap<a class="anchor" aria-label="anchor" href="#bootstrap-1"></a>
</h3>
<p>If we instead want to conduct standard bootstrapping <span class="citation">(Felsenstein 1985; Penny and Hendy 1985)</span>, we can
do so with the function <code>bootstrap.pml</code>:</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">bs</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/bootstrap.pml.html">bootstrap.pml</a></span><span class="op">(</span><span class="va">fit_mt</span>, bs<span class="op">=</span><span class="fl">100</span>, optNni<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p>Now we can plot the tree with the bootstrap support values on the
edges and compare the standard bootstrap values to the ultrafast
bootstrap values. With the function <code>plotBS</code> it is not only
possible to plot these two, but also the transfer bootstraps <span class="citation">(Lemoine et al. 2018)</span> which are especially
useful for large data sets.</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotBS.html">plotBS</a></span><span class="op">(</span><span class="fu"><a href="../reference/midpoint.html">midpoint</a></span><span class="op">(</span><span class="va">fit_mt</span><span class="op">$</span><span class="va">tree</span><span class="op">)</span>, p <span class="op">=</span> <span class="fl">.5</span>, type<span class="op">=</span><span class="st">"p"</span>, digits<span class="op">=</span><span class="fl">2</span>, main<span class="op">=</span><span class="st">"Ultrafast bootstrap"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/plotBS.html">plotBS</a></span><span class="op">(</span><span class="fu"><a href="../reference/midpoint.html">midpoint</a></span><span class="op">(</span><span class="va">fit_mt</span><span class="op">$</span><span class="va">tree</span><span class="op">)</span>, <span class="va">bs</span>, p <span class="op">=</span> <span class="fl">50</span>, type<span class="op">=</span><span class="st">"p"</span>, main<span class="op">=</span><span class="st">"Standard bootstrap"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/plotBS.html">plotBS</a></span><span class="op">(</span><span class="fu"><a href="../reference/midpoint.html">midpoint</a></span><span class="op">(</span><span class="va">fit_mt</span><span class="op">$</span><span class="va">tree</span><span class="op">)</span>, <span class="va">bs</span>, p <span class="op">=</span> <span class="fl">50</span>, type<span class="op">=</span><span class="st">"p"</span>, digits<span class="op">=</span><span class="fl">0</span>, method <span class="op">=</span> <span class="st">"TBE"</span>, main<span class="op">=</span><span class="st">"Transfer bootstrap"</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="Trees_files/figure-html/plotBS_ultrafast_bs-1.png" class="r-plt" alt="Unrooted tree (midpoint rooted) with ultrafast, standard and transfer bootstrap support values." width="33%"><img src="Trees_files/figure-html/plotBS_ultrafast_bs-2.png" class="r-plt" alt="Unrooted tree (midpoint rooted) with ultrafast, standard and transfer bootstrap support values." width="33%"><img src="Trees_files/figure-html/plotBS_ultrafast_bs-3.png" class="r-plt" alt="Unrooted tree (midpoint rooted) with ultrafast, standard and transfer bootstrap support values." width="33%"><p class="caption">
Unrooted tree (midpoint rooted) with ultrafast, standard and transfer
bootstrap support values.
</p>
</div>
<p>If we want to assign the standard or transfer bootstrap values to the
node labels in our tree instead of plotting it (e.g. to export the tree
somewhere else), <code>plotBS</code> gives that option with
<code>type = "n"</code>:</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># assigning standard bootstrap values to our tree; this is the default method</span></span>
<span><span class="va">tree_stdbs</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plotBS.html">plotBS</a></span><span class="op">(</span><span class="va">fit_mt</span><span class="op">$</span><span class="va">tree</span>, <span class="va">bs</span>, type <span class="op">=</span> <span class="st">"n"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># assigning transfer bootstrap values to our tree</span></span>
<span><span class="va">tree_tfbs</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plotBS.html">plotBS</a></span><span class="op">(</span><span class="va">fit_mt</span><span class="op">$</span><span class="va">tree</span>, <span class="va">bs</span>, type <span class="op">=</span> <span class="st">"n"</span>, method <span class="op">=</span> <span class="st">"TBE"</span><span class="op">)</span></span></code></pre></div>
<p>It is also possible to look at <code>consensusNet</code> to identify
potential conflict.</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cnet</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/consensusNet.html">consensusNet</a></span><span class="op">(</span><span class="va">bs</span>, p<span class="op">=</span><span class="fl">0.2</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">cnet</span>, show.edge.label<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="Trees_files/figure-html/ConsensusNet-1.png" class="r-plt" alt="ConsensusNet from the standard bootstrap sample." width="768"><p class="caption">
ConsensusNet from the standard bootstrap sample.
</p>
</div>
<p>Several analyses, e.g.<code>bootstrap</code> and
<code>modelTest</code>, can be computationally demanding, but as
nowadays most computers have several cores, one can distribute the
computations using the <em>parallel</em> package. However, it is only
possible to use this approach if R is running from command line (“X11”),
but not using a GUI (for example “Aqua” on Macs) and unfortunately the
<em>parallel</em> package does not work at all under Windows.</p>
</div>
<div class="section level3">
<h3 id="exporting-a-tree">Exporting a tree<a class="anchor" aria-label="anchor" href="#exporting-a-tree"></a>
</h3>
<p>Now that we have our tree with bootstrap values, we can easily write
it to a file in <em>Newick</em>-format:</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># tree with ultrafast bootstrap values</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/ape/man/write.tree.html" class="external-link">write.tree</a></span><span class="op">(</span><span class="va">fit_mt</span><span class="op">$</span><span class="va">tree</span>, <span class="st">"primates.tree"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># tree with standard bootstrap values</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/ape/man/write.tree.html" class="external-link">write.tree</a></span><span class="op">(</span><span class="va">tree_stdbs</span>, <span class="st">"primates.tree"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># tree with transfer bootstrap values</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/ape/man/write.tree.html" class="external-link">write.tree</a></span><span class="op">(</span><span class="va">tree_tfbs</span>, <span class="st">"primates.tree"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="molecular-dating-with-a-strict-clock-for-ultrametric-and-tipdated-phylogenies">Molecular dating with a strict clock for ultrametric and tipdated
phylogenies<a class="anchor" aria-label="anchor" href="#molecular-dating-with-a-strict-clock-for-ultrametric-and-tipdated-phylogenies"></a>
</h3>
<p>When we assume a “molecular clock” phylogenies can be used to infer
divergence times <span class="citation">(Zuckerkandl and Pauling
1965)</span>. We implemented a strict clock as described in <span class="citation">(Felsenstein 2004)</span>, p. 266, allowing to infer
ultrametric and tip-dated phylogenies. The function <code>pml_bb</code>
ensures that the tree is ultrametric, or the constraints given by the
tip dates are fulfilled. That differs from the function
<code>optim.pml</code> where th tree supplied to the function has to
fulfill the constraints. In this case for an ultrametric starting tree
we can use an UPGMA or WPGMA tree.</p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fit_strict</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/pml_bb.html">pml_bb</a></span><span class="op">(</span><span class="va">primates</span>, model<span class="op">=</span><span class="st">"HKY+G(4)"</span>, method<span class="op">=</span><span class="st">"ultrametric"</span>,</span>
<span>                     rearrangement<span class="op">=</span><span class="st">"NNI"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">fit_strict</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="Trees_files/figure-html/plot_strict_primates-1.png" class="r-plt" alt="Ultrametric ML phylogeny." width="768"><p class="caption">
Ultrametric ML phylogeny.
</p>
</div>
<p>With <em>phangorn</em> we also can estimate tipdated phylogenies.
Here we use a H3N2 virus data set from <em>treetime</em> <span class="citation">(Sagulenko, Puller, and Neher 2018)</span> as an
example. Additionally to the alignment we also need to read in data
containing the dates of the tips.</p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fdir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/trees"</span>, package <span class="op">=</span> <span class="st">"phangorn"</span><span class="op">)</span></span>
<span><span class="va">tmp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.csv</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">fdir</span>,<span class="st">"H3N2_NA_20.csv"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">H3N2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/read.phyDat.html">read.phyDat</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">fdir</span>,<span class="st">"H3N2_NA_20.fasta"</span><span class="op">)</span>, format<span class="op">=</span><span class="st">"fasta"</span><span class="op">)</span></span></code></pre></div>
<p>We first process the sampling dates and create a named vector. The
<em>lubridate</em> package <span class="citation">(Grolemund and Wickham
2011)</span> comes in very handy dates in case one has to recode dates,
e.g. days and months.</p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dates</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/setNames.html" class="external-link">setNames</a></span><span class="op">(</span><span class="va">tmp</span><span class="op">$</span><span class="va">numdate_given</span>, <span class="va">tmp</span><span class="op">$</span><span class="va">name</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">dates</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##               A/Hawaii/02/2013|KF789866|05/28/2013|USA|12_13|H3N2/1-1409 </span></span>
<span><span class="co">##                                                                 2013.405 </span></span>
<span><span class="co">##         A/Boston/DOA2_107/2012|CY148382|11/01/2012|USA|12_13|H3N2/1-1409 </span></span>
<span><span class="co">##                                                                 2012.838 </span></span>
<span><span class="co">##               A/Oregon/15/2009|GQ895004|06/25/2009|USA|08_09|H3N2/1-1409 </span></span>
<span><span class="co">##                                                                 2009.482 </span></span>
<span><span class="co">## A/Hong_Kong/H090_695_V10/2009|CY115546|07/10/2009|Hong_Kong||H3N2/8-1416 </span></span>
<span><span class="co">##                                                                 2009.523 </span></span>
<span><span class="co">##            A/New_York/182/2000|CY001279|02/18/2000|USA|99_00|H3N2/1-1409 </span></span>
<span><span class="co">##                                                                 2000.134 </span></span>
<span><span class="co">##        A/Canterbury/58/2000|CY009150|09/05/2000|New_Zealand||H3N2/8-1416 </span></span>
<span><span class="co">##                                                                 2000.682</span></span></code></pre>
<p>Again we use the <code>pml_bb</code> function, which optimizes the
tree given the constraints of the <code>tip.dates</code> vector.</p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fit_td</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/pml_bb.html">pml_bb</a></span><span class="op">(</span><span class="va">H3N2</span>, model<span class="op">=</span><span class="st">"HKY+I"</span>, method<span class="op">=</span><span class="st">"tipdated"</span>, tip.dates<span class="op">=</span><span class="va">dates</span>, </span>
<span>               rearrangement<span class="op">=</span><span class="st">"NNI"</span><span class="op">)</span></span>
<span><span class="va">fit_td</span></span></code></pre></div>
<pre><code><span><span class="co">## model: HKY+I </span></span>
<span><span class="co">## loglikelihood: -3117.866 </span></span>
<span><span class="co">## unconstrained loglikelihood: -2883.911 </span></span>
<span><span class="co">## Total tree length: 52.13048 </span></span>
<span><span class="co">## Proportion of invariant sites: 0.6865064 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Rate: 0.002534975 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Rates:</span></span>
<span><span class="co">## a &lt;-&gt; c : 1 </span></span>
<span><span class="co">## a &lt;-&gt; g : 9.866433 </span></span>
<span><span class="co">## a &lt;-&gt; t : 1 </span></span>
<span><span class="co">## c &lt;-&gt; g : 1 </span></span>
<span><span class="co">## c &lt;-&gt; t : 9.866433 </span></span>
<span><span class="co">## g &lt;-&gt; t : 1 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Base frequencies:  </span></span>
<span><span class="co">##         a         c         g         t </span></span>
<span><span class="co">## 0.3097759 0.1928617 0.2376819 0.2596805 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Rate: 0.002534975</span></span></code></pre>
<p>While the loglikelihood is lower than for an unrooted tree, we have
to keep in mind that rooted trees use less parameters. In unrooted trees
we estimate one edge length parameter for each tree, for ultrametric
trees we only estimate a parameter for each internal node and for
tipdated trees we have one additional parameter for the rate. The rate
is here comparable to the slope fo the tip-to-root regression in
programs like <em>TempEst</em> <span class="citation">(Rambaut et al.
2016)</span>.</p>
<p>And at last we plot the tree with a timescale.</p>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">fit_td</span>, align.tip.label<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="Trees_files/figure-html/tipdated_plot-1.png" class="r-plt" alt="Tip dated ML phylogeny with time scale in years." width="768"><p class="caption">
Tip dated ML phylogeny with time scale in years.
</p>
</div>
<div style="page-break-after: always;"></div>
</div>
</div>
<div class="section level2">
<h2 class="unnumbered" id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<pre><code><span><span class="co">## R version 4.5.2 (2025-10-31)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">## Running under: Ubuntu 24.04.3 LTS</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.26.so;  LAPACK version 3.12.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">##  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">## [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: UTC</span></span>
<span><span class="co">## tzcode source: system (glibc)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">## [1] knitr_1.51        future_1.69.0     phangorn_2.12.1.3 ape_5.8-1        </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##  [1] Matrix_1.7-4        future.apply_1.20.1 jsonlite_2.0.0     </span></span>
<span><span class="co">##  [4] compiler_4.5.2      Rcpp_1.1.1          parallel_4.5.2     </span></span>
<span><span class="co">##  [7] jquerylib_0.1.4     globals_0.18.0      systemfonts_1.3.1  </span></span>
<span><span class="co">## [10] textshaping_1.0.4   yaml_2.3.12         fastmap_1.2.0      </span></span>
<span><span class="co">## [13] lattice_0.22-7      R6_2.6.1            generics_0.1.4     </span></span>
<span><span class="co">## [16] igraph_2.2.1        htmlwidgets_1.6.4   backports_1.5.0    </span></span>
<span><span class="co">## [19] checkmate_2.3.3     desc_1.4.3          osqp_0.6.3.3       </span></span>
<span><span class="co">## [22] bslib_0.9.0         rlang_1.1.7         fastmatch_1.1-8    </span></span>
<span><span class="co">## [25] cachem_1.1.0        xfun_0.56           fs_1.6.6           </span></span>
<span><span class="co">## [28] sass_0.4.10         otel_0.2.0          cli_3.6.5          </span></span>
<span><span class="co">## [31] pkgdown_2.2.0       magrittr_2.0.4      digest_0.6.39      </span></span>
<span><span class="co">## [34] grid_4.5.2          lifecycle_1.0.5     nlme_3.1-168       </span></span>
<span><span class="co">## [37] evaluate_1.0.5      listenv_0.10.0      codetools_0.2-20   </span></span>
<span><span class="co">## [40] ragg_1.5.0          parallelly_1.46.1   rmarkdown_2.30     </span></span>
<span><span class="co">## [43] pkgconfig_2.0.3     tools_4.5.2         htmltools_0.5.9</span></span></code></pre>
</div>
<div class="section level2">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent" entry-spacing="0">
<div id="ref-Abascal2005" class="csl-entry">
Abascal, Federico, Rafael Zardoya, and David Posada. 2005.
<span>“ProtTest: Selection of Best-Fit Models of Protein
Evolution.”</span> <em>Bioinformatics</em> 21 (9): 2104–5. <a href="https://doi.org/10.1093/bioinformatics/bti263" class="external-link">https://doi.org/10.1093/bioinformatics/bti263</a>.
</div>
<div id="ref-Felsenstein1981" class="csl-entry">
Felsenstein, Joseph. 1981. <span>“Evolutionary Trees from DNA Sequences:
A Maxumum Likelihood Approach.”</span> <em>Journal of Molecular
Evolution</em> 17: 368–76.
</div>
<div id="ref-Felsenstein1985" class="csl-entry">
———. 1985. <span>“Confidence Limits on Phylogenies. An Approach Using
the Bootstrap.”</span> <em>Evolution</em> 39: 783–91.
</div>
<div id="ref-Felsenstein2004" class="csl-entry">
———. 2004. <em>Inferring Phylogenies</em>. Sunderland: Sinauer
Associates.
</div>
<div id="ref-lubridate" class="csl-entry">
Grolemund, Garrett, and Hadley Wickham. 2011. <span>“Dates and Times
Made Easy with <span class="nocase">lubridate</span>.”</span>
<em>Journal of Statistical Software</em> 40 (3): 1–25. <a href="https://www.jstatsoft.org/v40/i03/" class="external-link">https://www.jstatsoft.org/v40/i03/</a>.
</div>
<div id="ref-Hendy1982" class="csl-entry">
Hendy, M. D., and D. Penny. 1982. <span>“Branch and Bound Algorithms to
Determine Minimal Evolutionary Trees.”</span> <em>Math. Biosc.</em> 59:
277–90.
</div>
<div id="ref-Lemoine2018" class="csl-entry">
Lemoine, Fréderic, J-B Domelevo Entfellner, Eduan Wilkinson, Damien
Correia, M Dávila Felipe, Tulio De Oliveira, and Olivier Gascuel. 2018.
<span>“Renewing Felsenstein’s Phylogenetic Bootstrap in the Era of Big
Data.”</span> <em>Nature</em> 556 (7702): 452–56.
</div>
<div id="ref-minh2013" class="csl-entry">
Minh, Bui Quang, Minh Anh Thi Nguyen, and Arndt von Haeseler. 2013.
<span>“Ultrafast Approximation for Phylogenetic Bootstrap.”</span>
<em>Molecular Biology and Evolution</em> 30 (5): 1188–95.
</div>
<div id="ref-Nixon1999" class="csl-entry">
Nixon, K. 1999. <span>“The Parsimony Ratchet, a New Method for Rapid
Rarsimony Analysis.”</span> <em>Cladistics</em> 15: 407–14.
</div>
<div id="ref-Paradis2012" class="csl-entry">
Paradis, Emmanuel. 2012. <em>Analysis of Phylogenetics and Evolution
with r</em>. Second. New York: Springer.
</div>
<div id="ref-Paradis2018" class="csl-entry">
Paradis, Emmanuel, and Klaus Schliep. 2019. <span>“Ape 5.0: An
Environment for Modern Phylogenetics and Evolutionary Analyses in
r.”</span> <em>Bioinformatics</em> 35 (3): 526–28. <a href="https://doi.org/10.1093/bioinformatics/bty633" class="external-link">https://doi.org/10.1093/bioinformatics/bty633</a>.
</div>
<div id="ref-Penny1985" class="csl-entry">
Penny, D., and M. D. Hendy. 1985. <span>“Testing Methods Evolutionary
Tree Construction.”</span> <em>Cladistics</em> 1: 266–78.
</div>
<div id="ref-Posada2008" class="csl-entry">
Posada, David. 2008. <span>“<span class="nocase">jModelTest</span>:
Phylogenetic Model Averaging.”</span> <em>Molecular Biology and
Evolution</em> 25 (7): 1253–56. <a href="https://doi.org/10.1093/molbev/msn083" class="external-link">https://doi.org/10.1093/molbev/msn083</a>.
</div>
<div id="ref-Posada1998" class="csl-entry">
Posada, D., and K. A. Crandall. 1998. <span>“<span>MODELTEST</span>:
Testing the Model of <span>DNA</span> Substitution.”</span>
<em>Bioinformatics</em> 14 (9): 817–18.
</div>
<div id="ref-TempEst" class="csl-entry">
Rambaut, Andrew, Tommy T. Lam, Luiz Max Carvalho, and Oliver G. Pybus.
2016. <span>“<span class="nocase">Exploring the temporal structure of
heterochronous sequences using TempEst (formerly
Path-O-Gen)</span>.”</span> <em>Virus Evolution</em> 2 (1). <a href="https://doi.org/10.1093/ve/vew007" class="external-link">https://doi.org/10.1093/ve/vew007</a>.
</div>
<div id="ref-treetime" class="csl-entry">
Sagulenko, Pavel, Vadim Puller, and Richard A Neher. 2018.
<span>“TreeTime: Maximum-Likelihood Phylodynamic Analysis.”</span>
<em>Virus Evolution</em> 4 (1): vex042.
</div>
<div id="ref-Saitou1987" class="csl-entry">
Saitou, N., and M. Nei. 1987. <span>“The Neighbor-Joining Method - a New
Method for Reconstructing Phylogenetic Trees.”</span> <em>Molecular
Biology and Evolution</em> 4 (4): 406–25.
</div>
<div id="ref-Schliep2011" class="csl-entry">
Schliep, Klaus Peter. 2011. <span>“Phangorn: Phylogenetic Analysis in
<span>R</span>.”</span> <em>Bioinformatics</em> 27 (4): 592–93. <a href="https://doi.org/10.1093/bioinformatics/btq706" class="external-link">https://doi.org/10.1093/bioinformatics/btq706</a>.
</div>
<div id="ref-Studier1988" class="csl-entry">
Studier, J. A., and K. J. Keppler. 1988. <span>“A Note on the
Neighbor-Joining Algorithm of Saitou and Nei.”</span> <em>Molecular
Biology and Evolution</em> 5 (6): 729–31.
</div>
<div id="ref-Yang2006" class="csl-entry">
Yang, Ziheng. 2006. <em>Computational Molecular Evolution</em>. Oxford:
Oxford University Press.
</div>
<div id="ref-Zuckerkandl1965" class="csl-entry">
Zuckerkandl, Emile, and Linus Pauling. 1965. <span>“Molecules as
Documents of Evolutionary History.”</span> <em>Journal of Theoretical
Biology</em> 8 (2): 357–66.
</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Klaus Schliep, Emmanuel Paradis, Leonardo de Oliveira Martins, Alastair Potts, Iris Bardel-Kahr.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

    </footer>
</div>





  </body>
</html>
